// @openrouter/sdk/esm/models/errors/badgatewayresponseerror.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 6264bb9f01d6
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as models from "../index.js";
import { OpenRouterError } from "./openroutererror.js";
/**
 * Bad Gateway - Provider/upstream API failure
 */
export class BadGatewayResponseError extends OpenRouterError {
    constructor(err, httpMeta) {
        const message = err.error?.message
            || `API error occurred: ${JSON.stringify(err)}`;
        super(message, httpMeta);
        this.data$ = err;
        this.error = err.error;
        if (err.userId != null)
            this.userId = err.userId;
        this.name = "BadGatewayResponseError";
    }
}
/** @internal */
export const BadGatewayResponseError$inboundSchema = z.object({
    error: models.BadGatewayResponseErrorData$inboundSchema,
    user_id: z.nullable(z.string()).optional(),
    request$: z.custom(x => x instanceof Request),
    response$: z.custom(x => x instanceof Response),
    body$: z.string(),
})
    .transform((v) => {
    const remapped = remap$(v, {
        "user_id": "userId",
    });
    return new BadGatewayResponseError(remapped, {
        request: v.request$,
        response: v.response$,
        body: v.body$,
    });
});
