// @openrouter/sdk/esm/models/errors/conflictresponseerror.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 3d49b17cca5f
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as models from "../index.js";
import { OpenRouterError } from "./openroutererror.js";
/**
 * Conflict - Resource conflict or concurrent modification
 */
export class ConflictResponseError extends OpenRouterError {
    constructor(err, httpMeta) {
        const message = err.error?.message
            || `API error occurred: ${JSON.stringify(err)}`;
        super(message, httpMeta);
        this.data$ = err;
        this.error = err.error;
        if (err.userId != null)
            this.userId = err.userId;
        this.name = "ConflictResponseError";
    }
}
/** @internal */
export const ConflictResponseError$inboundSchema = z.object({
    error: models.ConflictResponseErrorData$inboundSchema,
    user_id: z.nullable(z.string()).optional(),
    request$: z.custom(x => x instanceof Request),
    response$: z.custom(x => x instanceof Response),
    body$: z.string(),
})
    .transform((v) => {
    const remapped = remap$(v, {
        "user_id": "userId",
    });
    return new ConflictResponseError(remapped, {
        request: v.request$,
        response: v.response$,
        body: v.body$,
    });
});
