// @openrouter/sdk/esm/models/operations/createresponses.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 7ad611b27e9d
 */
import * as z from "zod/v4";
import { EventStream } from "../../lib/event-streams.js";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as models from "../index.js";
/** @internal */
export const CreateResponsesRequest$outboundSchema = z.object({
    httpReferer: z.string().optional(),
    appTitle: z.string().optional(),
    appCategories: z.string().optional(),
    responsesRequest: models.ResponsesRequest$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        httpReferer: "HTTP-Referer",
        responsesRequest: "ResponsesRequest",
    });
});
export function createResponsesRequestToJSON(createResponsesRequest) {
    return JSON.stringify(CreateResponsesRequest$outboundSchema.parse(createResponsesRequest));
}
/** @internal */
export const CreateResponsesResponseBody$inboundSchema = z.object({
    data: z.string().transform((v, ctx) => {
        try {
            return JSON.parse(v);
        }
        catch (err) {
            ctx.addIssue({
                input: v,
                code: "custom",
                message: `malformed json: ${err}`,
            });
            return z.NEVER;
        }
    }).pipe(models.StreamEvents$inboundSchema),
});
export function createResponsesResponseBodyFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CreateResponsesResponseBody$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CreateResponsesResponseBody' from JSON`);
}
/** @internal */
export const CreateResponsesResponse$inboundSchema = z.union([
    models.OpenResponsesResult$inboundSchema,
    z.custom(x => x instanceof ReadableStream)
        .transform(stream => {
        return new EventStream(stream, rawEvent => {
            if (rawEvent.data === "[DONE]")
                return { done: true, value: undefined };
            return {
                done: false,
                value: z.lazy(() => CreateResponsesResponseBody$inboundSchema).parse(rawEvent)?.data,
            };
        });
    }),
]);
export function createResponsesResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CreateResponsesResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CreateResponsesResponse' from JSON`);
}
