// @openrouter/sdk/esm/models/operations/createembeddings.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 7c210337fdff
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as openEnums from "../../types/enums.js";
import * as models from "../index.js";
/**
 * The format of the output embeddings
 */
export const EncodingFormat = {
    Float: "float",
    Base64: "base64",
};
export const ObjectEmbedding = {
    Embedding: "embedding",
};
export const ObjectT = {
    List: "list",
};
/** @internal */
export const EncodingFormat$outboundSchema = openEnums.outboundSchema(EncodingFormat);
/** @internal */
export const ImageUrl$outboundSchema = z
    .object({
    url: z.string(),
});
export function imageUrlToJSON(imageUrl) {
    return JSON.stringify(ImageUrl$outboundSchema.parse(imageUrl));
}
/** @internal */
export const ContentImageURL$outboundSchema = z.object({
    imageUrl: z.lazy(() => ImageUrl$outboundSchema),
    type: z.literal("image_url"),
}).transform((v) => {
    return remap$(v, {
        imageUrl: "image_url",
    });
});
export function contentImageURLToJSON(contentImageURL) {
    return JSON.stringify(ContentImageURL$outboundSchema.parse(contentImageURL));
}
/** @internal */
export const ContentText$outboundSchema = z.object({
    text: z.string(),
    type: z.literal("text"),
});
export function contentTextToJSON(contentText) {
    return JSON.stringify(ContentText$outboundSchema.parse(contentText));
}
/** @internal */
export const Content$outboundSchema = z
    .union([
    z.lazy(() => ContentText$outboundSchema),
    z.lazy(() => ContentImageURL$outboundSchema),
]);
export function contentToJSON(content) {
    return JSON.stringify(Content$outboundSchema.parse(content));
}
/** @internal */
export const Input$outboundSchema = z.object({
    content: z.array(z.union([
        z.lazy(() => ContentText$outboundSchema),
        z.lazy(() => ContentImageURL$outboundSchema),
    ])),
});
export function inputToJSON(input) {
    return JSON.stringify(Input$outboundSchema.parse(input));
}
/** @internal */
export const InputUnion$outboundSchema = z.union([
    z.string(),
    z.array(z.string()),
    z.array(z.number()),
    z.array(z.array(z.number())),
    z.array(z.lazy(() => Input$outboundSchema)),
]);
export function inputUnionToJSON(inputUnion) {
    return JSON.stringify(InputUnion$outboundSchema.parse(inputUnion));
}
/** @internal */
export const CreateEmbeddingsRequestBody$outboundSchema = z.object({
    dimensions: z.int().optional(),
    encodingFormat: EncodingFormat$outboundSchema.optional(),
    input: z.union([
        z.string(),
        z.array(z.string()),
        z.array(z.number()),
        z.array(z.array(z.number())),
        z.array(z.lazy(() => Input$outboundSchema)),
    ]),
    inputType: z.string().optional(),
    model: z.string(),
    provider: z.nullable(models.ProviderPreferences$outboundSchema).optional(),
    user: z.string().optional(),
}).transform((v) => {
    return remap$(v, {
        encodingFormat: "encoding_format",
        inputType: "input_type",
    });
});
export function createEmbeddingsRequestBodyToJSON(createEmbeddingsRequestBody) {
    return JSON.stringify(CreateEmbeddingsRequestBody$outboundSchema.parse(createEmbeddingsRequestBody));
}
/** @internal */
export const CreateEmbeddingsRequest$outboundSchema = z.object({
    httpReferer: z.string().optional(),
    appTitle: z.string().optional(),
    appCategories: z.string().optional(),
    requestBody: z.lazy(() => CreateEmbeddingsRequestBody$outboundSchema),
}).transform((v) => {
    return remap$(v, {
        httpReferer: "HTTP-Referer",
        requestBody: "RequestBody",
    });
});
export function createEmbeddingsRequestToJSON(createEmbeddingsRequest) {
    return JSON.stringify(CreateEmbeddingsRequest$outboundSchema.parse(createEmbeddingsRequest));
}
/** @internal */
export const Embedding$inboundSchema = z.union([
    z.array(z.number()),
    z.string(),
]);
export function embeddingFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Embedding$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Embedding' from JSON`);
}
/** @internal */
export const ObjectEmbedding$inboundSchema = z.enum(ObjectEmbedding);
/** @internal */
export const CreateEmbeddingsData$inboundSchema = z.object({
    embedding: z.union([z.array(z.number()), z.string()]),
    index: z.int().optional(),
    object: ObjectEmbedding$inboundSchema,
});
export function createEmbeddingsDataFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CreateEmbeddingsData$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CreateEmbeddingsData' from JSON`);
}
/** @internal */
export const ObjectT$inboundSchema = z.enum(ObjectT);
/** @internal */
export const PromptTokensDetails$inboundSchema = z.object({
    audio_tokens: z.int().optional(),
    image_tokens: z.int().optional(),
    text_tokens: z.int().optional(),
    video_tokens: z.int().optional(),
}).transform((v) => {
    return remap$(v, {
        "audio_tokens": "audioTokens",
        "image_tokens": "imageTokens",
        "text_tokens": "textTokens",
        "video_tokens": "videoTokens",
    });
});
export function promptTokensDetailsFromJSON(jsonString) {
    return safeParse(jsonString, (x) => PromptTokensDetails$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'PromptTokensDetails' from JSON`);
}
/** @internal */
export const CreateEmbeddingsUsage$inboundSchema = z.object({
    cost: z.number().optional(),
    prompt_tokens: z.int(),
    prompt_tokens_details: z.lazy(() => PromptTokensDetails$inboundSchema)
        .optional(),
    total_tokens: z.int(),
}).transform((v) => {
    return remap$(v, {
        "prompt_tokens": "promptTokens",
        "prompt_tokens_details": "promptTokensDetails",
        "total_tokens": "totalTokens",
    });
});
export function createEmbeddingsUsageFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CreateEmbeddingsUsage$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CreateEmbeddingsUsage' from JSON`);
}
/** @internal */
export const CreateEmbeddingsResponseBody$inboundSchema = z.object({
    data: z.array(z.lazy(() => CreateEmbeddingsData$inboundSchema)),
    id: z.string().optional(),
    model: z.string(),
    object: ObjectT$inboundSchema,
    usage: z.lazy(() => CreateEmbeddingsUsage$inboundSchema).optional(),
});
export function createEmbeddingsResponseBodyFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CreateEmbeddingsResponseBody$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CreateEmbeddingsResponseBody' from JSON`);
}
/** @internal */
export const CreateEmbeddingsResponse$inboundSchema = z.union([
    z.lazy(() => CreateEmbeddingsResponseBody$inboundSchema),
    z.string(),
]);
export function createEmbeddingsResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CreateEmbeddingsResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CreateEmbeddingsResponse' from JSON`);
}
