// @openrouter/sdk/esm/models/operations/creatererank.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: e1de55b94363
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as models from "../index.js";
/** @internal */
export const CreateRerankRequestBody$outboundSchema = z.object({
    documents: z.array(z.string()),
    model: z.string(),
    provider: z.nullable(models.ProviderPreferences$outboundSchema).optional(),
    query: z.string(),
    topN: z.int().optional(),
}).transform((v) => {
    return remap$(v, {
        topN: "top_n",
    });
});
export function createRerankRequestBodyToJSON(createRerankRequestBody) {
    return JSON.stringify(CreateRerankRequestBody$outboundSchema.parse(createRerankRequestBody));
}
/** @internal */
export const CreateRerankRequest$outboundSchema = z.object({
    httpReferer: z.string().optional(),
    appTitle: z.string().optional(),
    appCategories: z.string().optional(),
    requestBody: z.lazy(() => CreateRerankRequestBody$outboundSchema),
}).transform((v) => {
    return remap$(v, {
        httpReferer: "HTTP-Referer",
        requestBody: "RequestBody",
    });
});
export function createRerankRequestToJSON(createRerankRequest) {
    return JSON.stringify(CreateRerankRequest$outboundSchema.parse(createRerankRequest));
}
/** @internal */
export const Document$inboundSchema = z.object({
    text: z.string(),
});
export function documentFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Document$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Document' from JSON`);
}
/** @internal */
export const Result$inboundSchema = z.object({
    document: z.lazy(() => Document$inboundSchema),
    index: z.int(),
    relevance_score: z.number(),
}).transform((v) => {
    return remap$(v, {
        "relevance_score": "relevanceScore",
    });
});
export function resultFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Result$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Result' from JSON`);
}
/** @internal */
export const CreateRerankUsage$inboundSchema = z.object({
    cost: z.number().optional(),
    search_units: z.int().optional(),
    total_tokens: z.int().optional(),
}).transform((v) => {
    return remap$(v, {
        "search_units": "searchUnits",
        "total_tokens": "totalTokens",
    });
});
export function createRerankUsageFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CreateRerankUsage$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CreateRerankUsage' from JSON`);
}
/** @internal */
export const CreateRerankResponseBody$inboundSchema = z.object({
    id: z.string().optional(),
    model: z.string(),
    provider: z.string().optional(),
    results: z.array(z.lazy(() => Result$inboundSchema)),
    usage: z.lazy(() => CreateRerankUsage$inboundSchema).optional(),
});
export function createRerankResponseBodyFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CreateRerankResponseBody$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CreateRerankResponseBody' from JSON`);
}
/** @internal */
export const CreateRerankResponse$inboundSchema = z.union([z.lazy(() => CreateRerankResponseBody$inboundSchema), z.string()]);
export function createRerankResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CreateRerankResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CreateRerankResponse' from JSON`);
}
