// @openrouter/sdk/esm/models/operations/createtts.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: fd4cb8ec3732
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as openEnums from "../../types/enums.js";
/**
 * Audio output format
 */
export const ResponseFormat = {
    Mp3: "mp3",
    Pcm: "pcm",
};
/** @internal */
export const Options$outboundSchema = z
    .object({
    oneai: z.record(z.string(), z.nullable(z.any())).optional(),
    ai21: z.record(z.string(), z.nullable(z.any())).optional(),
    aionLabs: z.record(z.string(), z.nullable(z.any())).optional(),
    akashml: z.record(z.string(), z.nullable(z.any())).optional(),
    alibaba: z.record(z.string(), z.nullable(z.any())).optional(),
    amazonBedrock: z.record(z.string(), z.nullable(z.any())).optional(),
    amazonNova: z.record(z.string(), z.nullable(z.any())).optional(),
    ambient: z.record(z.string(), z.nullable(z.any())).optional(),
    anthropic: z.record(z.string(), z.nullable(z.any())).optional(),
    anyscale: z.record(z.string(), z.nullable(z.any())).optional(),
    arceeAi: z.record(z.string(), z.nullable(z.any())).optional(),
    atlasCloud: z.record(z.string(), z.nullable(z.any())).optional(),
    atoma: z.record(z.string(), z.nullable(z.any())).optional(),
    avian: z.record(z.string(), z.nullable(z.any())).optional(),
    azure: z.record(z.string(), z.nullable(z.any())).optional(),
    baidu: z.record(z.string(), z.nullable(z.any())).optional(),
    baseten: z.record(z.string(), z.nullable(z.any())).optional(),
    blackForestLabs: z.record(z.string(), z.nullable(z.any())).optional(),
    byteplus: z.record(z.string(), z.nullable(z.any())).optional(),
    centml: z.record(z.string(), z.nullable(z.any())).optional(),
    cerebras: z.record(z.string(), z.nullable(z.any())).optional(),
    chutes: z.record(z.string(), z.nullable(z.any())).optional(),
    cirrascale: z.record(z.string(), z.nullable(z.any())).optional(),
    clarifai: z.record(z.string(), z.nullable(z.any())).optional(),
    cloudflare: z.record(z.string(), z.nullable(z.any())).optional(),
    cohere: z.record(z.string(), z.nullable(z.any())).optional(),
    crofai: z.record(z.string(), z.nullable(z.any())).optional(),
    crusoe: z.record(z.string(), z.nullable(z.any())).optional(),
    deepinfra: z.record(z.string(), z.nullable(z.any())).optional(),
    deepseek: z.record(z.string(), z.nullable(z.any())).optional(),
    dekallm: z.record(z.string(), z.nullable(z.any())).optional(),
    enfer: z.record(z.string(), z.nullable(z.any())).optional(),
    fakeProvider: z.record(z.string(), z.nullable(z.any())).optional(),
    featherless: z.record(z.string(), z.nullable(z.any())).optional(),
    fireworks: z.record(z.string(), z.nullable(z.any())).optional(),
    friendli: z.record(z.string(), z.nullable(z.any())).optional(),
    gmicloud: z.record(z.string(), z.nullable(z.any())).optional(),
    googleAiStudio: z.record(z.string(), z.nullable(z.any())).optional(),
    googleVertex: z.record(z.string(), z.nullable(z.any())).optional(),
    gopomelo: z.record(z.string(), z.nullable(z.any())).optional(),
    groq: z.record(z.string(), z.nullable(z.any())).optional(),
    huggingface: z.record(z.string(), z.nullable(z.any())).optional(),
    hyperbolic: z.record(z.string(), z.nullable(z.any())).optional(),
    hyperbolicQuantized: z.record(z.string(), z.nullable(z.any())).optional(),
    inception: z.record(z.string(), z.nullable(z.any())).optional(),
    inceptron: z.record(z.string(), z.nullable(z.any())).optional(),
    inferenceNet: z.record(z.string(), z.nullable(z.any())).optional(),
    infermatic: z.record(z.string(), z.nullable(z.any())).optional(),
    inflection: z.record(z.string(), z.nullable(z.any())).optional(),
    inocloud: z.record(z.string(), z.nullable(z.any())).optional(),
    ioNet: z.record(z.string(), z.nullable(z.any())).optional(),
    ionstream: z.record(z.string(), z.nullable(z.any())).optional(),
    klusterai: z.record(z.string(), z.nullable(z.any())).optional(),
    lambda: z.record(z.string(), z.nullable(z.any())).optional(),
    lepton: z.record(z.string(), z.nullable(z.any())).optional(),
    liquid: z.record(z.string(), z.nullable(z.any())).optional(),
    lynn: z.record(z.string(), z.nullable(z.any())).optional(),
    lynnPrivate: z.record(z.string(), z.nullable(z.any())).optional(),
    mancer: z.record(z.string(), z.nullable(z.any())).optional(),
    mancerOld: z.record(z.string(), z.nullable(z.any())).optional(),
    mara: z.record(z.string(), z.nullable(z.any())).optional(),
    meta: z.record(z.string(), z.nullable(z.any())).optional(),
    minimax: z.record(z.string(), z.nullable(z.any())).optional(),
    mistral: z.record(z.string(), z.nullable(z.any())).optional(),
    modal: z.record(z.string(), z.nullable(z.any())).optional(),
    modelrun: z.record(z.string(), z.nullable(z.any())).optional(),
    modular: z.record(z.string(), z.nullable(z.any())).optional(),
    moonshotai: z.record(z.string(), z.nullable(z.any())).optional(),
    morph: z.record(z.string(), z.nullable(z.any())).optional(),
    ncompass: z.record(z.string(), z.nullable(z.any())).optional(),
    nebius: z.record(z.string(), z.nullable(z.any())).optional(),
    nextbit: z.record(z.string(), z.nullable(z.any())).optional(),
    nineteen: z.record(z.string(), z.nullable(z.any())).optional(),
    novita: z.record(z.string(), z.nullable(z.any())).optional(),
    nvidia: z.record(z.string(), z.nullable(z.any())).optional(),
    octoai: z.record(z.string(), z.nullable(z.any())).optional(),
    openInference: z.record(z.string(), z.nullable(z.any())).optional(),
    openai: z.record(z.string(), z.nullable(z.any())).optional(),
    parasail: z.record(z.string(), z.nullable(z.any())).optional(),
    perplexity: z.record(z.string(), z.nullable(z.any())).optional(),
    phala: z.record(z.string(), z.nullable(z.any())).optional(),
    recraft: z.record(z.string(), z.nullable(z.any())).optional(),
    recursal: z.record(z.string(), z.nullable(z.any())).optional(),
    reflection: z.record(z.string(), z.nullable(z.any())).optional(),
    reka: z.record(z.string(), z.nullable(z.any())).optional(),
    relace: z.record(z.string(), z.nullable(z.any())).optional(),
    replicate: z.record(z.string(), z.nullable(z.any())).optional(),
    sambanova: z.record(z.string(), z.nullable(z.any())).optional(),
    sambanovaCloaked: z.record(z.string(), z.nullable(z.any())).optional(),
    seed: z.record(z.string(), z.nullable(z.any())).optional(),
    sfCompute: z.record(z.string(), z.nullable(z.any())).optional(),
    siliconflow: z.record(z.string(), z.nullable(z.any())).optional(),
    sourceful: z.record(z.string(), z.nullable(z.any())).optional(),
    stealth: z.record(z.string(), z.nullable(z.any())).optional(),
    stepfun: z.record(z.string(), z.nullable(z.any())).optional(),
    streamlake: z.record(z.string(), z.nullable(z.any())).optional(),
    switchpoint: z.record(z.string(), z.nullable(z.any())).optional(),
    targon: z.record(z.string(), z.nullable(z.any())).optional(),
    together: z.record(z.string(), z.nullable(z.any())).optional(),
    togetherLite: z.record(z.string(), z.nullable(z.any())).optional(),
    ubicloud: z.record(z.string(), z.nullable(z.any())).optional(),
    upstage: z.record(z.string(), z.nullable(z.any())).optional(),
    venice: z.record(z.string(), z.nullable(z.any())).optional(),
    wandb: z.record(z.string(), z.nullable(z.any())).optional(),
    xai: z.record(z.string(), z.nullable(z.any())).optional(),
    xiaomi: z.record(z.string(), z.nullable(z.any())).optional(),
    zAi: z.record(z.string(), z.nullable(z.any())).optional(),
}).transform((v) => {
    return remap$(v, {
        oneai: "01ai",
        aionLabs: "aion-labs",
        amazonBedrock: "amazon-bedrock",
        amazonNova: "amazon-nova",
        arceeAi: "arcee-ai",
        atlasCloud: "atlas-cloud",
        blackForestLabs: "black-forest-labs",
        fakeProvider: "fake-provider",
        googleAiStudio: "google-ai-studio",
        googleVertex: "google-vertex",
        hyperbolicQuantized: "hyperbolic-quantized",
        inferenceNet: "inference-net",
        ioNet: "io-net",
        lynnPrivate: "lynn-private",
        mancerOld: "mancer-old",
        openInference: "open-inference",
        sambanovaCloaked: "sambanova-cloaked",
        sfCompute: "sf-compute",
        togetherLite: "together-lite",
        zAi: "z-ai",
    });
});
export function optionsToJSON(options) {
    return JSON.stringify(Options$outboundSchema.parse(options));
}
/** @internal */
export const Provider$outboundSchema = z
    .object({
    options: z.lazy(() => Options$outboundSchema).optional(),
});
export function providerToJSON(provider) {
    return JSON.stringify(Provider$outboundSchema.parse(provider));
}
/** @internal */
export const ResponseFormat$outboundSchema = openEnums.outboundSchema(ResponseFormat);
/** @internal */
export const CreateTtsRequestBody$outboundSchema = z.object({
    input: z.string(),
    model: z.string(),
    provider: z.lazy(() => Provider$outboundSchema).optional(),
    responseFormat: ResponseFormat$outboundSchema.default("pcm"),
    speed: z.number().optional(),
    voice: z.string(),
}).transform((v) => {
    return remap$(v, {
        responseFormat: "response_format",
    });
});
export function createTtsRequestBodyToJSON(createTtsRequestBody) {
    return JSON.stringify(CreateTtsRequestBody$outboundSchema.parse(createTtsRequestBody));
}
/** @internal */
export const CreateTtsRequest$outboundSchema = z.object({
    httpReferer: z.string().optional(),
    appTitle: z.string().optional(),
    appCategories: z.string().optional(),
    requestBody: z.lazy(() => CreateTtsRequestBody$outboundSchema),
}).transform((v) => {
    return remap$(v, {
        httpReferer: "HTTP-Referer",
        requestBody: "RequestBody",
    });
});
export function createTtsRequestToJSON(createTtsRequest) {
    return JSON.stringify(CreateTtsRequest$outboundSchema.parse(createTtsRequest));
}
