// @openrouter/sdk/esm/models/operations/createauthkeyscode.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 1443f6afbf40
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as openEnums from "../../types/enums.js";
/**
 * The method used to generate the code challenge
 */
export const CreateAuthKeysCodeCodeChallengeMethod = {
    S256: "S256",
    Plain: "plain",
};
/**
 * Optional credit limit reset interval. When set, the credit limit resets on this interval.
 */
export const UsageLimitType = {
    Daily: "daily",
    Weekly: "weekly",
    Monthly: "monthly",
};
/** @internal */
export const CreateAuthKeysCodeCodeChallengeMethod$outboundSchema = openEnums.outboundSchema(CreateAuthKeysCodeCodeChallengeMethod);
/** @internal */
export const UsageLimitType$outboundSchema = openEnums.outboundSchema(UsageLimitType);
/** @internal */
export const CreateAuthKeysCodeRequestBody$outboundSchema = z.object({
    callbackUrl: z.string(),
    codeChallenge: z.string().optional(),
    codeChallengeMethod: CreateAuthKeysCodeCodeChallengeMethod$outboundSchema
        .optional(),
    expiresAt: z.nullable(z.date().transform(v => v.toISOString())).optional(),
    keyLabel: z.string().optional(),
    limit: z.number().optional(),
    usageLimitType: UsageLimitType$outboundSchema.optional(),
}).transform((v) => {
    return remap$(v, {
        callbackUrl: "callback_url",
        codeChallenge: "code_challenge",
        codeChallengeMethod: "code_challenge_method",
        expiresAt: "expires_at",
        keyLabel: "key_label",
        usageLimitType: "usage_limit_type",
    });
});
export function createAuthKeysCodeRequestBodyToJSON(createAuthKeysCodeRequestBody) {
    return JSON.stringify(CreateAuthKeysCodeRequestBody$outboundSchema.parse(createAuthKeysCodeRequestBody));
}
/** @internal */
export const CreateAuthKeysCodeRequest$outboundSchema = z.object({
    httpReferer: z.string().optional(),
    appTitle: z.string().optional(),
    appCategories: z.string().optional(),
    requestBody: z.lazy(() => CreateAuthKeysCodeRequestBody$outboundSchema),
}).transform((v) => {
    return remap$(v, {
        httpReferer: "HTTP-Referer",
        requestBody: "RequestBody",
    });
});
export function createAuthKeysCodeRequestToJSON(createAuthKeysCodeRequest) {
    return JSON.stringify(CreateAuthKeysCodeRequest$outboundSchema.parse(createAuthKeysCodeRequest));
}
/** @internal */
export const CreateAuthKeysCodeData$inboundSchema = z.object({
    app_id: z.int(),
    created_at: z.string(),
    id: z.string(),
}).transform((v) => {
    return remap$(v, {
        "app_id": "appId",
        "created_at": "createdAt",
    });
});
export function createAuthKeysCodeDataFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CreateAuthKeysCodeData$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CreateAuthKeysCodeData' from JSON`);
}
/** @internal */
export const CreateAuthKeysCodeResponse$inboundSchema = z.object({
    data: z.lazy(() => CreateAuthKeysCodeData$inboundSchema),
});
export function createAuthKeysCodeResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CreateAuthKeysCodeResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CreateAuthKeysCodeResponse' from JSON`);
}
