// @openrouter/sdk/esm/models/operations/deletekeys.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: efa22339a3d6
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const DeleteKeysRequest$outboundSchema = z.object({
    httpReferer: z.string().optional(),
    appTitle: z.string().optional(),
    appCategories: z.string().optional(),
    hash: z.string(),
}).transform((v) => {
    return remap$(v, {
        httpReferer: "HTTP-Referer",
    });
});
export function deleteKeysRequestToJSON(deleteKeysRequest) {
    return JSON.stringify(DeleteKeysRequest$outboundSchema.parse(deleteKeysRequest));
}
/** @internal */
export const DeleteKeysResponse$inboundSchema = z.object({
    deleted: z.literal(true),
});
export function deleteKeysResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => DeleteKeysResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'DeleteKeysResponse' from JSON`);
}
