// @openrouter/sdk/esm/models/operations/exchangeauthcodeforapikey.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 440595fac2ad
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as openEnums from "../../types/enums.js";
/**
 * The method used to generate the code challenge
 */
export const ExchangeAuthCodeForAPIKeyCodeChallengeMethod = {
    S256: "S256",
    Plain: "plain",
};
/** @internal */
export const ExchangeAuthCodeForAPIKeyCodeChallengeMethod$outboundSchema = openEnums
    .outboundSchema(ExchangeAuthCodeForAPIKeyCodeChallengeMethod);
/** @internal */
export const ExchangeAuthCodeForAPIKeyRequestBody$outboundSchema = z.object({
    code: z.string(),
    codeChallengeMethod: z.nullable(ExchangeAuthCodeForAPIKeyCodeChallengeMethod$outboundSchema).optional(),
    codeVerifier: z.string().optional(),
}).transform((v) => {
    return remap$(v, {
        codeChallengeMethod: "code_challenge_method",
        codeVerifier: "code_verifier",
    });
});
export function exchangeAuthCodeForAPIKeyRequestBodyToJSON(exchangeAuthCodeForAPIKeyRequestBody) {
    return JSON.stringify(ExchangeAuthCodeForAPIKeyRequestBody$outboundSchema.parse(exchangeAuthCodeForAPIKeyRequestBody));
}
/** @internal */
export const ExchangeAuthCodeForAPIKeyRequest$outboundSchema = z.object({
    httpReferer: z.string().optional(),
    appTitle: z.string().optional(),
    appCategories: z.string().optional(),
    requestBody: z.lazy(() => ExchangeAuthCodeForAPIKeyRequestBody$outboundSchema),
}).transform((v) => {
    return remap$(v, {
        httpReferer: "HTTP-Referer",
        requestBody: "RequestBody",
    });
});
export function exchangeAuthCodeForAPIKeyRequestToJSON(exchangeAuthCodeForAPIKeyRequest) {
    return JSON.stringify(ExchangeAuthCodeForAPIKeyRequest$outboundSchema.parse(exchangeAuthCodeForAPIKeyRequest));
}
/** @internal */
export const ExchangeAuthCodeForAPIKeyResponse$inboundSchema = z.object({
    key: z.string(),
    user_id: z.nullable(z.string()),
}).transform((v) => {
    return remap$(v, {
        "user_id": "userId",
    });
});
export function exchangeAuthCodeForAPIKeyResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ExchangeAuthCodeForAPIKeyResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ExchangeAuthCodeForAPIKeyResponse' from JSON`);
}
