// @openrouter/sdk/esm/models/operations/getcredits.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 0e5a438307ae
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const GetCreditsRequest$outboundSchema = z.object({
    httpReferer: z.string().optional(),
    appTitle: z.string().optional(),
    appCategories: z.string().optional(),
}).transform((v) => {
    return remap$(v, {
        httpReferer: "HTTP-Referer",
    });
});
export function getCreditsRequestToJSON(getCreditsRequest) {
    return JSON.stringify(GetCreditsRequest$outboundSchema.parse(getCreditsRequest));
}
/** @internal */
export const GetCreditsData$inboundSchema = z.object({
    total_credits: z.number(),
    total_usage: z.number(),
}).transform((v) => {
    return remap$(v, {
        "total_credits": "totalCredits",
        "total_usage": "totalUsage",
    });
});
export function getCreditsDataFromJSON(jsonString) {
    return safeParse(jsonString, (x) => GetCreditsData$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'GetCreditsData' from JSON`);
}
/** @internal */
export const GetCreditsResponse$inboundSchema = z.object({
    data: z.lazy(() => GetCreditsData$inboundSchema),
});
export function getCreditsResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => GetCreditsResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'GetCreditsResponse' from JSON`);
}
