// @openrouter/sdk/esm/models/operations/getgeneration.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 5cdb2959d2a5
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as openEnums from "../../types/enums.js";
import * as models from "../index.js";
/**
 * Type of API used for the generation
 */
export const ApiType = {
    Completions: "completions",
    Embeddings: "embeddings",
    Rerank: "rerank",
    Tts: "tts",
    Video: "video",
};
/** @internal */
export const GetGenerationRequest$outboundSchema = z.object({
    httpReferer: z.string().optional(),
    appTitle: z.string().optional(),
    appCategories: z.string().optional(),
    id: z.string(),
}).transform((v) => {
    return remap$(v, {
        httpReferer: "HTTP-Referer",
    });
});
export function getGenerationRequestToJSON(getGenerationRequest) {
    return JSON.stringify(GetGenerationRequest$outboundSchema.parse(getGenerationRequest));
}
/** @internal */
export const ApiType$inboundSchema = openEnums
    .inboundSchema(ApiType);
/** @internal */
export const GetGenerationData$inboundSchema = z.object({
    api_type: z.nullable(ApiType$inboundSchema),
    app_id: z.nullable(z.int()),
    cache_discount: z.nullable(z.number()),
    cancelled: z.nullable(z.boolean()),
    created_at: z.string(),
    external_user: z.nullable(z.string()),
    finish_reason: z.nullable(z.string()),
    generation_time: z.nullable(z.number()),
    http_referer: z.nullable(z.string()),
    id: z.string(),
    is_byok: z.boolean(),
    latency: z.nullable(z.number()),
    model: z.string(),
    moderation_latency: z.nullable(z.number()),
    native_finish_reason: z.nullable(z.string()),
    native_tokens_cached: z.nullable(z.int()),
    native_tokens_completion: z.nullable(z.int()),
    native_tokens_completion_images: z.nullable(z.int()),
    native_tokens_prompt: z.nullable(z.int()),
    native_tokens_reasoning: z.nullable(z.int()),
    num_input_audio_prompt: z.nullable(z.int()),
    num_media_completion: z.nullable(z.int()),
    num_media_prompt: z.nullable(z.int()),
    num_search_results: z.nullable(z.int()),
    origin: z.string(),
    provider_name: z.nullable(z.string()),
    provider_responses: z.nullable(z.array(models.ProviderResponse$inboundSchema)),
    request_id: z.nullable(z.string()).optional(),
    router: z.nullable(z.string()),
    session_id: z.nullable(z.string()).optional(),
    streamed: z.nullable(z.boolean()),
    tokens_completion: z.nullable(z.int()),
    tokens_prompt: z.nullable(z.int()),
    total_cost: z.number(),
    upstream_id: z.nullable(z.string()),
    upstream_inference_cost: z.nullable(z.number()),
    usage: z.number(),
    user_agent: z.nullable(z.string()),
    web_search_engine: z.nullable(z.string()),
}).transform((v) => {
    return remap$(v, {
        "api_type": "apiType",
        "app_id": "appId",
        "cache_discount": "cacheDiscount",
        "created_at": "createdAt",
        "external_user": "externalUser",
        "finish_reason": "finishReason",
        "generation_time": "generationTime",
        "http_referer": "httpReferer",
        "is_byok": "isByok",
        "moderation_latency": "moderationLatency",
        "native_finish_reason": "nativeFinishReason",
        "native_tokens_cached": "nativeTokensCached",
        "native_tokens_completion": "nativeTokensCompletion",
        "native_tokens_completion_images": "nativeTokensCompletionImages",
        "native_tokens_prompt": "nativeTokensPrompt",
        "native_tokens_reasoning": "nativeTokensReasoning",
        "num_input_audio_prompt": "numInputAudioPrompt",
        "num_media_completion": "numMediaCompletion",
        "num_media_prompt": "numMediaPrompt",
        "num_search_results": "numSearchResults",
        "provider_name": "providerName",
        "provider_responses": "providerResponses",
        "request_id": "requestId",
        "session_id": "sessionId",
        "tokens_completion": "tokensCompletion",
        "tokens_prompt": "tokensPrompt",
        "total_cost": "totalCost",
        "upstream_id": "upstreamId",
        "upstream_inference_cost": "upstreamInferenceCost",
        "user_agent": "userAgent",
        "web_search_engine": "webSearchEngine",
    });
});
export function getGenerationDataFromJSON(jsonString) {
    return safeParse(jsonString, (x) => GetGenerationData$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'GetGenerationData' from JSON`);
}
/** @internal */
export const GetGenerationResponse$inboundSchema = z.object({
    data: z.lazy(() => GetGenerationData$inboundSchema),
});
export function getGenerationResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => GetGenerationResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'GetGenerationResponse' from JSON`);
}
