// @openrouter/sdk/esm/models/operations/getmodels.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: f1343fbaaed5
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import * as openEnums from "../../types/enums.js";
/**
 * Filter models by use case category
 */
export const Category = {
    Programming: "programming",
    Roleplay: "roleplay",
    Marketing: "marketing",
    MarketingSeo: "marketing/seo",
    Technology: "technology",
    Science: "science",
    Translation: "translation",
    Legal: "legal",
    Finance: "finance",
    Health: "health",
    Trivia: "trivia",
    Academia: "academia",
};
/** @internal */
export const Category$outboundSchema = openEnums
    .outboundSchema(Category);
/** @internal */
export const GetModelsRequest$outboundSchema = z.object({
    httpReferer: z.string().optional(),
    appTitle: z.string().optional(),
    appCategories: z.string().optional(),
    category: Category$outboundSchema.optional(),
    supportedParameters: z.string().optional(),
    outputModalities: z.string().optional(),
}).transform((v) => {
    return remap$(v, {
        httpReferer: "HTTP-Referer",
        supportedParameters: "supported_parameters",
        outputModalities: "output_modalities",
    });
});
export function getModelsRequestToJSON(getModelsRequest) {
    return JSON.stringify(GetModelsRequest$outboundSchema.parse(getModelsRequest));
}
