// @openrouter/sdk/esm/models/operations/listorganizationmembers.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: cc21259f09b8
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as openEnums from "../../types/enums.js";
/**
 * Role of the member in the organization
 */
export const Role = {
    OrgAdmin: "org:admin",
    OrgMember: "org:member",
};
/** @internal */
export const ListOrganizationMembersRequest$outboundSchema = z.object({
    httpReferer: z.string().optional(),
    appTitle: z.string().optional(),
    appCategories: z.string().optional(),
    offset: z.nullable(z.int()).optional(),
    limit: z.int().optional(),
}).transform((v) => {
    return remap$(v, {
        httpReferer: "HTTP-Referer",
    });
});
export function listOrganizationMembersRequestToJSON(listOrganizationMembersRequest) {
    return JSON.stringify(ListOrganizationMembersRequest$outboundSchema.parse(listOrganizationMembersRequest));
}
/** @internal */
export const Role$inboundSchema = openEnums
    .inboundSchema(Role);
/** @internal */
export const ListOrganizationMembersData$inboundSchema = z.object({
    email: z.string(),
    first_name: z.nullable(z.string()),
    id: z.string(),
    last_name: z.nullable(z.string()),
    role: Role$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "first_name": "firstName",
        "last_name": "lastName",
    });
});
export function listOrganizationMembersDataFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ListOrganizationMembersData$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ListOrganizationMembersData' from JSON`);
}
/** @internal */
export const ListOrganizationMembersResponseBody$inboundSchema = z.object({
    data: z.array(z.lazy(() => ListOrganizationMembersData$inboundSchema)),
    total_count: z.int(),
}).transform((v) => {
    return remap$(v, {
        "total_count": "totalCount",
    });
});
export function listOrganizationMembersResponseBodyFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ListOrganizationMembersResponseBody$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ListOrganizationMembersResponseBody' from JSON`);
}
/** @internal */
export const ListOrganizationMembersResponse$inboundSchema = z.object({
    Result: z.lazy(() => ListOrganizationMembersResponseBody$inboundSchema),
}).transform((v) => {
    return remap$(v, {
        "Result": "result",
    });
});
export function listOrganizationMembersResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ListOrganizationMembersResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ListOrganizationMembersResponse' from JSON`);
}
