// @openrouter/sdk/esm/models/operations/listkeyassignments.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: c4b0c0303428
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as models from "../index.js";
/** @internal */
export const ListKeyAssignmentsRequest$outboundSchema = z.object({
    httpReferer: z.string().optional(),
    appTitle: z.string().optional(),
    appCategories: z.string().optional(),
    offset: z.nullable(z.int()).optional(),
    limit: z.int().optional(),
}).transform((v) => {
    return remap$(v, {
        httpReferer: "HTTP-Referer",
    });
});
export function listKeyAssignmentsRequestToJSON(listKeyAssignmentsRequest) {
    return JSON.stringify(ListKeyAssignmentsRequest$outboundSchema.parse(listKeyAssignmentsRequest));
}
/** @internal */
export const ListKeyAssignmentsResponse$inboundSchema = z.object({
    Result: models.ListKeyAssignmentsResponse$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "Result": "result",
    });
});
export function listKeyAssignmentsResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ListKeyAssignmentsResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ListKeyAssignmentsResponse' from JSON`);
}
