// @openrouter/sdk/esm/models/operations/listguardrailmemberassignments.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 37ee24c57c33
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as models from "../index.js";
/** @internal */
export const ListGuardrailMemberAssignmentsRequest$outboundSchema = z.object({
    httpReferer: z.string().optional(),
    appTitle: z.string().optional(),
    appCategories: z.string().optional(),
    id: z.string(),
    offset: z.nullable(z.int()).optional(),
    limit: z.int().optional(),
}).transform((v) => {
    return remap$(v, {
        httpReferer: "HTTP-Referer",
    });
});
export function listGuardrailMemberAssignmentsRequestToJSON(listGuardrailMemberAssignmentsRequest) {
    return JSON.stringify(ListGuardrailMemberAssignmentsRequest$outboundSchema.parse(listGuardrailMemberAssignmentsRequest));
}
/** @internal */
export const ListGuardrailMemberAssignmentsResponse$inboundSchema = z.object({
    Result: models.ListMemberAssignmentsResponse$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "Result": "result",
    });
});
export function listGuardrailMemberAssignmentsResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ListGuardrailMemberAssignmentsResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ListGuardrailMemberAssignmentsResponse' from JSON`);
}
