// @openrouter/sdk/esm/models/operations/listendpoints.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 253a81bc10af
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as models from "../index.js";
/** @internal */
export const ListEndpointsRequest$outboundSchema = z.object({
    httpReferer: z.string().optional(),
    appTitle: z.string().optional(),
    appCategories: z.string().optional(),
    author: z.string(),
    slug: z.string(),
}).transform((v) => {
    return remap$(v, {
        httpReferer: "HTTP-Referer",
    });
});
export function listEndpointsRequestToJSON(listEndpointsRequest) {
    return JSON.stringify(ListEndpointsRequest$outboundSchema.parse(listEndpointsRequest));
}
/** @internal */
export const ListEndpointsResponse$inboundSchema = z.object({
    data: models.ListEndpointsResponse$inboundSchema,
});
export function listEndpointsResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ListEndpointsResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ListEndpointsResponse' from JSON`);
}
