// @openrouter/sdk/esm/models/operations/listendpointszdr.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: d52e6aeefb5e
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as models from "../index.js";
/** @internal */
export const ListEndpointsZdrRequest$outboundSchema = z.object({
    httpReferer: z.string().optional(),
    appTitle: z.string().optional(),
    appCategories: z.string().optional(),
}).transform((v) => {
    return remap$(v, {
        httpReferer: "HTTP-Referer",
    });
});
export function listEndpointsZdrRequestToJSON(listEndpointsZdrRequest) {
    return JSON.stringify(ListEndpointsZdrRequest$outboundSchema.parse(listEndpointsZdrRequest));
}
/** @internal */
export const ListEndpointsZdrResponse$inboundSchema = z.object({
    data: z.array(models.PublicEndpoint$inboundSchema),
});
export function listEndpointsZdrResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ListEndpointsZdrResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ListEndpointsZdrResponse' from JSON`);
}
