// @openrouter/sdk/esm/models/operations/listguardrails.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 4c71cf4db173
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import * as models from "../index.js";
/** @internal */
export const ListGuardrailsRequest$outboundSchema = z.object({
    httpReferer: z.string().optional(),
    appTitle: z.string().optional(),
    appCategories: z.string().optional(),
    offset: z.nullable(z.int()).optional(),
    limit: z.int().optional(),
    workspaceId: z.string().optional(),
}).transform((v) => {
    return remap$(v, {
        httpReferer: "HTTP-Referer",
        workspaceId: "workspace_id",
    });
});
export function listGuardrailsRequestToJSON(listGuardrailsRequest) {
    return JSON.stringify(ListGuardrailsRequest$outboundSchema.parse(listGuardrailsRequest));
}
/** @internal */
export const ListGuardrailsResponse$inboundSchema = z.object({
    Result: models.ListGuardrailsResponse$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "Result": "result",
    });
});
export function listGuardrailsResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ListGuardrailsResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ListGuardrailsResponse' from JSON`);
}
