// @openrouter/sdk/esm/sdk/guardrails.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: c4e976dbf057
 */
import { guardrailsBulkAssignKeys } from "../funcs/guardrailsBulkAssignKeys.js";
import { guardrailsBulkAssignMembers } from "../funcs/guardrailsBulkAssignMembers.js";
import { guardrailsBulkUnassignKeys } from "../funcs/guardrailsBulkUnassignKeys.js";
import { guardrailsBulkUnassignMembers } from "../funcs/guardrailsBulkUnassignMembers.js";
import { guardrailsCreate } from "../funcs/guardrailsCreate.js";
import { guardrailsDelete } from "../funcs/guardrailsDelete.js";
import { guardrailsGet } from "../funcs/guardrailsGet.js";
import { guardrailsList } from "../funcs/guardrailsList.js";
import { guardrailsListGuardrailKeyAssignments } from "../funcs/guardrailsListGuardrailKeyAssignments.js";
import { guardrailsListGuardrailMemberAssignments } from "../funcs/guardrailsListGuardrailMemberAssignments.js";
import { guardrailsListKeyAssignments } from "../funcs/guardrailsListKeyAssignments.js";
import { guardrailsListMemberAssignments } from "../funcs/guardrailsListMemberAssignments.js";
import { guardrailsUpdate } from "../funcs/guardrailsUpdate.js";
import { ClientSDK } from "../lib/sdks.js";
import { unwrapAsync } from "../types/fp.js";
import { unwrapResultIterator } from "../types/operations.js";
export class Guardrails extends ClientSDK {
    /**
     * List guardrails
     *
     * @remarks
     * List all guardrails for the authenticated user. [Management key](/docs/guides/overview/auth/management-api-keys) required.
     */
    async list(request, options) {
        return unwrapResultIterator(guardrailsList(this, request, options));
    }
    /**
     * Create a guardrail
     *
     * @remarks
     * Create a new guardrail for the authenticated user. [Management key](/docs/guides/overview/auth/management-api-keys) required.
     */
    async create(request, options) {
        return unwrapAsync(guardrailsCreate(this, request, options));
    }
    /**
     * Delete a guardrail
     *
     * @remarks
     * Delete an existing guardrail. [Management key](/docs/guides/overview/auth/management-api-keys) required.
     */
    async delete(request, options) {
        return unwrapAsync(guardrailsDelete(this, request, options));
    }
    /**
     * Get a guardrail
     *
     * @remarks
     * Get a single guardrail by ID. [Management key](/docs/guides/overview/auth/management-api-keys) required.
     */
    async get(request, options) {
        return unwrapAsync(guardrailsGet(this, request, options));
    }
    /**
     * Update a guardrail
     *
     * @remarks
     * Update an existing guardrail. [Management key](/docs/guides/overview/auth/management-api-keys) required.
     */
    async update(request, options) {
        return unwrapAsync(guardrailsUpdate(this, request, options));
    }
    /**
     * List key assignments for a guardrail
     *
     * @remarks
     * List all API key assignments for a specific guardrail. [Management key](/docs/guides/overview/auth/management-api-keys) required.
     */
    async listGuardrailKeyAssignments(request, options) {
        return unwrapResultIterator(guardrailsListGuardrailKeyAssignments(this, request, options));
    }
    /**
     * Bulk assign keys to a guardrail
     *
     * @remarks
     * Assign multiple API keys to a specific guardrail. [Management key](/docs/guides/overview/auth/management-api-keys) required.
     */
    async bulkAssignKeys(request, options) {
        return unwrapAsync(guardrailsBulkAssignKeys(this, request, options));
    }
    /**
     * Bulk unassign keys from a guardrail
     *
     * @remarks
     * Unassign multiple API keys from a specific guardrail. [Management key](/docs/guides/overview/auth/management-api-keys) required.
     */
    async bulkUnassignKeys(request, options) {
        return unwrapAsync(guardrailsBulkUnassignKeys(this, request, options));
    }
    /**
     * List member assignments for a guardrail
     *
     * @remarks
     * List all organization member assignments for a specific guardrail. [Management key](/docs/guides/overview/auth/management-api-keys) required.
     */
    async listGuardrailMemberAssignments(request, options) {
        return unwrapResultIterator(guardrailsListGuardrailMemberAssignments(this, request, options));
    }
    /**
     * Bulk assign members to a guardrail
     *
     * @remarks
     * Assign multiple organization members to a specific guardrail. [Management key](/docs/guides/overview/auth/management-api-keys) required.
     */
    async bulkAssignMembers(request, options) {
        return unwrapAsync(guardrailsBulkAssignMembers(this, request, options));
    }
    /**
     * Bulk unassign members from a guardrail
     *
     * @remarks
     * Unassign multiple organization members from a specific guardrail. [Management key](/docs/guides/overview/auth/management-api-keys) required.
     */
    async bulkUnassignMembers(request, options) {
        return unwrapAsync(guardrailsBulkUnassignMembers(this, request, options));
    }
    /**
     * List all key assignments
     *
     * @remarks
     * List all API key guardrail assignments for the authenticated user. [Management key](/docs/guides/overview/auth/management-api-keys) required.
     */
    async listKeyAssignments(request, options) {
        return unwrapResultIterator(guardrailsListKeyAssignments(this, request, options));
    }
    /**
     * List all member assignments
     *
     * @remarks
     * List all organization member guardrail assignments for the authenticated user. [Management key](/docs/guides/overview/auth/management-api-keys) required.
     */
    async listMemberAssignments(request, options) {
        return unwrapResultIterator(guardrailsListMemberAssignments(this, request, options));
    }
}
