// @openrouter/sdk/esm/sdk/sdk.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 784571af2f69
 */
import { ClientSDK } from "../lib/sdks.js";
import { Analytics } from "./analytics.js";
import { APIKeys } from "./apikeys.js";
import { Beta } from "./beta.js";
import { Chat } from "./chat.js";
import { Credits } from "./credits.js";
import { Embeddings } from "./embeddings.js";
import { Endpoints } from "./endpoints.js";
import { Generations } from "./generations.js";
import { Guardrails } from "./guardrails.js";
import { Models } from "./models.js";
import { OAuth } from "./oauth.js";
import { Organization } from "./organization.js";
import { Providers } from "./providers.js";
import { Rerank } from "./rerank.js";
import { Tts } from "./tts.js";
import { VideoGeneration } from "./videogeneration.js";
import { callModel as callModelFunc, } from "../funcs/call-model.js";
import { ToolType } from "../lib/tool-types.js";
export { ToolType };
// #endregion imports
export class OpenRouter extends ClientSDK {
    get analytics() {
        return (this._analytics ?? (this._analytics = new Analytics(this._options)));
    }
    get oAuth() {
        return (this._oAuth ?? (this._oAuth = new OAuth(this._options)));
    }
    get chat() {
        return (this._chat ?? (this._chat = new Chat(this._options)));
    }
    get credits() {
        return (this._credits ?? (this._credits = new Credits(this._options)));
    }
    get embeddings() {
        return (this._embeddings ?? (this._embeddings = new Embeddings(this._options)));
    }
    get endpoints() {
        return (this._endpoints ?? (this._endpoints = new Endpoints(this._options)));
    }
    get generations() {
        return (this._generations ?? (this._generations = new Generations(this._options)));
    }
    get guardrails() {
        return (this._guardrails ?? (this._guardrails = new Guardrails(this._options)));
    }
    get apiKeys() {
        return (this._apiKeys ?? (this._apiKeys = new APIKeys(this._options)));
    }
    get models() {
        return (this._models ?? (this._models = new Models(this._options)));
    }
    get organization() {
        return (this._organization ?? (this._organization = new Organization(this._options)));
    }
    get providers() {
        return (this._providers ?? (this._providers = new Providers(this._options)));
    }
    get rerank() {
        return (this._rerank ?? (this._rerank = new Rerank(this._options)));
    }
    get beta() {
        return (this._beta ?? (this._beta = new Beta(this._options)));
    }
    get tts() {
        return (this._tts ?? (this._tts = new Tts(this._options)));
    }
    get videoGeneration() {
        return (this._videoGeneration ?? (this._videoGeneration = new VideoGeneration(this._options)));
    }
    // #region sdk-class-body
    callModel(request, options) {
        return callModelFunc(this, request, options);
    }
}
