// @openrouter/sdk/esm/lib/stream-type-guards.js
/**
 * Type guards for OpenResponses stream events
 * These enable proper TypeScript narrowing without type casts
 */
// Stream event type guards
export function isOutputTextDeltaEvent(event) {
    return 'type' in event && event.type === 'response.output_text.delta';
}
export function isReasoningDeltaEvent(event) {
    return 'type' in event && event.type === 'response.reasoning_text.delta';
}
export function isFunctionCallArgumentsDeltaEvent(event) {
    return 'type' in event && event.type === 'response.function_call_arguments.delta';
}
export function isOutputItemAddedEvent(event) {
    return 'type' in event && event.type === 'response.output_item.added';
}
export function isOutputItemDoneEvent(event) {
    return 'type' in event && event.type === 'response.output_item.done';
}
export function isResponseCompletedEvent(event) {
    return 'type' in event && event.type === 'response.completed';
}
export function isResponseFailedEvent(event) {
    return 'type' in event && event.type === 'response.failed';
}
export function isResponseIncompleteEvent(event) {
    return 'type' in event && event.type === 'response.incomplete';
}
export function isFunctionCallArgumentsDoneEvent(event) {
    return 'type' in event && event.type === 'response.function_call_arguments.done';
}
// Output item type guards
export function isOutputMessage(item) {
    return typeof item === 'object' && item !== null && 'type' in item && item.type === 'message';
}
export function isFunctionCallItem(item) {
    return (typeof item === 'object' && item !== null && 'type' in item && item.type === 'function_call');
}
export function isReasoningOutputItem(item) {
    return typeof item === 'object' && item !== null && 'type' in item && item.type === 'reasoning';
}
export function isWebSearchCallOutputItem(item) {
    return (typeof item === 'object' && item !== null && 'type' in item && item.type === 'web_search_call');
}
export function isFileSearchCallOutputItem(item) {
    return (typeof item === 'object' && item !== null && 'type' in item && item.type === 'file_search_call');
}
export function isImageGenerationCallOutputItem(item) {
    return (typeof item === 'object' &&
        item !== null &&
        'type' in item &&
        item.type === 'image_generation_call');
}
// Content part type guards
export function isOutputTextPart(part) {
    return typeof part === 'object' && part !== null && 'type' in part && part.type === 'output_text';
}
export function isRefusalPart(part) {
    return typeof part === 'object' && part !== null && 'type' in part && part.type === 'refusal';
}
// Annotation type guards for Claude conversion
export function isFileCitationAnnotation(annotation) {
    return (typeof annotation === 'object' &&
        annotation !== null &&
        'type' in annotation &&
        annotation.type === 'file_citation');
}
export function isURLCitationAnnotation(annotation) {
    return (typeof annotation === 'object' &&
        annotation !== null &&
        'type' in annotation &&
        annotation.type === 'url_citation');
}
export function isFilePathAnnotation(annotation) {
    return (typeof annotation === 'object' &&
        annotation !== null &&
        'type' in annotation &&
        annotation.type === 'file_path');
}
// Helper to check if output has a type property
export function hasTypeProperty(item) {
    return (typeof item === 'object' && item !== null && 'type' in item && typeof item.type === 'string');
}
