// @openrouter/sdk/esm/lib/tool-executor.js
import * as z4 from 'zod/v4';
import { hasExecuteFunction, isGeneratorTool, isRegularExecuteTool } from './tool-types.js';
import { buildToolExecuteContext } from './tool-context.js';
// Re-export ZodError for convenience
export const ZodError = z4.ZodError;
/**
 * Typeguard to check if a value is a non-null object (not an array).
 */
function isNonNullObject(value) {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
}
export function sanitizeJsonSchema(obj) {
    if (obj === null || typeof obj !== 'object') {
        return obj;
    }
    if (Array.isArray(obj)) {
        return obj.map(sanitizeJsonSchema);
    }
    // At this point, obj is a non-null, non-array object
    // Use typeguard to narrow the type for type-safe property access
    if (!isNonNullObject(obj)) {
        return obj;
    }
    const result = {};
    for (const key of Object.keys(obj)) {
        if (!key.startsWith('~')) {
            result[key] = sanitizeJsonSchema(obj[key]);
        }
    }
    return result;
}
/**
 * Typeguard to check if a value is a valid Zod schema compatible with zod/v4.
 * Zod schemas have a _zod property that contains schema metadata.
 */
function isZodSchema(value) {
    if (typeof value !== 'object' || value === null) {
        return false;
    }
    if (!('_zod' in value)) {
        return false;
    }
    // After the 'in' check, TypeScript knows value has _zod property
    return typeof value._zod === 'object';
}
/**
 * Convert a Zod schema to JSON Schema using Zod v4's toJSONSchema function.
 * Accepts ZodType from the main zod package for user compatibility.
 * The resulting schema is sanitized to remove metadata properties (like ~standard)
 * that would cause 400 errors with downstream providers.
 */
export function convertZodToJsonSchema(zodSchema) {
    if (!isZodSchema(zodSchema)) {
        throw new Error('Invalid Zod schema provided');
    }
    // Use draft-7 as it's closest to OpenAPI 3.0's JSON Schema variant
    const jsonSchema = z4.toJSONSchema(zodSchema, {
        target: 'draft-7',
    });
    // jsonSchema is always a Record<string, unknown> from toJSONSchema
    // The overloaded sanitizeJsonSchema preserves this type
    return sanitizeJsonSchema(jsonSchema);
}
/**
 * Convert tools to OpenRouter API format
 * Accepts readonly arrays for better type compatibility
 */
export function convertToolsToAPIFormat(tools) {
    return tools.map((tool) => ({
        type: 'function',
        name: tool.function.name,
        description: tool.function.description || null,
        strict: null,
        parameters: convertZodToJsonSchema(tool.function.inputSchema),
    }));
}
/**
 * Validate tool input against Zod schema
 * @throws ZodError if validation fails
 */
export function validateToolInput(schema, args) {
    return z4.parse(schema, args);
}
/**
 * Validate tool output against Zod schema
 * @throws ZodError if validation fails
 */
export function validateToolOutput(schema, result) {
    return z4.parse(schema, result);
}
/**
 * Try to validate a value against a Zod schema without throwing
 * @returns true if validation succeeds, false otherwise
 */
function tryValidate(schema, value) {
    const result = z4.safeParse(schema, value);
    return result.success;
}
/**
 * Parse tool call arguments from JSON string.
 * Treats empty/whitespace-only strings as an empty object — some providers
 * return `arguments: ""` for tools that take no parameters.
 */
export function parseToolCallArguments(argumentsString) {
    const trimmed = argumentsString.trim();
    if (!trimmed) {
        return {};
    }
    try {
        return JSON.parse(trimmed);
    }
    catch (error) {
        throw new Error(`Failed to parse tool call arguments: ${error instanceof Error ? error.message : String(error)}`);
    }
}
/**
 * Build a ToolExecuteContext for a tool from a TurnContext and optional context store
 */
function buildExecuteCtx(tool, turnContext, contextStore, sharedSchema) {
    return buildToolExecuteContext(turnContext, contextStore, tool.function.name, tool.function.contextSchema, sharedSchema);
}
/**
 * Execute a regular (non-generator) tool
 */
export async function executeRegularTool(tool, toolCall, context, contextStore, sharedSchema) {
    if (!isRegularExecuteTool(tool)) {
        throw new Error(`Tool "${toolCall.name}" is not a regular execute tool or has no execute function`);
    }
    try {
        const validatedInput = validateToolInput(tool.function.inputSchema, toolCall.arguments);
        const executeContext = buildExecuteCtx(tool, context, contextStore, sharedSchema);
        // Execute tool with context
        const result = await Promise.resolve(tool.function.execute(validatedInput, executeContext));
        // Validate output if schema is provided
        if (tool.function.outputSchema) {
            const validatedOutput = validateToolOutput(tool.function.outputSchema, result);
            return {
                toolCallId: toolCall.id,
                toolName: toolCall.name,
                result: validatedOutput,
            };
        }
        return {
            toolCallId: toolCall.id,
            toolName: toolCall.name,
            result,
        };
    }
    catch (error) {
        return {
            toolCallId: toolCall.id,
            toolName: toolCall.name,
            result: null,
            error: error instanceof Error ? error : new Error(String(error)),
        };
    }
}
/**
 * Execute a generator tool and collect preliminary and final results
 * - Intermediate yields are validated against eventSchema (preliminary events)
 * - Last yield is validated against outputSchema (final result sent to model)
 * - Generator must emit at least one value
 */
export async function executeGeneratorTool(tool, toolCall, context, onPreliminaryResult, contextStore, sharedSchema) {
    if (!isGeneratorTool(tool)) {
        throw new Error(`Tool "${toolCall.name}" is not a generator tool`);
    }
    try {
        const validatedInput = validateToolInput(tool.function.inputSchema, toolCall.arguments);
        const executeContext = buildExecuteCtx(tool, context, contextStore, sharedSchema);
        const preliminaryResults = [];
        let finalResult;
        let hasFinalResult = false;
        let lastEmittedValue;
        let hasEmittedValue = false;
        const iterator = tool.function.execute(validatedInput, executeContext);
        let iterResult = await iterator.next();
        while (!iterResult.done) {
            const event = iterResult.value;
            lastEmittedValue = event;
            hasEmittedValue = true;
            const matchesOutputSchema = tryValidate(tool.function.outputSchema, event);
            const matchesEventSchema = tryValidate(tool.function.eventSchema, event);
            if (matchesOutputSchema && !matchesEventSchema && !hasFinalResult) {
                finalResult = validateToolOutput(tool.function.outputSchema, event);
                hasFinalResult = true;
            }
            else {
                const validatedPreliminary = validateToolOutput(tool.function.eventSchema, event);
                preliminaryResults.push(validatedPreliminary);
                if (onPreliminaryResult) {
                    onPreliminaryResult(toolCall.id, validatedPreliminary);
                }
            }
            iterResult = await iterator.next();
        }
        if (iterResult.value !== undefined) {
            finalResult = validateToolOutput(tool.function.outputSchema, iterResult.value);
            hasFinalResult = true;
        }
        if (!hasFinalResult) {
            if (!hasEmittedValue) {
                throw new Error(`Generator tool "${toolCall.name}" completed without emitting any values or returning a result`);
            }
            finalResult = validateToolOutput(tool.function.outputSchema, lastEmittedValue);
        }
        return {
            toolCallId: toolCall.id,
            toolName: toolCall.name,
            result: finalResult,
            preliminaryResults,
        };
    }
    catch (error) {
        return {
            toolCallId: toolCall.id,
            toolName: toolCall.name,
            result: null,
            error: error instanceof Error ? error : new Error(String(error)),
        };
    }
}
/**
 * Execute a tool call
 * Automatically detects if it's a regular or generator tool
 */
export async function executeTool(tool, toolCall, context, onPreliminaryResult, contextStore, sharedSchema) {
    if (!hasExecuteFunction(tool)) {
        throw new Error(`Tool "${toolCall.name}" has no execute function. Use manual tool execution.`);
    }
    if (isGeneratorTool(tool)) {
        return executeGeneratorTool(tool, toolCall, context, onPreliminaryResult, contextStore, sharedSchema);
    }
    return executeRegularTool(tool, toolCall, context, contextStore, sharedSchema);
}
/**
 * Find a tool by name in the tools array
 */
export function findToolByName(tools, name) {
    return tools.find((tool) => tool.function.name === name);
}
/**
 * Format tool execution result as a string for sending to the model
 */
export function formatToolResultForModel(result) {
    if (result.error) {
        return JSON.stringify({
            error: result.error.message,
            toolName: result.toolName,
        });
    }
    return JSON.stringify(result.result);
}
/**
 * Create a user-friendly error message for tool execution errors
 */
export function formatToolExecutionError(error, toolCall) {
    if (error instanceof ZodError) {
        const issues = error.issues.map((issue) => ({
            path: issue.path.join('.'),
            message: issue.message,
        }));
        return `Tool "${toolCall.name}" validation error:\n${JSON.stringify(issues, null, 2)}`;
    }
    return `Tool "${toolCall.name}" execution error: ${error.message}`;
}
