// @openrouter/sdk/esm/models/activityitem.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: ee2d86453873
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
/** @internal */
export const ActivityItem$inboundSchema = z
    .object({
    byok_usage_inference: z.number(),
    completion_tokens: z.int(),
    date: z.string(),
    endpoint_id: z.string(),
    model: z.string(),
    model_permaslug: z.string(),
    prompt_tokens: z.int(),
    provider_name: z.string(),
    reasoning_tokens: z.int(),
    requests: z.int(),
    usage: z.number(),
}).transform((v) => {
    return remap$(v, {
        "byok_usage_inference": "byokUsageInference",
        "completion_tokens": "completionTokens",
        "endpoint_id": "endpointId",
        "model_permaslug": "modelPermaslug",
        "prompt_tokens": "promptTokens",
        "provider_name": "providerName",
        "reasoning_tokens": "reasoningTokens",
    });
});
export function activityItemFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ActivityItem$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ActivityItem' from JSON`);
}
