// @openrouter/sdk/esm/models/annotationaddedevent.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: e5c5f2adf5f6
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { OpenAIResponsesAnnotation$inboundSchema, } from "./openairesponsesannotation.js";
/** @internal */
export const AnnotationAddedEvent$inboundSchema = z.object({
    annotation: OpenAIResponsesAnnotation$inboundSchema,
    annotation_index: z.int(),
    content_index: z.int(),
    item_id: z.string(),
    output_index: z.int(),
    sequence_number: z.int(),
    type: z.literal("response.output_text.annotation.added"),
}).transform((v) => {
    return remap$(v, {
        "annotation_index": "annotationIndex",
        "content_index": "contentIndex",
        "item_id": "itemId",
        "output_index": "outputIndex",
        "sequence_number": "sequenceNumber",
    });
});
export function annotationAddedEventFromJSON(jsonString) {
    return safeParse(jsonString, (x) => AnnotationAddedEvent$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'AnnotationAddedEvent' from JSON`);
}
