// @openrouter/sdk/esm/models/baseinputsunion.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 7237afe9aafd
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { discriminatedUnion } from "../types/discriminatedUnion.js";
import { InputAudio$inboundSchema } from "./inputaudio.js";
import { InputFile$inboundSchema } from "./inputfile.js";
import { InputImage$inboundSchema } from "./inputimage.js";
import { InputText$inboundSchema } from "./inputtext.js";
import { OpenAIResponseFunctionToolCall$inboundSchema, } from "./openairesponsefunctiontoolcall.js";
import { OpenAIResponseFunctionToolCallOutput$inboundSchema, } from "./openairesponsefunctiontoolcalloutput.js";
import { OpenAIResponseInputMessageItem$inboundSchema, } from "./openairesponseinputmessageitem.js";
import { OutputItemImageGenerationCall$inboundSchema, } from "./outputitemimagegenerationcall.js";
import { OutputMessage$inboundSchema } from "./outputmessage.js";
export const BaseInputsPhaseFinalAnswer = {
    FinalAnswer: "final_answer",
};
export const BaseInputsPhaseCommentary = {
    Commentary: "commentary",
};
export const BaseInputsRoleDeveloper = {
    Developer: "developer",
};
export const BaseInputsRoleAssistant = {
    Assistant: "assistant",
};
export const BaseInputsRoleSystem = {
    System: "system",
};
export const BaseInputsRoleUser = {
    User: "user",
};
export const BaseInputsType = {
    Message: "message",
};
/** @internal */
export const BaseInputsContent1$inboundSchema = discriminatedUnion("type", {
    input_audio: InputAudio$inboundSchema,
    input_file: InputFile$inboundSchema,
    input_image: InputImage$inboundSchema.and(z.object({ type: z.literal("input_image") })),
    input_text: InputText$inboundSchema,
});
export function baseInputsContent1FromJSON(jsonString) {
    return safeParse(jsonString, (x) => BaseInputsContent1$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'BaseInputsContent1' from JSON`);
}
/** @internal */
export const BaseInputsContent2$inboundSchema = z.union([
    z.array(discriminatedUnion("type", {
        input_audio: InputAudio$inboundSchema,
        input_file: InputFile$inboundSchema,
        input_image: InputImage$inboundSchema.and(z.object({ type: z.literal("input_image") })),
        input_text: InputText$inboundSchema,
    })),
    z.string(),
]);
export function baseInputsContent2FromJSON(jsonString) {
    return safeParse(jsonString, (x) => BaseInputsContent2$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'BaseInputsContent2' from JSON`);
}
/** @internal */
export const BaseInputsPhaseFinalAnswer$inboundSchema = z.enum(BaseInputsPhaseFinalAnswer);
/** @internal */
export const BaseInputsPhaseCommentary$inboundSchema = z.enum(BaseInputsPhaseCommentary);
/** @internal */
export const BaseInputsPhaseUnion$inboundSchema = z.union([
    BaseInputsPhaseCommentary$inboundSchema,
    BaseInputsPhaseFinalAnswer$inboundSchema,
    z.any(),
]);
export function baseInputsPhaseUnionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => BaseInputsPhaseUnion$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'BaseInputsPhaseUnion' from JSON`);
}
/** @internal */
export const BaseInputsRoleDeveloper$inboundSchema = z.enum(BaseInputsRoleDeveloper);
/** @internal */
export const BaseInputsRoleAssistant$inboundSchema = z.enum(BaseInputsRoleAssistant);
/** @internal */
export const BaseInputsRoleSystem$inboundSchema = z.enum(BaseInputsRoleSystem);
/** @internal */
export const BaseInputsRoleUser$inboundSchema = z.enum(BaseInputsRoleUser);
/** @internal */
export const BaseInputsRoleUnion$inboundSchema = z.union([
    BaseInputsRoleUser$inboundSchema,
    BaseInputsRoleSystem$inboundSchema,
    BaseInputsRoleAssistant$inboundSchema,
    BaseInputsRoleDeveloper$inboundSchema,
]);
export function baseInputsRoleUnionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => BaseInputsRoleUnion$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'BaseInputsRoleUnion' from JSON`);
}
/** @internal */
export const BaseInputsType$inboundSchema = z
    .enum(BaseInputsType);
/** @internal */
export const BaseInputsMessage$inboundSchema = z.object({
    content: z.union([
        z.array(discriminatedUnion("type", {
            input_audio: InputAudio$inboundSchema,
            input_file: InputFile$inboundSchema,
            input_image: InputImage$inboundSchema.and(z.object({ type: z.literal("input_image") })),
            input_text: InputText$inboundSchema,
        })),
        z.string(),
    ]),
    phase: z.nullable(z.union([
        BaseInputsPhaseCommentary$inboundSchema,
        BaseInputsPhaseFinalAnswer$inboundSchema,
        z.any(),
    ])).optional(),
    role: z.union([
        BaseInputsRoleUser$inboundSchema,
        BaseInputsRoleSystem$inboundSchema,
        BaseInputsRoleAssistant$inboundSchema,
        BaseInputsRoleDeveloper$inboundSchema,
    ]),
    type: BaseInputsType$inboundSchema.optional(),
});
export function baseInputsMessageFromJSON(jsonString) {
    return safeParse(jsonString, (x) => BaseInputsMessage$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'BaseInputsMessage' from JSON`);
}
/** @internal */
export const BaseInputsUnion1$inboundSchema = z.union([
    OpenAIResponseFunctionToolCall$inboundSchema,
    OutputMessage$inboundSchema,
    OpenAIResponseInputMessageItem$inboundSchema,
    OpenAIResponseFunctionToolCallOutput$inboundSchema,
    OutputItemImageGenerationCall$inboundSchema,
    z.lazy(() => BaseInputsMessage$inboundSchema),
]);
export function baseInputsUnion1FromJSON(jsonString) {
    return safeParse(jsonString, (x) => BaseInputsUnion1$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'BaseInputsUnion1' from JSON`);
}
/** @internal */
export const BaseInputsUnion$inboundSchema = z.union([
    z.string(),
    z.array(z.union([
        OpenAIResponseFunctionToolCall$inboundSchema,
        OutputMessage$inboundSchema,
        OpenAIResponseInputMessageItem$inboundSchema,
        OpenAIResponseFunctionToolCallOutput$inboundSchema,
        OutputItemImageGenerationCall$inboundSchema,
        z.lazy(() => BaseInputsMessage$inboundSchema),
    ])),
    z.any(),
]);
export function baseInputsUnionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => BaseInputsUnion$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'BaseInputsUnion' from JSON`);
}
