// @openrouter/sdk/esm/models/chatstreamtoolcall.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: bad96c4eeaf7
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
/**
 * Tool call type
 */
export const ChatStreamToolCallType = {
    Function: "function",
};
/** @internal */
export const ChatStreamToolCallFunction$inboundSchema = z.object({
    arguments: z.string().optional(),
    name: z.string().optional(),
});
export function chatStreamToolCallFunctionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ChatStreamToolCallFunction$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatStreamToolCallFunction' from JSON`);
}
/** @internal */
export const ChatStreamToolCallType$inboundSchema = z.enum(ChatStreamToolCallType);
/** @internal */
export const ChatStreamToolCall$inboundSchema = z.object({
    function: z.lazy(() => ChatStreamToolCallFunction$inboundSchema).optional(),
    id: z.string().optional(),
    index: z.int(),
    type: ChatStreamToolCallType$inboundSchema.optional(),
});
export function chatStreamToolCallFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ChatStreamToolCall$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatStreamToolCall' from JSON`);
}
