// @openrouter/sdk/esm/models/chatfunctiontool.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 400104319b29
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { ChatContentCacheControl$outboundSchema, } from "./chatcontentcachecontrol.js";
import { ChatSearchModelsServerTool$outboundSchema, } from "./chatsearchmodelsservertool.js";
import { ChatWebSearchShorthand$outboundSchema, } from "./chatwebsearchshorthand.js";
import { DatetimeServerTool$outboundSchema, } from "./datetimeservertool.js";
import { ImageGenerationServerToolOpenRouter$outboundSchema, } from "./imagegenerationservertoolopenrouter.js";
import { OpenRouterWebSearchServerTool$outboundSchema, } from "./openrouterwebsearchservertool.js";
export const ChatFunctionToolType = {
    Function: "function",
};
/** @internal */
export const ChatFunctionToolFunctionFunction$outboundSchema = z.object({
    description: z.string().optional(),
    name: z.string(),
    parameters: z.record(z.string(), z.nullable(z.any())).optional(),
    strict: z.nullable(z.boolean()).optional(),
});
export function chatFunctionToolFunctionFunctionToJSON(chatFunctionToolFunctionFunction) {
    return JSON.stringify(ChatFunctionToolFunctionFunction$outboundSchema.parse(chatFunctionToolFunctionFunction));
}
/** @internal */
export const ChatFunctionToolType$outboundSchema = z.enum(ChatFunctionToolType);
/** @internal */
export const ChatFunctionToolFunction$outboundSchema = z.object({
    cacheControl: ChatContentCacheControl$outboundSchema.optional(),
    function: z.lazy(() => ChatFunctionToolFunctionFunction$outboundSchema),
    type: ChatFunctionToolType$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        cacheControl: "cache_control",
    });
});
export function chatFunctionToolFunctionToJSON(chatFunctionToolFunction) {
    return JSON.stringify(ChatFunctionToolFunction$outboundSchema.parse(chatFunctionToolFunction));
}
/** @internal */
export const ChatFunctionTool$outboundSchema = z.union([
    z.lazy(() => ChatFunctionToolFunction$outboundSchema),
    DatetimeServerTool$outboundSchema,
    ImageGenerationServerToolOpenRouter$outboundSchema,
    ChatSearchModelsServerTool$outboundSchema,
    OpenRouterWebSearchServerTool$outboundSchema,
    ChatWebSearchShorthand$outboundSchema,
]);
export function chatFunctionToolToJSON(chatFunctionTool) {
    return JSON.stringify(ChatFunctionTool$outboundSchema.parse(chatFunctionTool));
}
