// @openrouter/sdk/esm/models/chatresult.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: afefd8aadd09
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { ChatChoice$inboundSchema } from "./chatchoice.js";
import { ChatUsage$inboundSchema } from "./chatusage.js";
export const ChatResultObject = {
    ChatCompletion: "chat.completion",
};
/** @internal */
export const ChatResultObject$inboundSchema = z.enum(ChatResultObject);
/** @internal */
export const ChatResult$inboundSchema = z
    .object({
    choices: z.array(ChatChoice$inboundSchema),
    created: z.int(),
    id: z.string(),
    model: z.string(),
    object: ChatResultObject$inboundSchema,
    service_tier: z.nullable(z.string()).optional(),
    system_fingerprint: z.nullable(z.string()),
    usage: ChatUsage$inboundSchema.optional(),
}).transform((v) => {
    return remap$(v, {
        "service_tier": "serviceTier",
        "system_fingerprint": "systemFingerprint",
    });
});
export function chatResultFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ChatResult$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatResult' from JSON`);
}
