// @openrouter/sdk/esm/models/chatwebsearchshorthand.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: a67d6bbce478
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import * as openEnums from "../types/enums.js";
import { SearchQualityLevel$outboundSchema, } from "./searchqualitylevel.js";
import { WebSearchConfig$outboundSchema, } from "./websearchconfig.js";
import { WebSearchEngineEnum$outboundSchema, } from "./websearchengineenum.js";
import { WebSearchUserLocationServerTool$outboundSchema, } from "./websearchuserlocationservertool.js";
export const ChatWebSearchShorthandType = {
    WebSearch: "web_search",
    WebSearchPreview: "web_search_preview",
    WebSearchPreview20250311: "web_search_preview_2025_03_11",
    WebSearch20250826: "web_search_2025_08_26",
};
/** @internal */
export const ChatWebSearchShorthandType$outboundSchema = openEnums.outboundSchema(ChatWebSearchShorthandType);
/** @internal */
export const ChatWebSearchShorthand$outboundSchema = z.object({
    allowedDomains: z.array(z.string()).optional(),
    engine: WebSearchEngineEnum$outboundSchema.optional(),
    excludedDomains: z.array(z.string()).optional(),
    maxResults: z.int().optional(),
    maxTotalResults: z.int().optional(),
    parameters: WebSearchConfig$outboundSchema.optional(),
    searchContextSize: SearchQualityLevel$outboundSchema.optional(),
    type: ChatWebSearchShorthandType$outboundSchema,
    userLocation: WebSearchUserLocationServerTool$outboundSchema.optional(),
}).transform((v) => {
    return remap$(v, {
        allowedDomains: "allowed_domains",
        excludedDomains: "excluded_domains",
        maxResults: "max_results",
        maxTotalResults: "max_total_results",
        searchContextSize: "search_context_size",
        userLocation: "user_location",
    });
});
export function chatWebSearchShorthandToJSON(chatWebSearchShorthand) {
    return JSON.stringify(ChatWebSearchShorthand$outboundSchema.parse(chatWebSearchShorthand));
}
