// @openrouter/sdk/esm/models/chatstreamdelta.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 74a4be948d5f
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { ChatAudioOutput$inboundSchema, } from "./chataudiooutput.js";
import { ChatStreamToolCall$inboundSchema, } from "./chatstreamtoolcall.js";
import { ReasoningDetailUnion$inboundSchema, } from "./reasoningdetailunion.js";
/**
 * The role of the message author
 */
export const ChatStreamDeltaRole = {
    Assistant: "assistant",
};
/** @internal */
export const ChatStreamDeltaRole$inboundSchema = z.enum(ChatStreamDeltaRole);
/** @internal */
export const ChatStreamDelta$inboundSchema = z.object({
    audio: ChatAudioOutput$inboundSchema.optional(),
    content: z.nullable(z.string()).optional(),
    reasoning: z.nullable(z.string()).optional(),
    reasoning_details: z.array(ReasoningDetailUnion$inboundSchema).optional(),
    refusal: z.nullable(z.string()).optional(),
    role: ChatStreamDeltaRole$inboundSchema.optional(),
    tool_calls: z.array(ChatStreamToolCall$inboundSchema).optional(),
}).transform((v) => {
    return remap$(v, {
        "reasoning_details": "reasoningDetails",
        "tool_calls": "toolCalls",
    });
});
export function chatStreamDeltaFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ChatStreamDelta$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatStreamDelta' from JSON`);
}
