// @openrouter/sdk/esm/models/chatchoice.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: cb1eedd5b62e
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { ChatAssistantMessage$inboundSchema, } from "./chatassistantmessage.js";
import { ChatFinishReasonEnum$inboundSchema, } from "./chatfinishreasonenum.js";
import { ChatTokenLogprobs$inboundSchema, } from "./chattokenlogprobs.js";
/** @internal */
export const ChatChoice$inboundSchema = z
    .object({
    finish_reason: z.nullable(ChatFinishReasonEnum$inboundSchema),
    index: z.int(),
    logprobs: z.nullable(ChatTokenLogprobs$inboundSchema).optional(),
    message: ChatAssistantMessage$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "finish_reason": "finishReason",
    });
});
export function chatChoiceFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ChatChoice$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatChoice' from JSON`);
}
