// @openrouter/sdk/esm/models/chatstreamchunk.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 361f940e91e4
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { ChatStreamChoice$inboundSchema, } from "./chatstreamchoice.js";
import { ChatUsage$inboundSchema } from "./chatusage.js";
export const ChatStreamChunkObject = {
    ChatCompletionChunk: "chat.completion.chunk",
};
/** @internal */
export const ErrorT$inboundSchema = z.object({
    code: z.int(),
    message: z.string(),
});
export function errorFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ErrorT$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ErrorT' from JSON`);
}
/** @internal */
export const ChatStreamChunkObject$inboundSchema = z.enum(ChatStreamChunkObject);
/** @internal */
export const ChatStreamChunk$inboundSchema = z.object({
    choices: z.array(ChatStreamChoice$inboundSchema),
    created: z.int(),
    error: z.lazy(() => ErrorT$inboundSchema).optional(),
    id: z.string(),
    model: z.string(),
    object: ChatStreamChunkObject$inboundSchema,
    service_tier: z.nullable(z.string()).optional(),
    system_fingerprint: z.string().optional(),
    usage: ChatUsage$inboundSchema.optional(),
}).transform((v) => {
    return remap$(v, {
        "service_tier": "serviceTier",
        "system_fingerprint": "systemFingerprint",
    });
});
export function chatStreamChunkFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ChatStreamChunk$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatStreamChunk' from JSON`);
}
