// @openrouter/sdk/esm/models/chataudiooutput.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 692ef1eded11
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
/** @internal */
export const ChatAudioOutput$inboundSchema = z.object({
    data: z.string().optional(),
    expires_at: z.int().optional(),
    id: z.string().optional(),
    transcript: z.string().optional(),
}).transform((v) => {
    return remap$(v, {
        "expires_at": "expiresAt",
    });
});
/** @internal */
export const ChatAudioOutput$outboundSchema = z.object({
    data: z.string().optional(),
    expiresAt: z.int().optional(),
    id: z.string().optional(),
    transcript: z.string().optional(),
}).transform((v) => {
    return remap$(v, {
        expiresAt: "expires_at",
    });
});
export function chatAudioOutputToJSON(chatAudioOutput) {
    return JSON.stringify(ChatAudioOutput$outboundSchema.parse(chatAudioOutput));
}
export function chatAudioOutputFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ChatAudioOutput$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatAudioOutput' from JSON`);
}
