// @openrouter/sdk/esm/models/chatassistantimages.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 31fb8a4309f9
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
/** @internal */
export const ChatAssistantImagesImageUrl$inboundSchema = z.object({
    url: z.string(),
});
/** @internal */
export const ChatAssistantImagesImageUrl$outboundSchema = z.object({
    url: z.string(),
});
export function chatAssistantImagesImageUrlToJSON(chatAssistantImagesImageUrl) {
    return JSON.stringify(ChatAssistantImagesImageUrl$outboundSchema.parse(chatAssistantImagesImageUrl));
}
export function chatAssistantImagesImageUrlFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ChatAssistantImagesImageUrl$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatAssistantImagesImageUrl' from JSON`);
}
/** @internal */
export const ChatAssistantImages$inboundSchema = z.object({
    image_url: z.lazy(() => ChatAssistantImagesImageUrl$inboundSchema),
}).transform((v) => {
    return remap$(v, {
        "image_url": "imageUrl",
    });
});
/** @internal */
export const ChatAssistantImages$outboundSchema = z.object({
    imageUrl: z.lazy(() => ChatAssistantImagesImageUrl$outboundSchema),
}).transform((v) => {
    return remap$(v, {
        imageUrl: "image_url",
    });
});
export function chatAssistantImagesToJSON(chatAssistantImages) {
    return JSON.stringify(ChatAssistantImages$outboundSchema.parse(chatAssistantImages));
}
export function chatAssistantImagesFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ChatAssistantImages$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatAssistantImages' from JSON`);
}
