// @openrouter/sdk/esm/models/chatcontentvideo.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 353c68bbb3b2
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { ChatContentVideoInput$inboundSchema, ChatContentVideoInput$outboundSchema, } from "./chatcontentvideoinput.js";
/** @internal */
export const ChatContentVideo$inboundSchema = z.object({
    type: z.literal("video_url"),
    video_url: ChatContentVideoInput$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "video_url": "videoUrl",
    });
});
/** @internal */
export const ChatContentVideo$outboundSchema = z.object({
    type: z.literal("video_url"),
    videoUrl: ChatContentVideoInput$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        videoUrl: "video_url",
    });
});
export function chatContentVideoToJSON(chatContentVideo) {
    return JSON.stringify(ChatContentVideo$outboundSchema.parse(chatContentVideo));
}
export function chatContentVideoFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ChatContentVideo$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatContentVideo' from JSON`);
}
