// @openrouter/sdk/esm/models/chatusage.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 1e677d1ed4fc
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
/** @internal */
export const CompletionTokensDetails$inboundSchema = z.object({
    accepted_prediction_tokens: z.nullable(z.int()).optional(),
    audio_tokens: z.nullable(z.int()).optional(),
    reasoning_tokens: z.nullable(z.int()).optional(),
    rejected_prediction_tokens: z.nullable(z.int()).optional(),
}).transform((v) => {
    return remap$(v, {
        "accepted_prediction_tokens": "acceptedPredictionTokens",
        "audio_tokens": "audioTokens",
        "reasoning_tokens": "reasoningTokens",
        "rejected_prediction_tokens": "rejectedPredictionTokens",
    });
});
export function completionTokensDetailsFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CompletionTokensDetails$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CompletionTokensDetails' from JSON`);
}
/** @internal */
export const PromptTokensDetails$inboundSchema = z.object({
    audio_tokens: z.int().optional(),
    cache_write_tokens: z.int().optional(),
    cached_tokens: z.int().optional(),
    video_tokens: z.int().optional(),
}).transform((v) => {
    return remap$(v, {
        "audio_tokens": "audioTokens",
        "cache_write_tokens": "cacheWriteTokens",
        "cached_tokens": "cachedTokens",
        "video_tokens": "videoTokens",
    });
});
export function promptTokensDetailsFromJSON(jsonString) {
    return safeParse(jsonString, (x) => PromptTokensDetails$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'PromptTokensDetails' from JSON`);
}
/** @internal */
export const ChatUsage$inboundSchema = z.object({
    completion_tokens: z.int(),
    completion_tokens_details: z.nullable(z.lazy(() => CompletionTokensDetails$inboundSchema)).optional(),
    prompt_tokens: z.int(),
    prompt_tokens_details: z.nullable(z.lazy(() => PromptTokensDetails$inboundSchema)).optional(),
    total_tokens: z.int(),
}).transform((v) => {
    return remap$(v, {
        "completion_tokens": "completionTokens",
        "completion_tokens_details": "completionTokensDetails",
        "prompt_tokens": "promptTokens",
        "prompt_tokens_details": "promptTokensDetails",
        "total_tokens": "totalTokens",
    });
});
export function chatUsageFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ChatUsage$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatUsage' from JSON`);
}
