// @openrouter/sdk/esm/models/chatcontentitems.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 3d8c886d8084
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { discriminatedUnion } from "../types/discriminatedUnion.js";
import { ChatContentAudio$inboundSchema, ChatContentAudio$outboundSchema, } from "./chatcontentaudio.js";
import { ChatContentFile$inboundSchema, ChatContentFile$outboundSchema, } from "./chatcontentfile.js";
import { ChatContentImage$inboundSchema, ChatContentImage$outboundSchema, } from "./chatcontentimage.js";
import { ChatContentText$inboundSchema, ChatContentText$outboundSchema, } from "./chatcontenttext.js";
import { ChatContentVideo$inboundSchema, ChatContentVideo$outboundSchema, } from "./chatcontentvideo.js";
import { LegacyChatContentVideo$inboundSchema, LegacyChatContentVideo$outboundSchema, } from "./legacychatcontentvideo.js";
/** @internal */
export const ChatContentItems$inboundSchema = discriminatedUnion("type", {
    file: ChatContentFile$inboundSchema,
    image_url: ChatContentImage$inboundSchema,
    input_audio: ChatContentAudio$inboundSchema,
    input_video: LegacyChatContentVideo$inboundSchema,
    text: ChatContentText$inboundSchema.and(z.object({ type: z.literal("text") })),
    video_url: ChatContentVideo$inboundSchema,
});
/** @internal */
export const ChatContentItems$outboundSchema = z.union([
    ChatContentFile$outboundSchema,
    ChatContentImage$outboundSchema,
    ChatContentAudio$outboundSchema,
    LegacyChatContentVideo$outboundSchema,
    ChatContentText$outboundSchema.and(z.object({ type: z.literal("text") })),
    ChatContentVideo$outboundSchema,
]);
export function chatContentItemsToJSON(chatContentItems) {
    return JSON.stringify(ChatContentItems$outboundSchema.parse(chatContentItems));
}
export function chatContentItemsFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ChatContentItems$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatContentItems' from JSON`);
}
