// @openrouter/sdk/esm/models/chatcontentfile.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: ee5054f5da9f
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
/** @internal */
export const FileT$inboundSchema = z.object({
    file_data: z.string().optional(),
    file_id: z.string().optional(),
    filename: z.string().optional(),
}).transform((v) => {
    return remap$(v, {
        "file_data": "fileData",
        "file_id": "fileId",
    });
});
/** @internal */
export const FileT$outboundSchema = z.object({
    fileData: z.string().optional(),
    fileId: z.string().optional(),
    filename: z.string().optional(),
}).transform((v) => {
    return remap$(v, {
        fileData: "file_data",
        fileId: "file_id",
    });
});
export function fileToJSON(fileT) {
    return JSON.stringify(FileT$outboundSchema.parse(fileT));
}
export function fileFromJSON(jsonString) {
    return safeParse(jsonString, (x) => FileT$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'FileT' from JSON`);
}
/** @internal */
export const ChatContentFile$inboundSchema = z.object({
    file: z.lazy(() => FileT$inboundSchema),
    type: z.literal("file"),
});
/** @internal */
export const ChatContentFile$outboundSchema = z.object({
    file: z.lazy(() => FileT$outboundSchema),
    type: z.literal("file"),
});
export function chatContentFileToJSON(chatContentFile) {
    return JSON.stringify(ChatContentFile$outboundSchema.parse(chatContentFile));
}
export function chatContentFileFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ChatContentFile$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatContentFile' from JSON`);
}
