// @openrouter/sdk/esm/models/chattokenlogprob.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 267d5cc43c1b
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
/** @internal */
export const ChatTokenLogprobTopLogprob$inboundSchema = z.object({
    bytes: z.nullable(z.array(z.int())),
    logprob: z.number(),
    token: z.string(),
});
export function chatTokenLogprobTopLogprobFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ChatTokenLogprobTopLogprob$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatTokenLogprobTopLogprob' from JSON`);
}
/** @internal */
export const ChatTokenLogprob$inboundSchema = z.object({
    bytes: z.nullable(z.array(z.int())),
    logprob: z.number(),
    token: z.string(),
    top_logprobs: z.array(z.lazy(() => ChatTokenLogprobTopLogprob$inboundSchema)),
}).transform((v) => {
    return remap$(v, {
        "top_logprobs": "topLogprobs",
    });
});
export function chatTokenLogprobFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ChatTokenLogprob$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatTokenLogprob' from JSON`);
}
