// @openrouter/sdk/esm/models/chatcontentcachecontrol.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 2826e2627ca8
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { AnthropicCacheControlTtl$inboundSchema, AnthropicCacheControlTtl$outboundSchema, } from "./anthropiccachecontrolttl.js";
export const ChatContentCacheControlType = {
    Ephemeral: "ephemeral",
};
/** @internal */
export const ChatContentCacheControlType$inboundSchema = z.enum(ChatContentCacheControlType);
/** @internal */
export const ChatContentCacheControlType$outboundSchema = ChatContentCacheControlType$inboundSchema;
/** @internal */
export const ChatContentCacheControl$inboundSchema = z.object({
    ttl: AnthropicCacheControlTtl$inboundSchema.optional(),
    type: ChatContentCacheControlType$inboundSchema,
});
/** @internal */
export const ChatContentCacheControl$outboundSchema = z.object({
    ttl: AnthropicCacheControlTtl$outboundSchema.optional(),
    type: ChatContentCacheControlType$outboundSchema,
});
export function chatContentCacheControlToJSON(chatContentCacheControl) {
    return JSON.stringify(ChatContentCacheControl$outboundSchema.parse(chatContentCacheControl));
}
export function chatContentCacheControlFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ChatContentCacheControl$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatContentCacheControl' from JSON`);
}
