// @openrouter/sdk/esm/models/chatstreamchoice.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 3747d92a0ba4
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { ChatFinishReasonEnum$inboundSchema, } from "./chatfinishreasonenum.js";
import { ChatStreamDelta$inboundSchema, } from "./chatstreamdelta.js";
import { ChatTokenLogprobs$inboundSchema, } from "./chattokenlogprobs.js";
/** @internal */
export const ChatStreamChoice$inboundSchema = z.object({
    delta: ChatStreamDelta$inboundSchema,
    finish_reason: z.nullable(ChatFinishReasonEnum$inboundSchema),
    index: z.int(),
    logprobs: z.nullable(ChatTokenLogprobs$inboundSchema).optional(),
}).transform((v) => {
    return remap$(v, {
        "finish_reason": "finishReason",
    });
});
export function chatStreamChoiceFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ChatStreamChoice$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatStreamChoice' from JSON`);
}
