// @openrouter/sdk/esm/models/chattoolchoice.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: f553e5b216ba
 */
import * as z from "zod/v4";
import { ChatNamedToolChoice$outboundSchema, } from "./chatnamedtoolchoice.js";
export const ChatToolChoiceRequired = {
    Required: "required",
};
export const ChatToolChoiceAuto = {
    Auto: "auto",
};
export const ChatToolChoiceNone = {
    None: "none",
};
/** @internal */
export const ChatToolChoiceRequired$outboundSchema = z.enum(ChatToolChoiceRequired);
/** @internal */
export const ChatToolChoiceAuto$outboundSchema = z.enum(ChatToolChoiceAuto);
/** @internal */
export const ChatToolChoiceNone$outboundSchema = z.enum(ChatToolChoiceNone);
/** @internal */
export const ChatToolChoice$outboundSchema = z.union([
    ChatNamedToolChoice$outboundSchema,
    ChatToolChoiceNone$outboundSchema,
    ChatToolChoiceAuto$outboundSchema,
    ChatToolChoiceRequired$outboundSchema,
]);
export function chatToolChoiceToJSON(chatToolChoice) {
    return JSON.stringify(ChatToolChoice$outboundSchema.parse(chatToolChoice));
}
