// @openrouter/sdk/esm/models/chattoolmessage.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 446b465db30a
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { ChatContentItems$outboundSchema, } from "./chatcontentitems.js";
/** @internal */
export const ChatToolMessageContent$outboundSchema = z.union([z.string(), z.array(ChatContentItems$outboundSchema)]);
export function chatToolMessageContentToJSON(chatToolMessageContent) {
    return JSON.stringify(ChatToolMessageContent$outboundSchema.parse(chatToolMessageContent));
}
/** @internal */
export const ChatToolMessage$outboundSchema = z.object({
    content: z.union([z.string(), z.array(ChatContentItems$outboundSchema)]),
    role: z.literal("tool"),
    toolCallId: z.string(),
}).transform((v) => {
    return remap$(v, {
        toolCallId: "tool_call_id",
    });
});
export function chatToolMessageToJSON(chatToolMessage) {
    return JSON.stringify(ChatToolMessage$outboundSchema.parse(chatToolMessage));
}
