// @openrouter/sdk/esm/models/compoundfilter.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: beca2c6f6a75
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import * as openEnums from "../types/enums.js";
export const CompoundFilterType = {
    And: "and",
    Or: "or",
};
/** @internal */
export const CompoundFilterType$inboundSchema = openEnums.inboundSchema(CompoundFilterType);
/** @internal */
export const CompoundFilterType$outboundSchema = openEnums.outboundSchema(CompoundFilterType);
/** @internal */
export const CompoundFilter$inboundSchema = z.object({
    filters: z.array(z.record(z.string(), z.nullable(z.any()))),
    type: CompoundFilterType$inboundSchema,
});
/** @internal */
export const CompoundFilter$outboundSchema = z.object({
    filters: z.array(z.record(z.string(), z.nullable(z.any()))),
    type: CompoundFilterType$outboundSchema,
});
export function compoundFilterToJSON(compoundFilter) {
    return JSON.stringify(CompoundFilter$outboundSchema.parse(compoundFilter));
}
export function compoundFilterFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CompoundFilter$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CompoundFilter' from JSON`);
}
