// @openrouter/sdk/esm/models/contentpartdoneevent.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 943cfab7678a
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { discriminatedUnion } from "../types/discriminatedUnion.js";
import { OpenAIResponsesRefusalContent$inboundSchema, } from "./openairesponsesrefusalcontent.js";
import { ReasoningTextContent$inboundSchema, } from "./reasoningtextcontent.js";
import { ResponseOutputText$inboundSchema, } from "./responseoutputtext.js";
/** @internal */
export const ContentPartDoneEventPart$inboundSchema = discriminatedUnion("type", {
    output_text: ResponseOutputText$inboundSchema,
    reasoning_text: ReasoningTextContent$inboundSchema.and(z.object({ type: z.literal("reasoning_text") })),
    refusal: OpenAIResponsesRefusalContent$inboundSchema,
});
export function contentPartDoneEventPartFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ContentPartDoneEventPart$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ContentPartDoneEventPart' from JSON`);
}
/** @internal */
export const ContentPartDoneEvent$inboundSchema = z.object({
    content_index: z.int(),
    item_id: z.string(),
    output_index: z.int(),
    part: discriminatedUnion("type", {
        output_text: ResponseOutputText$inboundSchema,
        reasoning_text: ReasoningTextContent$inboundSchema.and(z.object({ type: z.literal("reasoning_text") })),
        refusal: OpenAIResponsesRefusalContent$inboundSchema,
    }),
    sequence_number: z.int(),
    type: z.literal("response.content_part.done"),
}).transform((v) => {
    return remap$(v, {
        "content_index": "contentIndex",
        "item_id": "itemId",
        "output_index": "outputIndex",
        "sequence_number": "sequenceNumber",
    });
});
export function contentPartDoneEventFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ContentPartDoneEvent$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ContentPartDoneEvent' from JSON`);
}
