// @openrouter/sdk/esm/models/contentpartimage.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 6b4d72ed5ae8
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
export const ContentPartImageType = {
    ImageUrl: "image_url",
};
/** @internal */
export const ContentPartImageImageUrl$outboundSchema = z.object({
    url: z.string(),
});
export function contentPartImageImageUrlToJSON(contentPartImageImageUrl) {
    return JSON.stringify(ContentPartImageImageUrl$outboundSchema.parse(contentPartImageImageUrl));
}
/** @internal */
export const ContentPartImageType$outboundSchema = z.enum(ContentPartImageType);
/** @internal */
export const ContentPartImage$outboundSchema = z.object({
    imageUrl: z.lazy(() => ContentPartImageImageUrl$outboundSchema),
    type: ContentPartImageType$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        imageUrl: "image_url",
    });
});
export function contentPartImageToJSON(contentPartImage) {
    return JSON.stringify(ContentPartImage$outboundSchema.parse(contentPartImage));
}
