// @openrouter/sdk/esm/models/customtool.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: b293ddf87321
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { discriminatedUnion } from "../types/discriminatedUnion.js";
import * as openEnums from "../types/enums.js";
export const Syntax = {
    Lark: "lark",
    Regex: "regex",
};
/** @internal */
export const Syntax$inboundSchema = openEnums
    .inboundSchema(Syntax);
/** @internal */
export const Syntax$outboundSchema = openEnums
    .outboundSchema(Syntax);
/** @internal */
export const FormatGrammar$inboundSchema = z
    .object({
    definition: z.string(),
    syntax: Syntax$inboundSchema,
    type: z.literal("grammar"),
});
/** @internal */
export const FormatGrammar$outboundSchema = z.object({
    definition: z.string(),
    syntax: Syntax$outboundSchema,
    type: z.literal("grammar"),
});
export function formatGrammarToJSON(formatGrammar) {
    return JSON.stringify(FormatGrammar$outboundSchema.parse(formatGrammar));
}
export function formatGrammarFromJSON(jsonString) {
    return safeParse(jsonString, (x) => FormatGrammar$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'FormatGrammar' from JSON`);
}
/** @internal */
export const FormatText$inboundSchema = z
    .object({
    type: z.literal("text"),
});
/** @internal */
export const FormatText$outboundSchema = z.object({
    type: z.literal("text"),
});
export function formatTextToJSON(formatText) {
    return JSON.stringify(FormatText$outboundSchema.parse(formatText));
}
export function formatTextFromJSON(jsonString) {
    return safeParse(jsonString, (x) => FormatText$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'FormatText' from JSON`);
}
/** @internal */
export const Format$inboundSchema = discriminatedUnion("type", {
    text: z.lazy(() => FormatText$inboundSchema),
    grammar: z.lazy(() => FormatGrammar$inboundSchema),
});
/** @internal */
export const Format$outboundSchema = z
    .union([
    z.lazy(() => FormatText$outboundSchema),
    z.lazy(() => FormatGrammar$outboundSchema),
]);
export function formatToJSON(format) {
    return JSON.stringify(Format$outboundSchema.parse(format));
}
export function formatFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Format$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Format' from JSON`);
}
/** @internal */
export const CustomTool$inboundSchema = z
    .object({
    description: z.string().optional(),
    format: discriminatedUnion("type", {
        text: z.lazy(() => FormatText$inboundSchema),
        grammar: z.lazy(() => FormatGrammar$inboundSchema),
    }).optional(),
    name: z.string(),
    type: z.literal("custom"),
});
/** @internal */
export const CustomTool$outboundSchema = z.object({
    description: z.string().optional(),
    format: z.union([
        z.lazy(() => FormatText$outboundSchema),
        z.lazy(() => FormatGrammar$outboundSchema),
    ]).optional(),
    name: z.string(),
    type: z.literal("custom"),
});
export function customToolToJSON(customTool) {
    return JSON.stringify(CustomTool$outboundSchema.parse(customTool));
}
export function customToolFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CustomTool$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CustomTool' from JSON`);
}
