// @openrouter/sdk/esm/models/defaultparameters.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 8da6fac53cb5
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
/** @internal */
export const DefaultParameters$inboundSchema = z.object({
    frequency_penalty: z.nullable(z.number()).optional(),
    presence_penalty: z.nullable(z.number()).optional(),
    repetition_penalty: z.nullable(z.number()).optional(),
    temperature: z.nullable(z.number()).optional(),
    top_k: z.nullable(z.int()).optional(),
    top_p: z.nullable(z.number()).optional(),
}).transform((v) => {
    return remap$(v, {
        "frequency_penalty": "frequencyPenalty",
        "presence_penalty": "presencePenalty",
        "repetition_penalty": "repetitionPenalty",
        "top_k": "topK",
        "top_p": "topP",
    });
});
export function defaultParametersFromJSON(jsonString) {
    return safeParse(jsonString, (x) => DefaultParameters$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'DefaultParameters' from JSON`);
}
