// @openrouter/sdk/esm/models/easyinputmessage.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 0e3bca037a2a
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import * as openEnums from "../types/enums.js";
import { InputAudio$outboundSchema, } from "./inputaudio.js";
import { InputFile$outboundSchema, } from "./inputfile.js";
import { InputText$outboundSchema, } from "./inputtext.js";
import { InputVideo$outboundSchema, } from "./inputvideo.js";
export const EasyInputMessageDetail = {
    Auto: "auto",
    High: "high",
    Low: "low",
};
export const EasyInputMessagePhaseFinalAnswer = {
    FinalAnswer: "final_answer",
};
export const EasyInputMessagePhaseCommentary = {
    Commentary: "commentary",
};
export const EasyInputMessageRoleDeveloper = {
    Developer: "developer",
};
export const EasyInputMessageRoleAssistant = {
    Assistant: "assistant",
};
export const EasyInputMessageRoleSystem = {
    System: "system",
};
export const EasyInputMessageRoleUser = {
    User: "user",
};
export const EasyInputMessageTypeMessage = {
    Message: "message",
};
/** @internal */
export const EasyInputMessageDetail$outboundSchema = openEnums.outboundSchema(EasyInputMessageDetail);
/** @internal */
export const EasyInputMessageContentInputImage$outboundSchema = z.object({
    detail: EasyInputMessageDetail$outboundSchema,
    imageUrl: z.nullable(z.string()).optional(),
    type: z.literal("input_image"),
}).transform((v) => {
    return remap$(v, {
        imageUrl: "image_url",
    });
});
export function easyInputMessageContentInputImageToJSON(easyInputMessageContentInputImage) {
    return JSON.stringify(EasyInputMessageContentInputImage$outboundSchema.parse(easyInputMessageContentInputImage));
}
/** @internal */
export const EasyInputMessageContentUnion1$outboundSchema = z.union([
    InputText$outboundSchema,
    z.lazy(() => EasyInputMessageContentInputImage$outboundSchema),
    InputFile$outboundSchema,
    InputAudio$outboundSchema,
    InputVideo$outboundSchema,
]);
export function easyInputMessageContentUnion1ToJSON(easyInputMessageContentUnion1) {
    return JSON.stringify(EasyInputMessageContentUnion1$outboundSchema.parse(easyInputMessageContentUnion1));
}
/** @internal */
export const EasyInputMessageContentUnion2$outboundSchema = z.union([
    z.array(z.union([
        InputText$outboundSchema,
        z.lazy(() => EasyInputMessageContentInputImage$outboundSchema),
        InputFile$outboundSchema,
        InputAudio$outboundSchema,
        InputVideo$outboundSchema,
    ])),
    z.string(),
    z.any(),
]);
export function easyInputMessageContentUnion2ToJSON(easyInputMessageContentUnion2) {
    return JSON.stringify(EasyInputMessageContentUnion2$outboundSchema.parse(easyInputMessageContentUnion2));
}
/** @internal */
export const EasyInputMessagePhaseFinalAnswer$outboundSchema = z.enum(EasyInputMessagePhaseFinalAnswer);
/** @internal */
export const EasyInputMessagePhaseCommentary$outboundSchema = z.enum(EasyInputMessagePhaseCommentary);
/** @internal */
export const EasyInputMessagePhaseUnion$outboundSchema = z.union([
    EasyInputMessagePhaseCommentary$outboundSchema,
    EasyInputMessagePhaseFinalAnswer$outboundSchema,
    z.any(),
]);
export function easyInputMessagePhaseUnionToJSON(easyInputMessagePhaseUnion) {
    return JSON.stringify(EasyInputMessagePhaseUnion$outboundSchema.parse(easyInputMessagePhaseUnion));
}
/** @internal */
export const EasyInputMessageRoleDeveloper$outboundSchema = z.enum(EasyInputMessageRoleDeveloper);
/** @internal */
export const EasyInputMessageRoleAssistant$outboundSchema = z.enum(EasyInputMessageRoleAssistant);
/** @internal */
export const EasyInputMessageRoleSystem$outboundSchema = z.enum(EasyInputMessageRoleSystem);
/** @internal */
export const EasyInputMessageRoleUser$outboundSchema = z.enum(EasyInputMessageRoleUser);
/** @internal */
export const EasyInputMessageRoleUnion$outboundSchema = z.union([
    EasyInputMessageRoleUser$outboundSchema,
    EasyInputMessageRoleSystem$outboundSchema,
    EasyInputMessageRoleAssistant$outboundSchema,
    EasyInputMessageRoleDeveloper$outboundSchema,
]);
export function easyInputMessageRoleUnionToJSON(easyInputMessageRoleUnion) {
    return JSON.stringify(EasyInputMessageRoleUnion$outboundSchema.parse(easyInputMessageRoleUnion));
}
/** @internal */
export const EasyInputMessageTypeMessage$outboundSchema = z.enum(EasyInputMessageTypeMessage);
/** @internal */
export const EasyInputMessage$outboundSchema = z.object({
    content: z.nullable(z.union([
        z.array(z.union([
            InputText$outboundSchema,
            z.lazy(() => EasyInputMessageContentInputImage$outboundSchema),
            InputFile$outboundSchema,
            InputAudio$outboundSchema,
            InputVideo$outboundSchema,
        ])),
        z.string(),
        z.any(),
    ])).optional(),
    phase: z.nullable(z.union([
        EasyInputMessagePhaseCommentary$outboundSchema,
        EasyInputMessagePhaseFinalAnswer$outboundSchema,
        z.any(),
    ])).optional(),
    role: z.union([
        EasyInputMessageRoleUser$outboundSchema,
        EasyInputMessageRoleSystem$outboundSchema,
        EasyInputMessageRoleAssistant$outboundSchema,
        EasyInputMessageRoleDeveloper$outboundSchema,
    ]),
    type: EasyInputMessageTypeMessage$outboundSchema.optional(),
});
export function easyInputMessageToJSON(easyInputMessage) {
    return JSON.stringify(EasyInputMessage$outboundSchema.parse(easyInputMessage));
}
