// @openrouter/sdk/esm/models/filecitation.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 0e609723dbc8
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
/** @internal */
export const FileCitation$inboundSchema = z
    .object({
    file_id: z.string(),
    filename: z.string(),
    index: z.int(),
    type: z.literal("file_citation"),
}).transform((v) => {
    return remap$(v, {
        "file_id": "fileId",
    });
});
/** @internal */
export const FileCitation$outboundSchema = z.object({
    fileId: z.string(),
    filename: z.string(),
    index: z.int(),
    type: z.literal("file_citation"),
}).transform((v) => {
    return remap$(v, {
        fileId: "file_id",
    });
});
export function fileCitationToJSON(fileCitation) {
    return JSON.stringify(FileCitation$outboundSchema.parse(fileCitation));
}
export function fileCitationFromJSON(jsonString) {
    return safeParse(jsonString, (x) => FileCitation$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'FileCitation' from JSON`);
}
