// @openrouter/sdk/esm/models/formats.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 6e4608e1920c
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { discriminatedUnion } from "../types/discriminatedUnion.js";
import { FormatJsonObjectConfig$inboundSchema, FormatJsonObjectConfig$outboundSchema, } from "./formatjsonobjectconfig.js";
import { FormatJsonSchemaConfig$inboundSchema, FormatJsonSchemaConfig$outboundSchema, } from "./formatjsonschemaconfig.js";
import { FormatTextConfig$inboundSchema, FormatTextConfig$outboundSchema, } from "./formattextconfig.js";
/** @internal */
export const Formats$inboundSchema = discriminatedUnion("type", {
    text: FormatTextConfig$inboundSchema,
    json_object: FormatJsonObjectConfig$inboundSchema,
    json_schema: FormatJsonSchemaConfig$inboundSchema,
});
/** @internal */
export const Formats$outboundSchema = z
    .union([
    FormatTextConfig$outboundSchema,
    FormatJsonObjectConfig$outboundSchema,
    FormatJsonSchemaConfig$outboundSchema,
]);
export function formatsToJSON(formats) {
    return JSON.stringify(Formats$outboundSchema.parse(formats));
}
export function formatsFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Formats$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Formats' from JSON`);
}
