// @openrouter/sdk/esm/models/frameimage.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 97fccf1b68ca
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import * as openEnums from "../types/enums.js";
export const FrameImageType = {
    ImageUrl: "image_url",
};
/**
 * Whether this image represents the first or last frame of the video
 */
export const FrameType = {
    FirstFrame: "first_frame",
    LastFrame: "last_frame",
};
/** @internal */
export const FrameImageImageUrl$outboundSchema = z.object({
    url: z.string(),
});
export function frameImageImageUrlToJSON(frameImageImageUrl) {
    return JSON.stringify(FrameImageImageUrl$outboundSchema.parse(frameImageImageUrl));
}
/** @internal */
export const FrameImageType$outboundSchema = z
    .enum(FrameImageType);
/** @internal */
export const FrameType$outboundSchema = openEnums
    .outboundSchema(FrameType);
/** @internal */
export const FrameImage$outboundSchema = z.object({
    imageUrl: z.lazy(() => FrameImageImageUrl$outboundSchema),
    type: FrameImageType$outboundSchema,
    frameType: FrameType$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        imageUrl: "image_url",
        frameType: "frame_type",
    });
});
export function frameImageToJSON(frameImage) {
    return JSON.stringify(FrameImage$outboundSchema.parse(frameImage));
}
