// @openrouter/sdk/esm/models/functioncalloutputitem.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 279b91cadffe
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import * as openEnums from "../types/enums.js";
import { InputFile$outboundSchema, } from "./inputfile.js";
import { InputText$outboundSchema, } from "./inputtext.js";
export const FunctionCallOutputItemDetail = {
    Auto: "auto",
    High: "high",
    Low: "low",
};
export const FunctionCallOutputItemStatus = {
    InProgress: "in_progress",
    Completed: "completed",
    Incomplete: "incomplete",
};
export const FunctionCallOutputItemTypeFunctionCallOutput = {
    FunctionCallOutput: "function_call_output",
};
/** @internal */
export const FunctionCallOutputItemDetail$outboundSchema = openEnums.outboundSchema(FunctionCallOutputItemDetail);
/** @internal */
export const OutputInputImage$outboundSchema = z.object({
    detail: FunctionCallOutputItemDetail$outboundSchema,
    imageUrl: z.nullable(z.string()).optional(),
    type: z.literal("input_image"),
}).transform((v) => {
    return remap$(v, {
        imageUrl: "image_url",
    });
});
export function outputInputImageToJSON(outputInputImage) {
    return JSON.stringify(OutputInputImage$outboundSchema.parse(outputInputImage));
}
/** @internal */
export const FunctionCallOutputItemOutputUnion1$outboundSchema = z.union([
    InputText$outboundSchema,
    z.lazy(() => OutputInputImage$outboundSchema),
    InputFile$outboundSchema,
]);
export function functionCallOutputItemOutputUnion1ToJSON(functionCallOutputItemOutputUnion1) {
    return JSON.stringify(FunctionCallOutputItemOutputUnion1$outboundSchema.parse(functionCallOutputItemOutputUnion1));
}
/** @internal */
export const FunctionCallOutputItemOutputUnion2$outboundSchema = z.union([
    z.string(),
    z.array(z.union([
        InputText$outboundSchema,
        z.lazy(() => OutputInputImage$outboundSchema),
        InputFile$outboundSchema,
    ])),
]);
export function functionCallOutputItemOutputUnion2ToJSON(functionCallOutputItemOutputUnion2) {
    return JSON.stringify(FunctionCallOutputItemOutputUnion2$outboundSchema.parse(functionCallOutputItemOutputUnion2));
}
/** @internal */
export const FunctionCallOutputItemStatus$outboundSchema = openEnums.outboundSchema(FunctionCallOutputItemStatus);
/** @internal */
export const FunctionCallOutputItemTypeFunctionCallOutput$outboundSchema = z.enum(FunctionCallOutputItemTypeFunctionCallOutput);
/** @internal */
export const FunctionCallOutputItem$outboundSchema = z.object({
    callId: z.string(),
    id: z.nullable(z.string()).optional(),
    output: z.union([
        z.string(),
        z.array(z.union([
            InputText$outboundSchema,
            z.lazy(() => OutputInputImage$outboundSchema),
            InputFile$outboundSchema,
        ])),
    ]),
    status: z.nullable(FunctionCallOutputItemStatus$outboundSchema).optional(),
    type: FunctionCallOutputItemTypeFunctionCallOutput$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        callId: "call_id",
    });
});
export function functionCallOutputItemToJSON(functionCallOutputItem) {
    return JSON.stringify(FunctionCallOutputItem$outboundSchema.parse(functionCallOutputItem));
}
