// @openrouter/sdk/esm/models/guardrail.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 240a4b8d7a7d
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { GuardrailInterval$inboundSchema, } from "./guardrailinterval.js";
/** @internal */
export const Guardrail$inboundSchema = z.object({
    allowed_models: z.nullable(z.array(z.string())).optional(),
    allowed_providers: z.nullable(z.array(z.string())).optional(),
    created_at: z.string(),
    description: z.nullable(z.string()).optional(),
    enforce_zdr: z.nullable(z.boolean()).optional(),
    id: z.string(),
    ignored_models: z.nullable(z.array(z.string())).optional(),
    ignored_providers: z.nullable(z.array(z.string())).optional(),
    limit_usd: z.nullable(z.number()).optional(),
    name: z.string(),
    reset_interval: z.nullable(GuardrailInterval$inboundSchema).optional(),
    updated_at: z.nullable(z.string()).optional(),
    workspace_id: z.string(),
}).transform((v) => {
    return remap$(v, {
        "allowed_models": "allowedModels",
        "allowed_providers": "allowedProviders",
        "created_at": "createdAt",
        "enforce_zdr": "enforceZdr",
        "ignored_models": "ignoredModels",
        "ignored_providers": "ignoredProviders",
        "limit_usd": "limitUsd",
        "reset_interval": "resetInterval",
        "updated_at": "updatedAt",
        "workspace_id": "workspaceId",
    });
});
export function guardrailFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Guardrail$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Guardrail' from JSON`);
}
