// @openrouter/sdk/esm/models/imagegenerationservertool.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 1efbad50f534
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import * as openEnums from "../types/enums.js";
export const Background = {
    Transparent: "transparent",
    Opaque: "opaque",
    Auto: "auto",
};
export const InputFidelity = {
    High: "high",
    Low: "low",
};
export const ModelEnum = {
    GptImage1: "gpt-image-1",
    GptImage1Mini: "gpt-image-1-mini",
};
export const Moderation = {
    Auto: "auto",
    Low: "low",
};
export const OutputFormat = {
    Png: "png",
    Webp: "webp",
    Jpeg: "jpeg",
};
export const Quality = {
    Low: "low",
    Medium: "medium",
    High: "high",
    Auto: "auto",
};
export const Size = {
    OneThousandAndTwentyFourx1024: "1024x1024",
    OneThousandAndTwentyFourx1536: "1024x1536",
    OneThousandFiveHundredAndThirtySixx1024: "1536x1024",
    Auto: "auto",
};
/** @internal */
export const Background$inboundSchema = openEnums.inboundSchema(Background);
/** @internal */
export const Background$outboundSchema = openEnums.outboundSchema(Background);
/** @internal */
export const InputFidelity$inboundSchema = openEnums.inboundSchema(InputFidelity);
/** @internal */
export const InputFidelity$outboundSchema = openEnums.outboundSchema(InputFidelity);
/** @internal */
export const InputImageMask$inboundSchema = z.object({
    file_id: z.string().optional(),
    image_url: z.string().optional(),
}).transform((v) => {
    return remap$(v, {
        "file_id": "fileId",
        "image_url": "imageUrl",
    });
});
/** @internal */
export const InputImageMask$outboundSchema = z.object({
    fileId: z.string().optional(),
    imageUrl: z.string().optional(),
}).transform((v) => {
    return remap$(v, {
        fileId: "file_id",
        imageUrl: "image_url",
    });
});
export function inputImageMaskToJSON(inputImageMask) {
    return JSON.stringify(InputImageMask$outboundSchema.parse(inputImageMask));
}
export function inputImageMaskFromJSON(jsonString) {
    return safeParse(jsonString, (x) => InputImageMask$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'InputImageMask' from JSON`);
}
/** @internal */
export const ModelEnum$inboundSchema = openEnums
    .inboundSchema(ModelEnum);
/** @internal */
export const ModelEnum$outboundSchema = openEnums
    .outboundSchema(ModelEnum);
/** @internal */
export const Moderation$inboundSchema = openEnums.inboundSchema(Moderation);
/** @internal */
export const Moderation$outboundSchema = openEnums.outboundSchema(Moderation);
/** @internal */
export const OutputFormat$inboundSchema = openEnums.inboundSchema(OutputFormat);
/** @internal */
export const OutputFormat$outboundSchema = openEnums.outboundSchema(OutputFormat);
/** @internal */
export const Quality$inboundSchema = openEnums
    .inboundSchema(Quality);
/** @internal */
export const Quality$outboundSchema = openEnums
    .outboundSchema(Quality);
/** @internal */
export const Size$inboundSchema = openEnums
    .inboundSchema(Size);
/** @internal */
export const Size$outboundSchema = openEnums
    .outboundSchema(Size);
/** @internal */
export const ImageGenerationServerTool$inboundSchema = z.object({
    background: Background$inboundSchema.optional(),
    input_fidelity: z.nullable(InputFidelity$inboundSchema).optional(),
    input_image_mask: z.lazy(() => InputImageMask$inboundSchema).optional(),
    model: ModelEnum$inboundSchema.optional(),
    moderation: Moderation$inboundSchema.optional(),
    output_compression: z.int().optional(),
    output_format: OutputFormat$inboundSchema.optional(),
    partial_images: z.int().optional(),
    quality: Quality$inboundSchema.optional(),
    size: Size$inboundSchema.optional(),
    type: z.literal("image_generation"),
}).transform((v) => {
    return remap$(v, {
        "input_fidelity": "inputFidelity",
        "input_image_mask": "inputImageMask",
        "output_compression": "outputCompression",
        "output_format": "outputFormat",
        "partial_images": "partialImages",
    });
});
/** @internal */
export const ImageGenerationServerTool$outboundSchema = z.object({
    background: Background$outboundSchema.optional(),
    inputFidelity: z.nullable(InputFidelity$outboundSchema).optional(),
    inputImageMask: z.lazy(() => InputImageMask$outboundSchema).optional(),
    model: ModelEnum$outboundSchema.optional(),
    moderation: Moderation$outboundSchema.optional(),
    outputCompression: z.int().optional(),
    outputFormat: OutputFormat$outboundSchema.optional(),
    partialImages: z.int().optional(),
    quality: Quality$outboundSchema.optional(),
    size: Size$outboundSchema.optional(),
    type: z.literal("image_generation"),
}).transform((v) => {
    return remap$(v, {
        inputFidelity: "input_fidelity",
        inputImageMask: "input_image_mask",
        outputCompression: "output_compression",
        outputFormat: "output_format",
        partialImages: "partial_images",
    });
});
export function imageGenerationServerToolToJSON(imageGenerationServerTool) {
    return JSON.stringify(ImageGenerationServerTool$outboundSchema.parse(imageGenerationServerTool));
}
export function imageGenerationServerToolFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ImageGenerationServerTool$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ImageGenerationServerTool' from JSON`);
}
