// @openrouter/sdk/esm/models/inputimage.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 7a718177a217
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import * as openEnums from "../types/enums.js";
export const InputImageDetail = {
    Auto: "auto",
    High: "high",
    Low: "low",
};
export const InputImageTypeEnum = {
    InputImage: "input_image",
};
/** @internal */
export const InputImageDetail$inboundSchema = openEnums.inboundSchema(InputImageDetail);
/** @internal */
export const InputImageDetail$outboundSchema = openEnums.outboundSchema(InputImageDetail);
/** @internal */
export const InputImageTypeEnum$inboundSchema = z.enum(InputImageTypeEnum);
/** @internal */
export const InputImageTypeEnum$outboundSchema = InputImageTypeEnum$inboundSchema;
/** @internal */
export const InputImage$inboundSchema = z
    .object({
    detail: InputImageDetail$inboundSchema,
    image_url: z.nullable(z.string()).optional(),
    type: InputImageTypeEnum$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "image_url": "imageUrl",
    });
});
/** @internal */
export const InputImage$outboundSchema = z.object({
    detail: InputImageDetail$outboundSchema,
    imageUrl: z.nullable(z.string()).optional(),
    type: InputImageTypeEnum$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        imageUrl: "image_url",
    });
});
export function inputImageToJSON(inputImage) {
    return JSON.stringify(InputImage$outboundSchema.parse(inputImage));
}
export function inputImageFromJSON(jsonString) {
    return safeParse(jsonString, (x) => InputImage$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'InputImage' from JSON`);
}
