// @openrouter/sdk/esm/models/inputaudio.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 9bdc14c7565f
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import * as openEnums from "../types/enums.js";
export const FormatEnum = {
    Mp3: "mp3",
    Wav: "wav",
};
/** @internal */
export const FormatEnum$inboundSchema = openEnums.inboundSchema(FormatEnum);
/** @internal */
export const FormatEnum$outboundSchema = openEnums.outboundSchema(FormatEnum);
/** @internal */
export const InputAudioInputAudio$inboundSchema = z.object({
    data: z.string(),
    format: FormatEnum$inboundSchema,
});
/** @internal */
export const InputAudioInputAudio$outboundSchema = z.object({
    data: z.string(),
    format: FormatEnum$outboundSchema,
});
export function inputAudioInputAudioToJSON(inputAudioInputAudio) {
    return JSON.stringify(InputAudioInputAudio$outboundSchema.parse(inputAudioInputAudio));
}
export function inputAudioInputAudioFromJSON(jsonString) {
    return safeParse(jsonString, (x) => InputAudioInputAudio$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'InputAudioInputAudio' from JSON`);
}
/** @internal */
export const InputAudio$inboundSchema = z
    .object({
    input_audio: z.lazy(() => InputAudioInputAudio$inboundSchema),
    type: z.literal("input_audio"),
}).transform((v) => {
    return remap$(v, {
        "input_audio": "inputAudio",
    });
});
/** @internal */
export const InputAudio$outboundSchema = z.object({
    inputAudio: z.lazy(() => InputAudioInputAudio$outboundSchema),
    type: z.literal("input_audio"),
}).transform((v) => {
    return remap$(v, {
        inputAudio: "input_audio",
    });
});
export function inputAudioToJSON(inputAudio) {
    return JSON.stringify(InputAudio$outboundSchema.parse(inputAudio));
}
export function inputAudioFromJSON(jsonString) {
    return safeParse(jsonString, (x) => InputAudio$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'InputAudio' from JSON`);
}
