// @openrouter/sdk/esm/models/legacywebsearchservertool.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 1a9c64204177
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { SearchContextSizeEnum$inboundSchema, SearchContextSizeEnum$outboundSchema, } from "./searchcontextsizeenum.js";
import { WebSearchDomainFilter$inboundSchema, WebSearchDomainFilter$outboundSchema, } from "./websearchdomainfilter.js";
import { WebSearchEngineEnum$inboundSchema, WebSearchEngineEnum$outboundSchema, } from "./websearchengineenum.js";
import { WebSearchUserLocation$inboundSchema, WebSearchUserLocation$outboundSchema, } from "./websearchuserlocation.js";
/** @internal */
export const LegacyWebSearchServerTool$inboundSchema = z.object({
    engine: WebSearchEngineEnum$inboundSchema.optional(),
    filters: z.nullable(WebSearchDomainFilter$inboundSchema).optional(),
    max_results: z.int().optional(),
    search_context_size: SearchContextSizeEnum$inboundSchema.optional(),
    type: z.literal("web_search"),
    user_location: z.nullable(WebSearchUserLocation$inboundSchema).optional(),
}).transform((v) => {
    return remap$(v, {
        "max_results": "maxResults",
        "search_context_size": "searchContextSize",
        "user_location": "userLocation",
    });
});
/** @internal */
export const LegacyWebSearchServerTool$outboundSchema = z.object({
    engine: WebSearchEngineEnum$outboundSchema.optional(),
    filters: z.nullable(WebSearchDomainFilter$outboundSchema).optional(),
    maxResults: z.int().optional(),
    searchContextSize: SearchContextSizeEnum$outboundSchema.optional(),
    type: z.literal("web_search"),
    userLocation: z.nullable(WebSearchUserLocation$outboundSchema).optional(),
}).transform((v) => {
    return remap$(v, {
        maxResults: "max_results",
        searchContextSize: "search_context_size",
        userLocation: "user_location",
    });
});
export function legacyWebSearchServerToolToJSON(legacyWebSearchServerTool) {
    return JSON.stringify(LegacyWebSearchServerTool$outboundSchema.parse(legacyWebSearchServerTool));
}
export function legacyWebSearchServerToolFromJSON(jsonString) {
    return safeParse(jsonString, (x) => LegacyWebSearchServerTool$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'LegacyWebSearchServerTool' from JSON`);
}
