// @openrouter/sdk/esm/models/legacychatcontentvideo.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: baf6fcb0034a
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { ChatContentVideoInput$inboundSchema, ChatContentVideoInput$outboundSchema, } from "./chatcontentvideoinput.js";
/** @internal */
export const LegacyChatContentVideo$inboundSchema = z.object({
    type: z.literal("input_video"),
    video_url: ChatContentVideoInput$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "video_url": "videoUrl",
    });
});
/** @internal */
export const LegacyChatContentVideo$outboundSchema = z.object({
    type: z.literal("input_video"),
    videoUrl: ChatContentVideoInput$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        videoUrl: "video_url",
    });
});
export function legacyChatContentVideoToJSON(legacyChatContentVideo) {
    return JSON.stringify(LegacyChatContentVideo$outboundSchema.parse(legacyChatContentVideo));
}
export function legacyChatContentVideoFromJSON(jsonString) {
    return safeParse(jsonString, (x) => LegacyChatContentVideo$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'LegacyChatContentVideo' from JSON`);
}
