// @openrouter/sdk/esm/models/mcpservertool.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 5e5a26b93ed7
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import * as openEnums from "../types/enums.js";
export const ConnectorId = {
    ConnectorDropbox: "connector_dropbox",
    ConnectorGmail: "connector_gmail",
    ConnectorGooglecalendar: "connector_googlecalendar",
    ConnectorGoogledrive: "connector_googledrive",
    ConnectorMicrosoftteams: "connector_microsoftteams",
    ConnectorOutlookcalendar: "connector_outlookcalendar",
    ConnectorOutlookemail: "connector_outlookemail",
    ConnectorSharepoint: "connector_sharepoint",
};
export const RequireApprovalNever = {
    Never: "never",
};
export const RequireApprovalAlways = {
    Always: "always",
};
/** @internal */
export const AllowedTools$inboundSchema = z
    .object({
    read_only: z.boolean().optional(),
    tool_names: z.array(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        "read_only": "readOnly",
        "tool_names": "toolNames",
    });
});
/** @internal */
export const AllowedTools$outboundSchema = z.object({
    readOnly: z.boolean().optional(),
    toolNames: z.array(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        readOnly: "read_only",
        toolNames: "tool_names",
    });
});
export function allowedToolsToJSON(allowedTools) {
    return JSON.stringify(AllowedTools$outboundSchema.parse(allowedTools));
}
export function allowedToolsFromJSON(jsonString) {
    return safeParse(jsonString, (x) => AllowedTools$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'AllowedTools' from JSON`);
}
/** @internal */
export const AllowedToolsUnion$inboundSchema = z.union([
    z.array(z.string()),
    z.lazy(() => AllowedTools$inboundSchema),
    z.any(),
]);
/** @internal */
export const AllowedToolsUnion$outboundSchema = z.union([
    z.array(z.string()),
    z.lazy(() => AllowedTools$outboundSchema),
    z.any(),
]);
export function allowedToolsUnionToJSON(allowedToolsUnion) {
    return JSON.stringify(AllowedToolsUnion$outboundSchema.parse(allowedToolsUnion));
}
export function allowedToolsUnionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => AllowedToolsUnion$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'AllowedToolsUnion' from JSON`);
}
/** @internal */
export const ConnectorId$inboundSchema = openEnums.inboundSchema(ConnectorId);
/** @internal */
export const ConnectorId$outboundSchema = openEnums.outboundSchema(ConnectorId);
/** @internal */
export const RequireApprovalNever$inboundSchema = z.enum(RequireApprovalNever);
/** @internal */
export const RequireApprovalNever$outboundSchema = RequireApprovalNever$inboundSchema;
/** @internal */
export const RequireApprovalAlways$inboundSchema = z.enum(RequireApprovalAlways);
/** @internal */
export const RequireApprovalAlways$outboundSchema = RequireApprovalAlways$inboundSchema;
/** @internal */
export const Always$inboundSchema = z.object({
    tool_names: z.array(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        "tool_names": "toolNames",
    });
});
/** @internal */
export const Always$outboundSchema = z
    .object({
    toolNames: z.array(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        toolNames: "tool_names",
    });
});
export function alwaysToJSON(always) {
    return JSON.stringify(Always$outboundSchema.parse(always));
}
export function alwaysFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Always$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Always' from JSON`);
}
/** @internal */
export const Never$inboundSchema = z.object({
    tool_names: z.array(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        "tool_names": "toolNames",
    });
});
/** @internal */
export const Never$outboundSchema = z.object({
    toolNames: z.array(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        toolNames: "tool_names",
    });
});
export function neverToJSON(never) {
    return JSON.stringify(Never$outboundSchema.parse(never));
}
export function neverFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Never$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Never' from JSON`);
}
/** @internal */
export const RequireApproval$inboundSchema = z.object({
    always: z.lazy(() => Always$inboundSchema).optional(),
    never: z.lazy(() => Never$inboundSchema).optional(),
});
/** @internal */
export const RequireApproval$outboundSchema = z.object({
    always: z.lazy(() => Always$outboundSchema).optional(),
    never: z.lazy(() => Never$outboundSchema).optional(),
});
export function requireApprovalToJSON(requireApproval) {
    return JSON.stringify(RequireApproval$outboundSchema.parse(requireApproval));
}
export function requireApprovalFromJSON(jsonString) {
    return safeParse(jsonString, (x) => RequireApproval$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'RequireApproval' from JSON`);
}
/** @internal */
export const RequireApprovalUnion$inboundSchema = z.union([
    z.lazy(() => RequireApproval$inboundSchema),
    RequireApprovalAlways$inboundSchema,
    RequireApprovalNever$inboundSchema,
    z.any(),
]);
/** @internal */
export const RequireApprovalUnion$outboundSchema = z.union([
    z.lazy(() => RequireApproval$outboundSchema),
    RequireApprovalAlways$outboundSchema,
    RequireApprovalNever$outboundSchema,
    z.any(),
]);
export function requireApprovalUnionToJSON(requireApprovalUnion) {
    return JSON.stringify(RequireApprovalUnion$outboundSchema.parse(requireApprovalUnion));
}
export function requireApprovalUnionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => RequireApprovalUnion$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'RequireApprovalUnion' from JSON`);
}
/** @internal */
export const McpServerTool$inboundSchema = z
    .object({
    allowed_tools: z.nullable(z.union([
        z.array(z.string()),
        z.lazy(() => AllowedTools$inboundSchema),
        z.any(),
    ])).optional(),
    authorization: z.string().optional(),
    connector_id: ConnectorId$inboundSchema.optional(),
    headers: z.nullable(z.record(z.string(), z.string())).optional(),
    require_approval: z.nullable(z.union([
        z.lazy(() => RequireApproval$inboundSchema),
        RequireApprovalAlways$inboundSchema,
        RequireApprovalNever$inboundSchema,
        z.any(),
    ])).optional(),
    server_description: z.string().optional(),
    server_label: z.string(),
    server_url: z.string().optional(),
    type: z.literal("mcp"),
}).transform((v) => {
    return remap$(v, {
        "allowed_tools": "allowedTools",
        "connector_id": "connectorId",
        "require_approval": "requireApproval",
        "server_description": "serverDescription",
        "server_label": "serverLabel",
        "server_url": "serverUrl",
    });
});
/** @internal */
export const McpServerTool$outboundSchema = z.object({
    allowedTools: z.nullable(z.union([
        z.array(z.string()),
        z.lazy(() => AllowedTools$outboundSchema),
        z.any(),
    ])).optional(),
    authorization: z.string().optional(),
    connectorId: ConnectorId$outboundSchema.optional(),
    headers: z.nullable(z.record(z.string(), z.string())).optional(),
    requireApproval: z.nullable(z.union([
        z.lazy(() => RequireApproval$outboundSchema),
        RequireApprovalAlways$outboundSchema,
        RequireApprovalNever$outboundSchema,
        z.any(),
    ])).optional(),
    serverDescription: z.string().optional(),
    serverLabel: z.string(),
    serverUrl: z.string().optional(),
    type: z.literal("mcp"),
}).transform((v) => {
    return remap$(v, {
        allowedTools: "allowed_tools",
        connectorId: "connector_id",
        requireApproval: "require_approval",
        serverDescription: "server_description",
        serverLabel: "server_label",
        serverUrl: "server_url",
    });
});
export function mcpServerToolToJSON(mcpServerTool) {
    return JSON.stringify(McpServerTool$outboundSchema.parse(mcpServerTool));
}
export function mcpServerToolFromJSON(jsonString) {
    return safeParse(jsonString, (x) => McpServerTool$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'McpServerTool' from JSON`);
}
