// @openrouter/sdk/esm/models/model.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: b99d4c14e794
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { DefaultParameters$inboundSchema, } from "./defaultparameters.js";
import { ModelArchitecture$inboundSchema, } from "./modelarchitecture.js";
import { ModelLinks$inboundSchema } from "./modellinks.js";
import { Parameter$inboundSchema } from "./parameter.js";
import { PerRequestLimits$inboundSchema, } from "./perrequestlimits.js";
import { PublicPricing$inboundSchema } from "./publicpricing.js";
import { TopProviderInfo$inboundSchema, } from "./topproviderinfo.js";
/** @internal */
export const Model$inboundSchema = z.object({
    architecture: ModelArchitecture$inboundSchema,
    canonical_slug: z.string(),
    context_length: z.nullable(z.int()),
    created: z.int(),
    default_parameters: z.nullable(DefaultParameters$inboundSchema),
    description: z.string().optional(),
    expiration_date: z.nullable(z.string()).optional(),
    hugging_face_id: z.nullable(z.string()).optional(),
    id: z.string(),
    knowledge_cutoff: z.nullable(z.string()).optional(),
    links: ModelLinks$inboundSchema,
    name: z.string(),
    per_request_limits: z.nullable(PerRequestLimits$inboundSchema),
    pricing: PublicPricing$inboundSchema,
    supported_parameters: z.array(Parameter$inboundSchema),
    top_provider: TopProviderInfo$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "canonical_slug": "canonicalSlug",
        "context_length": "contextLength",
        "default_parameters": "defaultParameters",
        "expiration_date": "expirationDate",
        "hugging_face_id": "huggingFaceId",
        "knowledge_cutoff": "knowledgeCutoff",
        "per_request_limits": "perRequestLimits",
        "supported_parameters": "supportedParameters",
        "top_provider": "topProvider",
    });
});
export function modelFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Model$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Model' from JSON`);
}
