// @openrouter/sdk/esm/models/modelarchitecture.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 57607576095f
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import * as openEnums from "../types/enums.js";
import { InputModality$inboundSchema } from "./inputmodality.js";
import { ModelGroup$inboundSchema } from "./modelgroup.js";
import { OutputModality$inboundSchema, } from "./outputmodality.js";
/**
 * Instruction format type
 */
export const ModelArchitectureInstructType = {
    None: "none",
    Airoboros: "airoboros",
    Alpaca: "alpaca",
    AlpacaModif: "alpaca-modif",
    Chatml: "chatml",
    Claude: "claude",
    CodeLlama: "code-llama",
    Gemma: "gemma",
    Llama2: "llama2",
    Llama3: "llama3",
    Mistral: "mistral",
    Nemotron: "nemotron",
    Neural: "neural",
    Openchat: "openchat",
    Phi3: "phi3",
    Rwkv: "rwkv",
    Vicuna: "vicuna",
    Zephyr: "zephyr",
    DeepseekR1: "deepseek-r1",
    DeepseekV31: "deepseek-v3.1",
    Qwq: "qwq",
    Qwen3: "qwen3",
};
/** @internal */
export const ModelArchitectureInstructType$inboundSchema = openEnums.inboundSchema(ModelArchitectureInstructType);
/** @internal */
export const ModelArchitecture$inboundSchema = z.object({
    input_modalities: z.array(InputModality$inboundSchema),
    instruct_type: z.nullable(ModelArchitectureInstructType$inboundSchema)
        .optional(),
    modality: z.nullable(z.string()),
    output_modalities: z.array(OutputModality$inboundSchema),
    tokenizer: ModelGroup$inboundSchema.optional(),
}).transform((v) => {
    return remap$(v, {
        "input_modalities": "inputModalities",
        "instruct_type": "instructType",
        "output_modalities": "outputModalities",
    });
});
export function modelArchitectureFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ModelArchitecture$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ModelArchitecture' from JSON`);
}
