// @openrouter/sdk/esm/models/openairesponsestoolchoiceunion.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 73d80db9cd46
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { ToolChoiceAllowed$inboundSchema, ToolChoiceAllowed$outboundSchema, } from "./toolchoiceallowed.js";
export const OpenAIResponsesToolChoiceTypeWebSearchPreview = {
    WebSearchPreview: "web_search_preview",
};
export const OpenAIResponsesToolChoiceTypeWebSearchPreview20250311 = {
    WebSearchPreview20250311: "web_search_preview_2025_03_11",
};
export const OpenAIResponsesToolChoiceTypeFunction = {
    Function: "function",
};
export const OpenAIResponsesToolChoiceRequired = {
    Required: "required",
};
export const OpenAIResponsesToolChoiceNone = {
    None: "none",
};
export const OpenAIResponsesToolChoiceAuto = {
    Auto: "auto",
};
/** @internal */
export const OpenAIResponsesToolChoiceTypeWebSearchPreview$inboundSchema = z.enum(OpenAIResponsesToolChoiceTypeWebSearchPreview);
/** @internal */
export const OpenAIResponsesToolChoiceTypeWebSearchPreview$outboundSchema = OpenAIResponsesToolChoiceTypeWebSearchPreview$inboundSchema;
/** @internal */
export const OpenAIResponsesToolChoiceTypeWebSearchPreview20250311$inboundSchema = z
    .enum(OpenAIResponsesToolChoiceTypeWebSearchPreview20250311);
/** @internal */
export const OpenAIResponsesToolChoiceTypeWebSearchPreview20250311$outboundSchema = OpenAIResponsesToolChoiceTypeWebSearchPreview20250311$inboundSchema;
/** @internal */
export const Type$inboundSchema = z.union([
    OpenAIResponsesToolChoiceTypeWebSearchPreview20250311$inboundSchema,
    OpenAIResponsesToolChoiceTypeWebSearchPreview$inboundSchema,
]);
/** @internal */
export const Type$outboundSchema = z.union([
    OpenAIResponsesToolChoiceTypeWebSearchPreview20250311$outboundSchema,
    OpenAIResponsesToolChoiceTypeWebSearchPreview$outboundSchema,
]);
export function typeToJSON(type) {
    return JSON.stringify(Type$outboundSchema.parse(type));
}
export function typeFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Type$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Type' from JSON`);
}
/** @internal */
export const OpenAIResponsesToolChoice$inboundSchema = z.object({
    type: z.union([
        OpenAIResponsesToolChoiceTypeWebSearchPreview20250311$inboundSchema,
        OpenAIResponsesToolChoiceTypeWebSearchPreview$inboundSchema,
    ]),
});
/** @internal */
export const OpenAIResponsesToolChoice$outboundSchema = z.object({
    type: z.union([
        OpenAIResponsesToolChoiceTypeWebSearchPreview20250311$outboundSchema,
        OpenAIResponsesToolChoiceTypeWebSearchPreview$outboundSchema,
    ]),
});
export function openAIResponsesToolChoiceToJSON(openAIResponsesToolChoice) {
    return JSON.stringify(OpenAIResponsesToolChoice$outboundSchema.parse(openAIResponsesToolChoice));
}
export function openAIResponsesToolChoiceFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OpenAIResponsesToolChoice$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OpenAIResponsesToolChoice' from JSON`);
}
/** @internal */
export const OpenAIResponsesToolChoiceTypeFunction$inboundSchema = z.enum(OpenAIResponsesToolChoiceTypeFunction);
/** @internal */
export const OpenAIResponsesToolChoiceTypeFunction$outboundSchema = OpenAIResponsesToolChoiceTypeFunction$inboundSchema;
/** @internal */
export const OpenAIResponsesToolChoiceFunction$inboundSchema = z.object({
    name: z.string(),
    type: OpenAIResponsesToolChoiceTypeFunction$inboundSchema,
});
/** @internal */
export const OpenAIResponsesToolChoiceFunction$outboundSchema = z.object({
    name: z.string(),
    type: OpenAIResponsesToolChoiceTypeFunction$outboundSchema,
});
export function openAIResponsesToolChoiceFunctionToJSON(openAIResponsesToolChoiceFunction) {
    return JSON.stringify(OpenAIResponsesToolChoiceFunction$outboundSchema.parse(openAIResponsesToolChoiceFunction));
}
export function openAIResponsesToolChoiceFunctionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OpenAIResponsesToolChoiceFunction$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OpenAIResponsesToolChoiceFunction' from JSON`);
}
/** @internal */
export const OpenAIResponsesToolChoiceRequired$inboundSchema = z.enum(OpenAIResponsesToolChoiceRequired);
/** @internal */
export const OpenAIResponsesToolChoiceRequired$outboundSchema = OpenAIResponsesToolChoiceRequired$inboundSchema;
/** @internal */
export const OpenAIResponsesToolChoiceNone$inboundSchema = z.enum(OpenAIResponsesToolChoiceNone);
/** @internal */
export const OpenAIResponsesToolChoiceNone$outboundSchema = OpenAIResponsesToolChoiceNone$inboundSchema;
/** @internal */
export const OpenAIResponsesToolChoiceAuto$inboundSchema = z.enum(OpenAIResponsesToolChoiceAuto);
/** @internal */
export const OpenAIResponsesToolChoiceAuto$outboundSchema = OpenAIResponsesToolChoiceAuto$inboundSchema;
/** @internal */
export const OpenAIResponsesToolChoiceUnion$inboundSchema = z.union([
    ToolChoiceAllowed$inboundSchema,
    z.lazy(() => OpenAIResponsesToolChoiceFunction$inboundSchema),
    z.lazy(() => OpenAIResponsesToolChoice$inboundSchema),
    OpenAIResponsesToolChoiceAuto$inboundSchema,
    OpenAIResponsesToolChoiceNone$inboundSchema,
    OpenAIResponsesToolChoiceRequired$inboundSchema,
]);
/** @internal */
export const OpenAIResponsesToolChoiceUnion$outboundSchema = z.union([
    ToolChoiceAllowed$outboundSchema,
    z.lazy(() => OpenAIResponsesToolChoiceFunction$outboundSchema),
    z.lazy(() => OpenAIResponsesToolChoice$outboundSchema),
    OpenAIResponsesToolChoiceAuto$outboundSchema,
    OpenAIResponsesToolChoiceNone$outboundSchema,
    OpenAIResponsesToolChoiceRequired$outboundSchema,
]);
export function openAIResponsesToolChoiceUnionToJSON(openAIResponsesToolChoiceUnion) {
    return JSON.stringify(OpenAIResponsesToolChoiceUnion$outboundSchema.parse(openAIResponsesToolChoiceUnion));
}
export function openAIResponsesToolChoiceUnionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OpenAIResponsesToolChoiceUnion$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OpenAIResponsesToolChoiceUnion' from JSON`);
}
