// @openrouter/sdk/esm/models/openresponsesresult.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 95391b93fcbf
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { discriminatedUnion } from "../types/discriminatedUnion.js";
import { ApplyPatchServerTool$inboundSchema, } from "./applypatchservertool.js";
import { BaseInputsUnion$inboundSchema, } from "./baseinputsunion.js";
import { BaseReasoningConfig$inboundSchema, } from "./basereasoningconfig.js";
import { CodeInterpreterServerTool$inboundSchema, } from "./codeinterpreterservertool.js";
import { CodexLocalShellTool$inboundSchema, } from "./codexlocalshelltool.js";
import { ComputerUseServerTool$inboundSchema, } from "./computeruseservertool.js";
import { CustomTool$inboundSchema } from "./customtool.js";
import { FileSearchServerTool$inboundSchema, } from "./filesearchservertool.js";
import { ImageGenerationServerTool$inboundSchema, } from "./imagegenerationservertool.js";
import { IncompleteDetails$inboundSchema, } from "./incompletedetails.js";
import { LegacyWebSearchServerTool$inboundSchema, } from "./legacywebsearchservertool.js";
import { McpServerTool$inboundSchema } from "./mcpservertool.js";
import { OpenAIResponsesResponseStatus$inboundSchema, } from "./openairesponsesresponsestatus.js";
import { OpenAIResponsesToolChoiceUnion$inboundSchema, } from "./openairesponsestoolchoiceunion.js";
import { OutputItems$inboundSchema } from "./outputitems.js";
import { Preview20250311WebSearchServerTool$inboundSchema, } from "./preview20250311websearchservertool.js";
import { PreviewWebSearchServerTool$inboundSchema, } from "./previewwebsearchservertool.js";
import { ResponsesErrorField$inboundSchema, } from "./responseserrorfield.js";
import { ShellServerTool$inboundSchema, } from "./shellservertool.js";
import { StoredPromptTemplate$inboundSchema, } from "./storedprompttemplate.js";
import { TextExtendedConfig$inboundSchema, } from "./textextendedconfig.js";
import { Truncation$inboundSchema } from "./truncation.js";
import { Usage$inboundSchema } from "./usage.js";
import { WebSearchServerTool$inboundSchema, } from "./websearchservertool.js";
export const OpenResponsesResultObject = {
    Response: "response",
};
/** @internal */
export const OpenResponsesResultObject$inboundSchema = z.enum(OpenResponsesResultObject);
/** @internal */
export const OpenResponsesResultToolFunction$inboundSchema = z.object({
    description: z.nullable(z.string()).optional(),
    name: z.string(),
    parameters: z.nullable(z.record(z.string(), z.nullable(z.any()))),
    strict: z.nullable(z.boolean()).optional(),
    type: z.literal("function"),
});
export function openResponsesResultToolFunctionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OpenResponsesResultToolFunction$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OpenResponsesResultToolFunction' from JSON`);
}
/** @internal */
export const OpenResponsesResultToolUnion$inboundSchema = discriminatedUnion("type", {
    function: z.lazy(() => OpenResponsesResultToolFunction$inboundSchema),
    web_search_preview: PreviewWebSearchServerTool$inboundSchema,
    web_search_preview_2025_03_11: Preview20250311WebSearchServerTool$inboundSchema,
    web_search: LegacyWebSearchServerTool$inboundSchema,
    web_search_2025_08_26: WebSearchServerTool$inboundSchema,
    file_search: FileSearchServerTool$inboundSchema,
    computer_use_preview: ComputerUseServerTool$inboundSchema,
    code_interpreter: CodeInterpreterServerTool$inboundSchema,
    mcp: McpServerTool$inboundSchema,
    image_generation: ImageGenerationServerTool$inboundSchema,
    local_shell: CodexLocalShellTool$inboundSchema,
    shell: ShellServerTool$inboundSchema,
    apply_patch: ApplyPatchServerTool$inboundSchema,
    custom: CustomTool$inboundSchema,
});
export function openResponsesResultToolUnionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OpenResponsesResultToolUnion$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OpenResponsesResultToolUnion' from JSON`);
}
/** @internal */
export const OpenResponsesResult$inboundSchema = z.object({
    background: z.nullable(z.boolean()).optional(),
    completed_at: z.nullable(z.int()),
    created_at: z.int(),
    error: z.nullable(ResponsesErrorField$inboundSchema),
    frequency_penalty: z.nullable(z.number()),
    id: z.string(),
    incomplete_details: z.nullable(IncompleteDetails$inboundSchema),
    instructions: z.nullable(BaseInputsUnion$inboundSchema),
    max_output_tokens: z.nullable(z.int()).optional(),
    max_tool_calls: z.nullable(z.int()).optional(),
    metadata: z.nullable(z.record(z.string(), z.string())),
    model: z.string(),
    object: OpenResponsesResultObject$inboundSchema,
    output: z.array(OutputItems$inboundSchema),
    output_text: z.string().optional(),
    parallel_tool_calls: z.boolean(),
    presence_penalty: z.nullable(z.number()),
    previous_response_id: z.nullable(z.string()).optional(),
    prompt: z.nullable(StoredPromptTemplate$inboundSchema).optional(),
    prompt_cache_key: z.nullable(z.string()).optional(),
    reasoning: z.nullable(BaseReasoningConfig$inboundSchema).optional(),
    safety_identifier: z.nullable(z.string()).optional(),
    service_tier: z.nullable(z.string()).optional(),
    status: OpenAIResponsesResponseStatus$inboundSchema,
    store: z.boolean().optional(),
    temperature: z.nullable(z.number()),
    text: TextExtendedConfig$inboundSchema.optional(),
    tool_choice: OpenAIResponsesToolChoiceUnion$inboundSchema,
    tools: z.array(discriminatedUnion("type", {
        function: z.lazy(() => OpenResponsesResultToolFunction$inboundSchema),
        web_search_preview: PreviewWebSearchServerTool$inboundSchema,
        web_search_preview_2025_03_11: Preview20250311WebSearchServerTool$inboundSchema,
        web_search: LegacyWebSearchServerTool$inboundSchema,
        web_search_2025_08_26: WebSearchServerTool$inboundSchema,
        file_search: FileSearchServerTool$inboundSchema,
        computer_use_preview: ComputerUseServerTool$inboundSchema,
        code_interpreter: CodeInterpreterServerTool$inboundSchema,
        mcp: McpServerTool$inboundSchema,
        image_generation: ImageGenerationServerTool$inboundSchema,
        local_shell: CodexLocalShellTool$inboundSchema,
        shell: ShellServerTool$inboundSchema,
        apply_patch: ApplyPatchServerTool$inboundSchema,
        custom: CustomTool$inboundSchema,
    })),
    top_logprobs: z.nullable(z.int()).optional(),
    top_p: z.nullable(z.number()),
    truncation: z.nullable(Truncation$inboundSchema).optional(),
    usage: z.nullable(Usage$inboundSchema).optional(),
    user: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        "completed_at": "completedAt",
        "created_at": "createdAt",
        "frequency_penalty": "frequencyPenalty",
        "incomplete_details": "incompleteDetails",
        "max_output_tokens": "maxOutputTokens",
        "max_tool_calls": "maxToolCalls",
        "output_text": "outputText",
        "parallel_tool_calls": "parallelToolCalls",
        "presence_penalty": "presencePenalty",
        "previous_response_id": "previousResponseId",
        "prompt_cache_key": "promptCacheKey",
        "safety_identifier": "safetyIdentifier",
        "service_tier": "serviceTier",
        "tool_choice": "toolChoice",
        "top_logprobs": "topLogprobs",
        "top_p": "topP",
    });
});
export function openResponsesResultFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OpenResponsesResult$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OpenResponsesResult' from JSON`);
}
