// @openrouter/sdk/esm/models/openairesponsefunctiontoolcall.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: d27a63e0fb93
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { ToolCallStatus$inboundSchema, } from "./toolcallstatus.js";
export const OpenAIResponseFunctionToolCallType = {
    FunctionCall: "function_call",
};
/** @internal */
export const OpenAIResponseFunctionToolCallType$inboundSchema = z.enum(OpenAIResponseFunctionToolCallType);
/** @internal */
export const OpenAIResponseFunctionToolCall$inboundSchema = z.object({
    arguments: z.string(),
    call_id: z.string(),
    id: z.string().optional(),
    name: z.string(),
    status: ToolCallStatus$inboundSchema.optional(),
    type: OpenAIResponseFunctionToolCallType$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "call_id": "callId",
    });
});
export function openAIResponseFunctionToolCallFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OpenAIResponseFunctionToolCall$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OpenAIResponseFunctionToolCall' from JSON`);
}
