// @openrouter/sdk/esm/models/openairesponseinputmessageitem.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 67004da2ad5c
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { discriminatedUnion } from "../types/discriminatedUnion.js";
import { InputAudio$inboundSchema } from "./inputaudio.js";
import { InputFile$inboundSchema } from "./inputfile.js";
import { InputImage$inboundSchema } from "./inputimage.js";
import { InputText$inboundSchema } from "./inputtext.js";
export const OpenAIResponseInputMessageItemRoleDeveloper = {
    Developer: "developer",
};
export const OpenAIResponseInputMessageItemRoleSystem = {
    System: "system",
};
export const OpenAIResponseInputMessageItemRoleUser = {
    User: "user",
};
export const OpenAIResponseInputMessageItemType = {
    Message: "message",
};
/** @internal */
export const OpenAIResponseInputMessageItemContent$inboundSchema = discriminatedUnion("type", {
    input_audio: InputAudio$inboundSchema,
    input_file: InputFile$inboundSchema,
    input_image: InputImage$inboundSchema.and(z.object({ type: z.literal("input_image") })),
    input_text: InputText$inboundSchema,
});
export function openAIResponseInputMessageItemContentFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OpenAIResponseInputMessageItemContent$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OpenAIResponseInputMessageItemContent' from JSON`);
}
/** @internal */
export const OpenAIResponseInputMessageItemRoleDeveloper$inboundSchema = z.enum(OpenAIResponseInputMessageItemRoleDeveloper);
/** @internal */
export const OpenAIResponseInputMessageItemRoleSystem$inboundSchema = z.enum(OpenAIResponseInputMessageItemRoleSystem);
/** @internal */
export const OpenAIResponseInputMessageItemRoleUser$inboundSchema = z.enum(OpenAIResponseInputMessageItemRoleUser);
/** @internal */
export const OpenAIResponseInputMessageItemRoleUnion$inboundSchema = z.union([
    OpenAIResponseInputMessageItemRoleUser$inboundSchema,
    OpenAIResponseInputMessageItemRoleSystem$inboundSchema,
    OpenAIResponseInputMessageItemRoleDeveloper$inboundSchema,
]);
export function openAIResponseInputMessageItemRoleUnionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OpenAIResponseInputMessageItemRoleUnion$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OpenAIResponseInputMessageItemRoleUnion' from JSON`);
}
/** @internal */
export const OpenAIResponseInputMessageItemType$inboundSchema = z.enum(OpenAIResponseInputMessageItemType);
/** @internal */
export const OpenAIResponseInputMessageItem$inboundSchema = z.object({
    content: z.array(discriminatedUnion("type", {
        input_audio: InputAudio$inboundSchema,
        input_file: InputFile$inboundSchema,
        input_image: InputImage$inboundSchema.and(z.object({ type: z.literal("input_image") })),
        input_text: InputText$inboundSchema,
    })),
    id: z.string(),
    role: z.union([
        OpenAIResponseInputMessageItemRoleUser$inboundSchema,
        OpenAIResponseInputMessageItemRoleSystem$inboundSchema,
        OpenAIResponseInputMessageItemRoleDeveloper$inboundSchema,
    ]),
    type: OpenAIResponseInputMessageItemType$inboundSchema.optional(),
});
export function openAIResponseInputMessageItemFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OpenAIResponseInputMessageItem$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OpenAIResponseInputMessageItem' from JSON`);
}
