// @openrouter/sdk/esm/models/openairesponsesannotation.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 5828fa2314e3
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { discriminatedUnion } from "../types/discriminatedUnion.js";
import { FileCitation$inboundSchema, FileCitation$outboundSchema, } from "./filecitation.js";
import { FilePath$inboundSchema, FilePath$outboundSchema, } from "./filepath.js";
import { URLCitation$inboundSchema, URLCitation$outboundSchema, } from "./urlcitation.js";
/** @internal */
export const OpenAIResponsesAnnotation$inboundSchema = discriminatedUnion("type", {
    file_citation: FileCitation$inboundSchema,
    url_citation: URLCitation$inboundSchema,
    file_path: FilePath$inboundSchema,
});
/** @internal */
export const OpenAIResponsesAnnotation$outboundSchema = z.union([
    FileCitation$outboundSchema,
    URLCitation$outboundSchema,
    FilePath$outboundSchema,
]);
export function openAIResponsesAnnotationToJSON(openAIResponsesAnnotation) {
    return JSON.stringify(OpenAIResponsesAnnotation$outboundSchema.parse(openAIResponsesAnnotation));
}
export function openAIResponsesAnnotationFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OpenAIResponsesAnnotation$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OpenAIResponsesAnnotation' from JSON`);
}
