// @openrouter/sdk/esm/models/openairesponsefunctiontoolcalloutput.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 78c25d89c63e
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { discriminatedUnion } from "../types/discriminatedUnion.js";
import * as openEnums from "../types/enums.js";
import { InputFile$inboundSchema } from "./inputfile.js";
import { InputImage$inboundSchema } from "./inputimage.js";
import { InputText$inboundSchema } from "./inputtext.js";
export const OpenAIResponseFunctionToolCallOutputStatus = {
    InProgress: "in_progress",
    Completed: "completed",
    Incomplete: "incomplete",
};
export const OpenAIResponseFunctionToolCallOutputType = {
    FunctionCallOutput: "function_call_output",
};
/** @internal */
export const OpenAIResponseFunctionToolCallOutputOutput1$inboundSchema = discriminatedUnion("type", {
    input_file: InputFile$inboundSchema,
    input_image: InputImage$inboundSchema.and(z.object({ type: z.literal("input_image") })),
    input_text: InputText$inboundSchema,
});
export function openAIResponseFunctionToolCallOutputOutput1FromJSON(jsonString) {
    return safeParse(jsonString, (x) => OpenAIResponseFunctionToolCallOutputOutput1$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OpenAIResponseFunctionToolCallOutputOutput1' from JSON`);
}
/** @internal */
export const OpenAIResponseFunctionToolCallOutputOutput2$inboundSchema = z.union([
    z.string(),
    z.array(discriminatedUnion("type", {
        input_file: InputFile$inboundSchema,
        input_image: InputImage$inboundSchema.and(z.object({ type: z.literal("input_image") })),
        input_text: InputText$inboundSchema,
    })),
]);
export function openAIResponseFunctionToolCallOutputOutput2FromJSON(jsonString) {
    return safeParse(jsonString, (x) => OpenAIResponseFunctionToolCallOutputOutput2$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OpenAIResponseFunctionToolCallOutputOutput2' from JSON`);
}
/** @internal */
export const OpenAIResponseFunctionToolCallOutputStatus$inboundSchema = openEnums
    .inboundSchema(OpenAIResponseFunctionToolCallOutputStatus);
/** @internal */
export const OpenAIResponseFunctionToolCallOutputType$inboundSchema = z.enum(OpenAIResponseFunctionToolCallOutputType);
/** @internal */
export const OpenAIResponseFunctionToolCallOutput$inboundSchema = z.object({
    call_id: z.string(),
    id: z.nullable(z.string()).optional(),
    output: z.union([
        z.string(),
        z.array(discriminatedUnion("type", {
            input_file: InputFile$inboundSchema,
            input_image: InputImage$inboundSchema.and(z.object({ type: z.literal("input_image") })),
            input_text: InputText$inboundSchema,
        })),
    ]),
    status: z.nullable(OpenAIResponseFunctionToolCallOutputStatus$inboundSchema)
        .optional(),
    type: OpenAIResponseFunctionToolCallOutputType$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "call_id": "callId",
    });
});
export function openAIResponseFunctionToolCallOutputFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OpenAIResponseFunctionToolCallOutput$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OpenAIResponseFunctionToolCallOutput' from JSON`);
}
