// @openrouter/sdk/esm/models/outputdatetimeitem.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 0c1828e72f98
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { ToolCallStatus$inboundSchema, ToolCallStatus$outboundSchema, } from "./toolcallstatus.js";
export const OutputDatetimeItemType = {
    OpenrouterDatetime: "openrouter:datetime",
};
/** @internal */
export const OutputDatetimeItemType$inboundSchema = z.enum(OutputDatetimeItemType);
/** @internal */
export const OutputDatetimeItemType$outboundSchema = OutputDatetimeItemType$inboundSchema;
/** @internal */
export const OutputDatetimeItem$inboundSchema = z.object({
    datetime: z.string(),
    id: z.string().optional(),
    status: ToolCallStatus$inboundSchema,
    timezone: z.string(),
    type: OutputDatetimeItemType$inboundSchema,
});
/** @internal */
export const OutputDatetimeItem$outboundSchema = z.object({
    datetime: z.string(),
    id: z.string().optional(),
    status: ToolCallStatus$outboundSchema,
    timezone: z.string(),
    type: OutputDatetimeItemType$outboundSchema,
});
export function outputDatetimeItemToJSON(outputDatetimeItem) {
    return JSON.stringify(OutputDatetimeItem$outboundSchema.parse(outputDatetimeItem));
}
export function outputDatetimeItemFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputDatetimeItem$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputDatetimeItem' from JSON`);
}
