// @openrouter/sdk/esm/models/outputmemoryservertoolitem.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 2fb051c8b4e9
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import * as openEnums from "../types/enums.js";
import { ToolCallStatus$inboundSchema, } from "./toolcallstatus.js";
export const ActionEnum = {
    Read: "read",
    Write: "write",
    Delete: "delete",
};
/** @internal */
export const ActionEnum$inboundSchema = openEnums.inboundSchema(ActionEnum);
/** @internal */
export const OutputMemoryServerToolItem$inboundSchema = z.object({
    action: ActionEnum$inboundSchema.optional(),
    id: z.string().optional(),
    key: z.string().optional(),
    status: ToolCallStatus$inboundSchema,
    type: z.literal("openrouter:memory"),
    value: z.nullable(z.any()).optional(),
});
export function outputMemoryServerToolItemFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputMemoryServerToolItem$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputMemoryServerToolItem' from JSON`);
}
