// @openrouter/sdk/esm/models/outputwebsearchservertoolitem.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 92114d9bad95
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { ToolCallStatus$inboundSchema, ToolCallStatus$outboundSchema, } from "./toolcallstatus.js";
export const OutputWebSearchServerToolItemType = {
    OpenrouterWebSearch: "openrouter:web_search",
};
/** @internal */
export const OutputWebSearchServerToolItemType$inboundSchema = z.enum(OutputWebSearchServerToolItemType);
/** @internal */
export const OutputWebSearchServerToolItemType$outboundSchema = OutputWebSearchServerToolItemType$inboundSchema;
/** @internal */
export const OutputWebSearchServerToolItem$inboundSchema = z.object({
    id: z.string().optional(),
    status: ToolCallStatus$inboundSchema,
    type: OutputWebSearchServerToolItemType$inboundSchema,
});
/** @internal */
export const OutputWebSearchServerToolItem$outboundSchema = z.object({
    id: z.string().optional(),
    status: ToolCallStatus$outboundSchema,
    type: OutputWebSearchServerToolItemType$outboundSchema,
});
export function outputWebSearchServerToolItemToJSON(outputWebSearchServerToolItem) {
    return JSON.stringify(OutputWebSearchServerToolItem$outboundSchema.parse(outputWebSearchServerToolItem));
}
export function outputWebSearchServerToolItemFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputWebSearchServerToolItem$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputWebSearchServerToolItem' from JSON`);
}
