// @openrouter/sdk/esm/models/outputtexteditorservertoolitem.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 6c55837a703d
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import * as openEnums from "../types/enums.js";
import { ToolCallStatus$inboundSchema, } from "./toolcallstatus.js";
export const Command = {
    View: "view",
    Create: "create",
    StrReplace: "str_replace",
    Insert: "insert",
};
/** @internal */
export const Command$inboundSchema = openEnums
    .inboundSchema(Command);
/** @internal */
export const OutputTextEditorServerToolItem$inboundSchema = z.object({
    command: Command$inboundSchema.optional(),
    filePath: z.string().optional(),
    id: z.string().optional(),
    status: ToolCallStatus$inboundSchema,
    type: z.literal("openrouter:text_editor"),
});
export function outputTextEditorServerToolItemFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputTextEditorServerToolItem$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputTextEditorServerToolItem' from JSON`);
}
