// @openrouter/sdk/esm/models/outputfunctioncallitem.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 9c462e9ff54b
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
export const OutputFunctionCallItemStatusInProgress = {
    InProgress: "in_progress",
};
export const OutputFunctionCallItemStatusIncomplete = {
    Incomplete: "incomplete",
};
export const OutputFunctionCallItemStatusCompleted = {
    Completed: "completed",
};
export const OutputFunctionCallItemType = {
    FunctionCall: "function_call",
};
/** @internal */
export const OutputFunctionCallItemStatusInProgress$inboundSchema = z.enum(OutputFunctionCallItemStatusInProgress);
/** @internal */
export const OutputFunctionCallItemStatusInProgress$outboundSchema = OutputFunctionCallItemStatusInProgress$inboundSchema;
/** @internal */
export const OutputFunctionCallItemStatusIncomplete$inboundSchema = z.enum(OutputFunctionCallItemStatusIncomplete);
/** @internal */
export const OutputFunctionCallItemStatusIncomplete$outboundSchema = OutputFunctionCallItemStatusIncomplete$inboundSchema;
/** @internal */
export const OutputFunctionCallItemStatusCompleted$inboundSchema = z.enum(OutputFunctionCallItemStatusCompleted);
/** @internal */
export const OutputFunctionCallItemStatusCompleted$outboundSchema = OutputFunctionCallItemStatusCompleted$inboundSchema;
/** @internal */
export const OutputFunctionCallItemStatusUnion$inboundSchema = z.union([
    OutputFunctionCallItemStatusCompleted$inboundSchema,
    OutputFunctionCallItemStatusIncomplete$inboundSchema,
    OutputFunctionCallItemStatusInProgress$inboundSchema,
]);
/** @internal */
export const OutputFunctionCallItemStatusUnion$outboundSchema = z.union([
    OutputFunctionCallItemStatusCompleted$outboundSchema,
    OutputFunctionCallItemStatusIncomplete$outboundSchema,
    OutputFunctionCallItemStatusInProgress$outboundSchema,
]);
export function outputFunctionCallItemStatusUnionToJSON(outputFunctionCallItemStatusUnion) {
    return JSON.stringify(OutputFunctionCallItemStatusUnion$outboundSchema.parse(outputFunctionCallItemStatusUnion));
}
export function outputFunctionCallItemStatusUnionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputFunctionCallItemStatusUnion$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputFunctionCallItemStatusUnion' from JSON`);
}
/** @internal */
export const OutputFunctionCallItemType$inboundSchema = z.enum(OutputFunctionCallItemType);
/** @internal */
export const OutputFunctionCallItemType$outboundSchema = OutputFunctionCallItemType$inboundSchema;
/** @internal */
export const OutputFunctionCallItem$inboundSchema = z.object({
    arguments: z.string(),
    call_id: z.string(),
    id: z.string().optional(),
    name: z.string(),
    status: z.union([
        OutputFunctionCallItemStatusCompleted$inboundSchema,
        OutputFunctionCallItemStatusIncomplete$inboundSchema,
        OutputFunctionCallItemStatusInProgress$inboundSchema,
    ]).optional(),
    type: OutputFunctionCallItemType$inboundSchema,
}).transform((v) => {
    return remap$(v, {
        "call_id": "callId",
    });
});
/** @internal */
export const OutputFunctionCallItem$outboundSchema = z.object({
    arguments: z.string(),
    callId: z.string(),
    id: z.string().optional(),
    name: z.string(),
    status: z.union([
        OutputFunctionCallItemStatusCompleted$outboundSchema,
        OutputFunctionCallItemStatusIncomplete$outboundSchema,
        OutputFunctionCallItemStatusInProgress$outboundSchema,
    ]).optional(),
    type: OutputFunctionCallItemType$outboundSchema,
}).transform((v) => {
    return remap$(v, {
        callId: "call_id",
    });
});
export function outputFunctionCallItemToJSON(outputFunctionCallItem) {
    return JSON.stringify(OutputFunctionCallItem$outboundSchema.parse(outputFunctionCallItem));
}
export function outputFunctionCallItemFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputFunctionCallItem$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputFunctionCallItem' from JSON`);
}
