// @openrouter/sdk/esm/models/outputwebsearchcallitem.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: b4585d176186
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { discriminatedUnion } from "../types/discriminatedUnion.js";
import { WebSearchSource$inboundSchema, WebSearchSource$outboundSchema, } from "./websearchsource.js";
import { WebSearchStatus$inboundSchema, WebSearchStatus$outboundSchema, } from "./websearchstatus.js";
export const TypeWebSearchCall = {
    WebSearchCall: "web_search_call",
};
/** @internal */
export const ActionFindInPage$inboundSchema = z.object({
    pattern: z.string(),
    type: z.literal("find_in_page"),
    url: z.string(),
});
/** @internal */
export const ActionFindInPage$outboundSchema = z.object({
    pattern: z.string(),
    type: z.literal("find_in_page"),
    url: z.string(),
});
export function actionFindInPageToJSON(actionFindInPage) {
    return JSON.stringify(ActionFindInPage$outboundSchema.parse(actionFindInPage));
}
export function actionFindInPageFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ActionFindInPage$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ActionFindInPage' from JSON`);
}
/** @internal */
export const ActionOpenPage$inboundSchema = z.object({
    type: z.literal("open_page"),
    url: z.nullable(z.string()).optional(),
});
/** @internal */
export const ActionOpenPage$outboundSchema = z.object({
    type: z.literal("open_page"),
    url: z.nullable(z.string()).optional(),
});
export function actionOpenPageToJSON(actionOpenPage) {
    return JSON.stringify(ActionOpenPage$outboundSchema.parse(actionOpenPage));
}
export function actionOpenPageFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ActionOpenPage$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ActionOpenPage' from JSON`);
}
/** @internal */
export const ActionSearch$inboundSchema = z
    .object({
    queries: z.array(z.string()).optional(),
    query: z.string(),
    sources: z.array(WebSearchSource$inboundSchema).optional(),
    type: z.literal("search"),
});
/** @internal */
export const ActionSearch$outboundSchema = z.object({
    queries: z.array(z.string()).optional(),
    query: z.string(),
    sources: z.array(WebSearchSource$outboundSchema).optional(),
    type: z.literal("search"),
});
export function actionSearchToJSON(actionSearch) {
    return JSON.stringify(ActionSearch$outboundSchema.parse(actionSearch));
}
export function actionSearchFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ActionSearch$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ActionSearch' from JSON`);
}
/** @internal */
export const Action$inboundSchema = discriminatedUnion("type", {
    search: z.lazy(() => ActionSearch$inboundSchema),
    open_page: z.lazy(() => ActionOpenPage$inboundSchema),
    find_in_page: z.lazy(() => ActionFindInPage$inboundSchema),
});
/** @internal */
export const Action$outboundSchema = z
    .union([
    z.lazy(() => ActionSearch$outboundSchema),
    z.lazy(() => ActionOpenPage$outboundSchema),
    z.lazy(() => ActionFindInPage$outboundSchema),
]);
export function actionToJSON(action) {
    return JSON.stringify(Action$outboundSchema.parse(action));
}
export function actionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Action$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Action' from JSON`);
}
/** @internal */
export const TypeWebSearchCall$inboundSchema = z.enum(TypeWebSearchCall);
/** @internal */
export const TypeWebSearchCall$outboundSchema = TypeWebSearchCall$inboundSchema;
/** @internal */
export const OutputWebSearchCallItem$inboundSchema = z.object({
    action: discriminatedUnion("type", {
        search: z.lazy(() => ActionSearch$inboundSchema),
        open_page: z.lazy(() => ActionOpenPage$inboundSchema),
        find_in_page: z.lazy(() => ActionFindInPage$inboundSchema),
    }),
    id: z.string(),
    status: WebSearchStatus$inboundSchema,
    type: TypeWebSearchCall$inboundSchema,
});
/** @internal */
export const OutputWebSearchCallItem$outboundSchema = z.object({
    action: z.union([
        z.lazy(() => ActionSearch$outboundSchema),
        z.lazy(() => ActionOpenPage$outboundSchema),
        z.lazy(() => ActionFindInPage$outboundSchema),
    ]),
    id: z.string(),
    status: WebSearchStatus$outboundSchema,
    type: TypeWebSearchCall$outboundSchema,
});
export function outputWebSearchCallItemToJSON(outputWebSearchCallItem) {
    return JSON.stringify(OutputWebSearchCallItem$outboundSchema.parse(outputWebSearchCallItem));
}
export function outputWebSearchCallItemFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputWebSearchCallItem$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputWebSearchCallItem' from JSON`);
}
