// @openrouter/sdk/esm/models/outputcomputercallitem.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: fd4439fdf358
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import * as openEnums from "../types/enums.js";
export const OutputComputerCallItemStatus = {
    Completed: "completed",
    Incomplete: "incomplete",
    InProgress: "in_progress",
};
/** @internal */
export const PendingSafetyCheck$inboundSchema = z.object({
    code: z.string(),
    id: z.string(),
    message: z.string(),
});
export function pendingSafetyCheckFromJSON(jsonString) {
    return safeParse(jsonString, (x) => PendingSafetyCheck$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'PendingSafetyCheck' from JSON`);
}
/** @internal */
export const OutputComputerCallItemStatus$inboundSchema = openEnums.inboundSchema(OutputComputerCallItemStatus);
/** @internal */
export const OutputComputerCallItem$inboundSchema = z.object({
    action: z.nullable(z.any()).optional(),
    call_id: z.string(),
    id: z.string().optional(),
    pending_safety_checks: z.array(z.lazy(() => PendingSafetyCheck$inboundSchema)),
    status: OutputComputerCallItemStatus$inboundSchema,
    type: z.literal("computer_call"),
}).transform((v) => {
    return remap$(v, {
        "call_id": "callId",
        "pending_safety_checks": "pendingSafetyChecks",
    });
});
export function outputComputerCallItemFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputComputerCallItem$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputComputerCallItem' from JSON`);
}
