// @openrouter/sdk/esm/models/outputreasoningitem.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 4ea30f8ad34c
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { ReasoningFormat$inboundSchema, } from "./reasoningformat.js";
import { ReasoningSummaryText$inboundSchema, } from "./reasoningsummarytext.js";
import { ReasoningTextContent$inboundSchema, } from "./reasoningtextcontent.js";
export const OutputReasoningItemStatusInProgress = {
    InProgress: "in_progress",
};
export const OutputReasoningItemStatusIncomplete = {
    Incomplete: "incomplete",
};
export const OutputReasoningItemStatusCompleted = {
    Completed: "completed",
};
/** @internal */
export const OutputReasoningItemStatusInProgress$inboundSchema = z.enum(OutputReasoningItemStatusInProgress);
/** @internal */
export const OutputReasoningItemStatusIncomplete$inboundSchema = z.enum(OutputReasoningItemStatusIncomplete);
/** @internal */
export const OutputReasoningItemStatusCompleted$inboundSchema = z.enum(OutputReasoningItemStatusCompleted);
/** @internal */
export const OutputReasoningItemStatusUnion$inboundSchema = z.union([
    OutputReasoningItemStatusCompleted$inboundSchema,
    OutputReasoningItemStatusIncomplete$inboundSchema,
    OutputReasoningItemStatusInProgress$inboundSchema,
]);
export function outputReasoningItemStatusUnionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputReasoningItemStatusUnion$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputReasoningItemStatusUnion' from JSON`);
}
/** @internal */
export const OutputReasoningItem$inboundSchema = z.object({
    content: z.nullable(z.array(ReasoningTextContent$inboundSchema)).optional(),
    encrypted_content: z.nullable(z.string()).optional(),
    id: z.string(),
    status: z.union([
        OutputReasoningItemStatusCompleted$inboundSchema,
        OutputReasoningItemStatusIncomplete$inboundSchema,
        OutputReasoningItemStatusInProgress$inboundSchema,
    ]).optional(),
    summary: z.array(ReasoningSummaryText$inboundSchema),
    type: z.literal("reasoning"),
    format: z.nullable(ReasoningFormat$inboundSchema).optional(),
    signature: z.nullable(z.string()).optional(),
}).transform((v) => {
    return remap$(v, {
        "encrypted_content": "encryptedContent",
    });
});
export function outputReasoningItemFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputReasoningItem$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputReasoningItem' from JSON`);
}
