// @openrouter/sdk/esm/models/outputcodeinterpretercallitem.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: ee080e2eccae
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { discriminatedUnion } from "../types/discriminatedUnion.js";
import { ToolCallStatus$inboundSchema, } from "./toolcallstatus.js";
/** @internal */
export const OutputLogs$inboundSchema = z
    .object({
    logs: z.string(),
    type: z.literal("logs"),
});
export function outputLogsFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputLogs$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputLogs' from JSON`);
}
/** @internal */
export const OutputImage$inboundSchema = z
    .object({
    type: z.literal("image"),
    url: z.string(),
});
export function outputImageFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputImage$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputImage' from JSON`);
}
/** @internal */
export const OutputCodeInterpreterCallItemOutputUnion$inboundSchema = discriminatedUnion("type", {
    image: z.lazy(() => OutputImage$inboundSchema),
    logs: z.lazy(() => OutputLogs$inboundSchema),
});
export function outputCodeInterpreterCallItemOutputUnionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputCodeInterpreterCallItemOutputUnion$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputCodeInterpreterCallItemOutputUnion' from JSON`);
}
/** @internal */
export const OutputCodeInterpreterCallItem$inboundSchema = z.object({
    code: z.nullable(z.string()),
    container_id: z.string(),
    id: z.string(),
    outputs: z.nullable(z.array(discriminatedUnion("type", {
        image: z.lazy(() => OutputImage$inboundSchema),
        logs: z.lazy(() => OutputLogs$inboundSchema),
    }))),
    status: ToolCallStatus$inboundSchema,
    type: z.literal("code_interpreter_call"),
}).transform((v) => {
    return remap$(v, {
        "container_id": "containerId",
    });
});
export function outputCodeInterpreterCallItemFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputCodeInterpreterCallItem$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputCodeInterpreterCallItem' from JSON`);
}
