// @openrouter/sdk/esm/models/outputitemimagegenerationcall.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 9c3ac6a7a9ad
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { ImageGenerationStatus$inboundSchema, } from "./imagegenerationstatus.js";
export const OutputItemImageGenerationCallType = {
    ImageGenerationCall: "image_generation_call",
};
/** @internal */
export const OutputItemImageGenerationCallType$inboundSchema = z.enum(OutputItemImageGenerationCallType);
/** @internal */
export const OutputItemImageGenerationCall$inboundSchema = z.object({
    id: z.string(),
    result: z.nullable(z.string()).default(null),
    status: ImageGenerationStatus$inboundSchema,
    type: OutputItemImageGenerationCallType$inboundSchema,
});
export function outputItemImageGenerationCallFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputItemImageGenerationCall$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputItemImageGenerationCall' from JSON`);
}
