// @openrouter/sdk/esm/models/outputmessage.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: fa128cf87d51
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { discriminatedUnion } from "../types/discriminatedUnion.js";
import { OpenAIResponsesRefusalContent$inboundSchema, } from "./openairesponsesrefusalcontent.js";
import { ResponseOutputText$inboundSchema, } from "./responseoutputtext.js";
export const OutputMessagePhaseFinalAnswer = {
    FinalAnswer: "final_answer",
};
export const OutputMessagePhaseCommentary = {
    Commentary: "commentary",
};
export const OutputMessageRole = {
    Assistant: "assistant",
};
export const OutputMessageStatusInProgress = {
    InProgress: "in_progress",
};
export const OutputMessageStatusIncomplete = {
    Incomplete: "incomplete",
};
export const OutputMessageStatusCompleted = {
    Completed: "completed",
};
export const OutputMessageType = {
    Message: "message",
};
/** @internal */
export const OutputMessageContent$inboundSchema = discriminatedUnion("type", {
    output_text: ResponseOutputText$inboundSchema,
    refusal: OpenAIResponsesRefusalContent$inboundSchema,
});
export function outputMessageContentFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputMessageContent$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputMessageContent' from JSON`);
}
/** @internal */
export const OutputMessagePhaseFinalAnswer$inboundSchema = z.enum(OutputMessagePhaseFinalAnswer);
/** @internal */
export const OutputMessagePhaseCommentary$inboundSchema = z.enum(OutputMessagePhaseCommentary);
/** @internal */
export const OutputMessagePhaseUnion$inboundSchema = z.union([
    OutputMessagePhaseCommentary$inboundSchema,
    OutputMessagePhaseFinalAnswer$inboundSchema,
    z.any(),
]);
export function outputMessagePhaseUnionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputMessagePhaseUnion$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputMessagePhaseUnion' from JSON`);
}
/** @internal */
export const OutputMessageRole$inboundSchema = z.enum(OutputMessageRole);
/** @internal */
export const OutputMessageStatusInProgress$inboundSchema = z.enum(OutputMessageStatusInProgress);
/** @internal */
export const OutputMessageStatusIncomplete$inboundSchema = z.enum(OutputMessageStatusIncomplete);
/** @internal */
export const OutputMessageStatusCompleted$inboundSchema = z.enum(OutputMessageStatusCompleted);
/** @internal */
export const OutputMessageStatusUnion$inboundSchema = z.union([
    OutputMessageStatusCompleted$inboundSchema,
    OutputMessageStatusIncomplete$inboundSchema,
    OutputMessageStatusInProgress$inboundSchema,
]);
export function outputMessageStatusUnionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputMessageStatusUnion$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputMessageStatusUnion' from JSON`);
}
/** @internal */
export const OutputMessageType$inboundSchema = z.enum(OutputMessageType);
/** @internal */
export const OutputMessage$inboundSchema = z
    .object({
    content: z.array(discriminatedUnion("type", {
        output_text: ResponseOutputText$inboundSchema,
        refusal: OpenAIResponsesRefusalContent$inboundSchema,
    })),
    id: z.string(),
    phase: z.nullable(z.union([
        OutputMessagePhaseCommentary$inboundSchema,
        OutputMessagePhaseFinalAnswer$inboundSchema,
        z.any(),
    ])).optional(),
    role: OutputMessageRole$inboundSchema,
    status: z.union([
        OutputMessageStatusCompleted$inboundSchema,
        OutputMessageStatusIncomplete$inboundSchema,
        OutputMessageStatusInProgress$inboundSchema,
    ]).optional(),
    type: OutputMessageType$inboundSchema,
});
export function outputMessageFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputMessage$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputMessage' from JSON`);
}
