// @openrouter/sdk/esm/models/outputfilesearchcallitem.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: eafceec94b81
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { WebSearchStatus$inboundSchema, WebSearchStatus$outboundSchema, } from "./websearchstatus.js";
export const OutputFileSearchCallItemType = {
    FileSearchCall: "file_search_call",
};
/** @internal */
export const OutputFileSearchCallItemType$inboundSchema = z.enum(OutputFileSearchCallItemType);
/** @internal */
export const OutputFileSearchCallItemType$outboundSchema = OutputFileSearchCallItemType$inboundSchema;
/** @internal */
export const OutputFileSearchCallItem$inboundSchema = z.object({
    id: z.string(),
    queries: z.array(z.string()),
    status: WebSearchStatus$inboundSchema,
    type: OutputFileSearchCallItemType$inboundSchema,
});
/** @internal */
export const OutputFileSearchCallItem$outboundSchema = z.object({
    id: z.string(),
    queries: z.array(z.string()),
    status: WebSearchStatus$outboundSchema,
    type: OutputFileSearchCallItemType$outboundSchema,
});
export function outputFileSearchCallItemToJSON(outputFileSearchCallItem) {
    return JSON.stringify(OutputFileSearchCallItem$outboundSchema.parse(outputFileSearchCallItem));
}
export function outputFileSearchCallItemFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputFileSearchCallItem$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputFileSearchCallItem' from JSON`);
}
