// @openrouter/sdk/esm/models/outputitems.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 0d7161b3403e
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { discriminatedUnion } from "../types/discriminatedUnion.js";
import { OutputApplyPatchServerToolItem$inboundSchema, } from "./outputapplypatchservertoolitem.js";
import { OutputBashServerToolItem$inboundSchema, } from "./outputbashservertoolitem.js";
import { OutputBrowserUseServerToolItem$inboundSchema, } from "./outputbrowseruseservertoolitem.js";
import { OutputCodeInterpreterCallItem$inboundSchema, } from "./outputcodeinterpretercallitem.js";
import { OutputCodeInterpreterServerToolItem$inboundSchema, } from "./outputcodeinterpreterservertoolitem.js";
import { OutputComputerCallItem$inboundSchema, } from "./outputcomputercallitem.js";
import { OutputDatetimeItem$inboundSchema, } from "./outputdatetimeitem.js";
import { OutputFileSearchCallItem$inboundSchema, } from "./outputfilesearchcallitem.js";
import { OutputFileSearchServerToolItem$inboundSchema, } from "./outputfilesearchservertoolitem.js";
import { OutputFunctionCallItem$inboundSchema, } from "./outputfunctioncallitem.js";
import { OutputImageGenerationCallItem$inboundSchema, } from "./outputimagegenerationcallitem.js";
import { OutputImageGenerationServerToolItem$inboundSchema, } from "./outputimagegenerationservertoolitem.js";
import { OutputMcpServerToolItem$inboundSchema, } from "./outputmcpservertoolitem.js";
import { OutputMemoryServerToolItem$inboundSchema, } from "./outputmemoryservertoolitem.js";
import { OutputMessageItem$inboundSchema, } from "./outputmessageitem.js";
import { OutputReasoningItem$inboundSchema, } from "./outputreasoningitem.js";
import { OutputSearchModelsServerToolItem$inboundSchema, } from "./outputsearchmodelsservertoolitem.js";
import { OutputTextEditorServerToolItem$inboundSchema, } from "./outputtexteditorservertoolitem.js";
import { OutputToolSearchServerToolItem$inboundSchema, } from "./outputtoolsearchservertoolitem.js";
import { OutputWebFetchServerToolItem$inboundSchema, } from "./outputwebfetchservertoolitem.js";
import { OutputWebSearchCallItem$inboundSchema, } from "./outputwebsearchcallitem.js";
import { OutputWebSearchServerToolItem$inboundSchema, } from "./outputwebsearchservertoolitem.js";
/** @internal */
export const OutputItems$inboundSchema = discriminatedUnion("type", {
    code_interpreter_call: OutputCodeInterpreterCallItem$inboundSchema,
    computer_call: OutputComputerCallItem$inboundSchema,
    file_search_call: OutputFileSearchCallItem$inboundSchema.and(z.object({ type: z.literal("file_search_call") })),
    function_call: OutputFunctionCallItem$inboundSchema.and(z.object({ type: z.literal("function_call") })),
    image_generation_call: OutputImageGenerationCallItem$inboundSchema.and(z.object({ type: z.literal("image_generation_call") })),
    message: OutputMessageItem$inboundSchema,
    ["openrouter:apply_patch"]: OutputApplyPatchServerToolItem$inboundSchema,
    ["openrouter:bash"]: OutputBashServerToolItem$inboundSchema,
    ["openrouter:browser_use"]: OutputBrowserUseServerToolItem$inboundSchema,
    ["openrouter:code_interpreter"]: OutputCodeInterpreterServerToolItem$inboundSchema,
    ["openrouter:datetime"]: OutputDatetimeItem$inboundSchema.and(z.object({ type: z.literal("openrouter:datetime") })),
    ["openrouter:experimental__search_models"]: OutputSearchModelsServerToolItem$inboundSchema,
    ["openrouter:file_search"]: OutputFileSearchServerToolItem$inboundSchema,
    ["openrouter:image_generation"]: OutputImageGenerationServerToolItem$inboundSchema,
    ["openrouter:mcp"]: OutputMcpServerToolItem$inboundSchema,
    ["openrouter:memory"]: OutputMemoryServerToolItem$inboundSchema,
    ["openrouter:text_editor"]: OutputTextEditorServerToolItem$inboundSchema,
    ["openrouter:tool_search"]: OutputToolSearchServerToolItem$inboundSchema,
    ["openrouter:web_fetch"]: OutputWebFetchServerToolItem$inboundSchema,
    ["openrouter:web_search"]: OutputWebSearchServerToolItem$inboundSchema.and(z.object({ type: z.literal("openrouter:web_search") })),
    reasoning: OutputReasoningItem$inboundSchema,
    web_search_call: OutputWebSearchCallItem$inboundSchema.and(z.object({ type: z.literal("web_search_call") })),
});
export function outputItemsFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputItems$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputItems' from JSON`);
}
