// @openrouter/sdk/esm/models/outputmessageitem.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 7eb39195329c
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { discriminatedUnion } from "../types/discriminatedUnion.js";
import { OpenAIResponsesRefusalContent$inboundSchema, } from "./openairesponsesrefusalcontent.js";
import { ResponseOutputText$inboundSchema, } from "./responseoutputtext.js";
export const OutputMessageItemPhaseFinalAnswer = {
    FinalAnswer: "final_answer",
};
export const OutputMessageItemPhaseCommentary = {
    Commentary: "commentary",
};
export const OutputMessageItemRole = {
    Assistant: "assistant",
};
export const OutputMessageItemStatusInProgress = {
    InProgress: "in_progress",
};
export const OutputMessageItemStatusIncomplete = {
    Incomplete: "incomplete",
};
export const OutputMessageItemStatusCompleted = {
    Completed: "completed",
};
/** @internal */
export const OutputMessageItemContent$inboundSchema = discriminatedUnion("type", {
    output_text: ResponseOutputText$inboundSchema,
    refusal: OpenAIResponsesRefusalContent$inboundSchema,
});
export function outputMessageItemContentFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputMessageItemContent$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputMessageItemContent' from JSON`);
}
/** @internal */
export const OutputMessageItemPhaseFinalAnswer$inboundSchema = z.enum(OutputMessageItemPhaseFinalAnswer);
/** @internal */
export const OutputMessageItemPhaseCommentary$inboundSchema = z.enum(OutputMessageItemPhaseCommentary);
/** @internal */
export const OutputMessageItemPhaseUnion$inboundSchema = z.union([
    OutputMessageItemPhaseCommentary$inboundSchema,
    OutputMessageItemPhaseFinalAnswer$inboundSchema,
    z.any(),
]);
export function outputMessageItemPhaseUnionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputMessageItemPhaseUnion$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputMessageItemPhaseUnion' from JSON`);
}
/** @internal */
export const OutputMessageItemRole$inboundSchema = z.enum(OutputMessageItemRole);
/** @internal */
export const OutputMessageItemStatusInProgress$inboundSchema = z.enum(OutputMessageItemStatusInProgress);
/** @internal */
export const OutputMessageItemStatusIncomplete$inboundSchema = z.enum(OutputMessageItemStatusIncomplete);
/** @internal */
export const OutputMessageItemStatusCompleted$inboundSchema = z.enum(OutputMessageItemStatusCompleted);
/** @internal */
export const OutputMessageItemStatusUnion$inboundSchema = z.union([
    OutputMessageItemStatusCompleted$inboundSchema,
    OutputMessageItemStatusIncomplete$inboundSchema,
    OutputMessageItemStatusInProgress$inboundSchema,
]);
export function outputMessageItemStatusUnionFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputMessageItemStatusUnion$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputMessageItemStatusUnion' from JSON`);
}
/** @internal */
export const OutputMessageItem$inboundSchema = z.object({
    content: z.array(discriminatedUnion("type", {
        output_text: ResponseOutputText$inboundSchema,
        refusal: OpenAIResponsesRefusalContent$inboundSchema,
    })),
    id: z.string(),
    phase: z.nullable(z.union([
        OutputMessageItemPhaseCommentary$inboundSchema,
        OutputMessageItemPhaseFinalAnswer$inboundSchema,
        z.any(),
    ])).optional(),
    role: OutputMessageItemRole$inboundSchema,
    status: z.union([
        OutputMessageItemStatusCompleted$inboundSchema,
        OutputMessageItemStatusIncomplete$inboundSchema,
        OutputMessageItemStatusInProgress$inboundSchema,
    ]).optional(),
    type: z.literal("message"),
});
export function outputMessageItemFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputMessageItem$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputMessageItem' from JSON`);
}
