// @openrouter/sdk/esm/models/outputimagegenerationcallitem.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 618bac10713d
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { ImageGenerationStatus$inboundSchema, ImageGenerationStatus$outboundSchema, } from "./imagegenerationstatus.js";
export const OutputImageGenerationCallItemType = {
    ImageGenerationCall: "image_generation_call",
};
/** @internal */
export const OutputImageGenerationCallItemType$inboundSchema = z.enum(OutputImageGenerationCallItemType);
/** @internal */
export const OutputImageGenerationCallItemType$outboundSchema = OutputImageGenerationCallItemType$inboundSchema;
/** @internal */
export const OutputImageGenerationCallItem$inboundSchema = z.object({
    id: z.string(),
    result: z.nullable(z.string()).default(null),
    status: ImageGenerationStatus$inboundSchema,
    type: OutputImageGenerationCallItemType$inboundSchema,
});
/** @internal */
export const OutputImageGenerationCallItem$outboundSchema = z.object({
    id: z.string(),
    result: z.nullable(z.string()).default(null),
    status: ImageGenerationStatus$outboundSchema,
    type: OutputImageGenerationCallItemType$outboundSchema,
});
export function outputImageGenerationCallItemToJSON(outputImageGenerationCallItem) {
    return JSON.stringify(OutputImageGenerationCallItem$outboundSchema.parse(outputImageGenerationCallItem));
}
export function outputImageGenerationCallItemFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputImageGenerationCallItem$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputImageGenerationCallItem' from JSON`);
}
