// @openrouter/sdk/esm/models/parameter.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 066287ded212
 */
import * as openEnums from "../types/enums.js";
export const Parameter = {
    Temperature: "temperature",
    TopP: "top_p",
    TopK: "top_k",
    MinP: "min_p",
    TopA: "top_a",
    FrequencyPenalty: "frequency_penalty",
    PresencePenalty: "presence_penalty",
    RepetitionPenalty: "repetition_penalty",
    MaxTokens: "max_tokens",
    MaxCompletionTokens: "max_completion_tokens",
    LogitBias: "logit_bias",
    Logprobs: "logprobs",
    TopLogprobs: "top_logprobs",
    Seed: "seed",
    ResponseFormat: "response_format",
    StructuredOutputs: "structured_outputs",
    Stop: "stop",
    Tools: "tools",
    ToolChoice: "tool_choice",
    ParallelToolCalls: "parallel_tool_calls",
    IncludeReasoning: "include_reasoning",
    Reasoning: "reasoning",
    ReasoningEffort: "reasoning_effort",
    WebSearchOptions: "web_search_options",
    Verbosity: "verbosity",
};
/** @internal */
export const Parameter$inboundSchema = openEnums
    .inboundSchema(Parameter);
