// @openrouter/sdk/esm/models/previewwebsearchuserlocation.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 64db0edfc37d
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
export const PreviewWebSearchUserLocationType = {
    Approximate: "approximate",
};
/** @internal */
export const PreviewWebSearchUserLocationType$inboundSchema = z.enum(PreviewWebSearchUserLocationType);
/** @internal */
export const PreviewWebSearchUserLocationType$outboundSchema = PreviewWebSearchUserLocationType$inboundSchema;
/** @internal */
export const PreviewWebSearchUserLocation$inboundSchema = z.object({
    city: z.nullable(z.string()).optional(),
    country: z.nullable(z.string()).optional(),
    region: z.nullable(z.string()).optional(),
    timezone: z.nullable(z.string()).optional(),
    type: PreviewWebSearchUserLocationType$inboundSchema,
});
/** @internal */
export const PreviewWebSearchUserLocation$outboundSchema = z.object({
    city: z.nullable(z.string()).optional(),
    country: z.nullable(z.string()).optional(),
    region: z.nullable(z.string()).optional(),
    timezone: z.nullable(z.string()).optional(),
    type: PreviewWebSearchUserLocationType$outboundSchema,
});
export function previewWebSearchUserLocationToJSON(previewWebSearchUserLocation) {
    return JSON.stringify(PreviewWebSearchUserLocation$outboundSchema.parse(previewWebSearchUserLocation));
}
export function previewWebSearchUserLocationFromJSON(jsonString) {
    return safeParse(jsonString, (x) => PreviewWebSearchUserLocation$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'PreviewWebSearchUserLocation' from JSON`);
}
