// @openrouter/sdk/esm/models/providersortconfig.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 1491329aae16
 */
import * as z from "zod/v4";
import * as openEnums from "../types/enums.js";
/**
 * The provider sorting strategy (price, throughput, latency)
 */
export const By = {
    Price: "price",
    Throughput: "throughput",
    Latency: "latency",
    Exacto: "exacto",
};
/**
 * Partitioning strategy for sorting: "model" (default) groups endpoints by model before sorting (fallback models remain fallbacks), "none" sorts all endpoints together regardless of model.
 */
export const Partition = {
    Model: "model",
    None: "none",
};
/** @internal */
export const By$outboundSchema = openEnums
    .outboundSchema(By);
/** @internal */
export const Partition$outboundSchema = openEnums
    .outboundSchema(Partition);
/** @internal */
export const ProviderSortConfig$outboundSchema = z.object({
    by: z.nullable(By$outboundSchema).optional(),
    partition: z.nullable(Partition$outboundSchema).optional(),
});
export function providerSortConfigToJSON(providerSortConfig) {
    return JSON.stringify(ProviderSortConfig$outboundSchema.parse(providerSortConfig));
}
