// @openrouter/sdk/esm/models/providerpreferences.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 3a47295d7b91
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import * as openEnums from "../types/enums.js";
import { PreferredMaxLatency$outboundSchema, } from "./preferredmaxlatency.js";
import { PreferredMinThroughput$outboundSchema, } from "./preferredminthroughput.js";
import { ProviderName$outboundSchema } from "./providername.js";
import { ProviderSort$outboundSchema } from "./providersort.js";
import { ProviderSortConfig$outboundSchema, } from "./providersortconfig.js";
import { Quantization$outboundSchema } from "./quantization.js";
/**
 * Data collection setting. If no available model provider meets the requirement, your request will return an error.
 *
 * @remarks
 * - allow: (default) allow providers which store user data non-transiently and may train on it
 *
 * - deny: use only providers which do not collect user data.
 */
export const DataCollection = {
    Deny: "deny",
    Allow: "allow",
};
/** @internal */
export const DataCollection$outboundSchema = openEnums.outboundSchema(DataCollection);
/** @internal */
export const Ignore$outboundSchema = z
    .union([ProviderName$outboundSchema, z.string()]);
export function ignoreToJSON(ignore) {
    return JSON.stringify(Ignore$outboundSchema.parse(ignore));
}
/** @internal */
export const MaxPrice$outboundSchema = z
    .object({
    audio: z.string().optional(),
    completion: z.string().optional(),
    image: z.string().optional(),
    prompt: z.string().optional(),
    request: z.string().optional(),
});
export function maxPriceToJSON(maxPrice) {
    return JSON.stringify(MaxPrice$outboundSchema.parse(maxPrice));
}
/** @internal */
export const Only$outboundSchema = z.union([
    ProviderName$outboundSchema,
    z.string(),
]);
export function onlyToJSON(only) {
    return JSON.stringify(Only$outboundSchema.parse(only));
}
/** @internal */
export const Order$outboundSchema = z.union([
    ProviderName$outboundSchema,
    z.string(),
]);
export function orderToJSON(order) {
    return JSON.stringify(Order$outboundSchema.parse(order));
}
/** @internal */
export const Sort$outboundSchema = z.union([
    ProviderSort$outboundSchema,
    ProviderSortConfig$outboundSchema,
    z.any(),
]);
export function sortToJSON(sort) {
    return JSON.stringify(Sort$outboundSchema.parse(sort));
}
/** @internal */
export const ProviderPreferences$outboundSchema = z.object({
    allowFallbacks: z.nullable(z.boolean()).optional(),
    dataCollection: z.nullable(DataCollection$outboundSchema).optional(),
    enforceDistillableText: z.nullable(z.boolean()).optional(),
    ignore: z.nullable(z.array(z.union([ProviderName$outboundSchema, z.string()]))).optional(),
    maxPrice: z.lazy(() => MaxPrice$outboundSchema).optional(),
    only: z.nullable(z.array(z.union([ProviderName$outboundSchema, z.string()])))
        .optional(),
    order: z.nullable(z.array(z.union([ProviderName$outboundSchema, z.string()])))
        .optional(),
    preferredMaxLatency: z.nullable(PreferredMaxLatency$outboundSchema)
        .optional(),
    preferredMinThroughput: z.nullable(PreferredMinThroughput$outboundSchema)
        .optional(),
    quantizations: z.nullable(z.array(Quantization$outboundSchema)).optional(),
    requireParameters: z.nullable(z.boolean()).optional(),
    sort: z.nullable(z.union([
        ProviderSort$outboundSchema,
        ProviderSortConfig$outboundSchema,
        z.any(),
    ])).optional(),
    zdr: z.nullable(z.boolean()).optional(),
}).transform((v) => {
    return remap$(v, {
        allowFallbacks: "allow_fallbacks",
        dataCollection: "data_collection",
        enforceDistillableText: "enforce_distillable_text",
        maxPrice: "max_price",
        preferredMaxLatency: "preferred_max_latency",
        preferredMinThroughput: "preferred_min_throughput",
        requireParameters: "require_parameters",
    });
});
export function providerPreferencesToJSON(providerPreferences) {
    return JSON.stringify(ProviderPreferences$outboundSchema.parse(providerPreferences));
}
