// @openrouter/sdk/esm/models/providerresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 7427d17a27bc
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import * as openEnums from "../types/enums.js";
/**
 * Name of the provider
 */
export const ProviderResponseProviderName = {
    AnyScale: "AnyScale",
    Atoma: "Atoma",
    CentML: "Cent-ML",
    CrofAI: "CrofAI",
    Enfer: "Enfer",
    GoPomelo: "GoPomelo",
    HuggingFace: "HuggingFace",
    Hyperbolic2: "Hyperbolic 2",
    InoCloud: "InoCloud",
    Kluster: "Kluster",
    Lambda: "Lambda",
    Lepton: "Lepton",
    Lynn2: "Lynn 2",
    Lynn: "Lynn",
    Mancer: "Mancer",
    Meta: "Meta",
    Modal: "Modal",
    Nineteen: "Nineteen",
    OctoAI: "OctoAI",
    Recursal: "Recursal",
    Reflection: "Reflection",
    Replicate: "Replicate",
    SambaNova2: "SambaNova 2",
    SFCompute: "SF Compute",
    Targon: "Targon",
    Together2: "Together 2",
    Ubicloud: "Ubicloud",
    OneDotAI: "01.AI",
    AkashML: "AkashML",
    Ai21: "AI21",
    AionLabs: "AionLabs",
    Alibaba: "Alibaba",
    Ambient: "Ambient",
    Baidu: "Baidu",
    AmazonBedrock: "Amazon Bedrock",
    AmazonNova: "Amazon Nova",
    Anthropic: "Anthropic",
    ArceeAI: "Arcee AI",
    AtlasCloud: "AtlasCloud",
    Avian: "Avian",
    Azure: "Azure",
    BaseTen: "BaseTen",
    BytePlus: "BytePlus",
    BlackForestLabs: "Black Forest Labs",
    Cerebras: "Cerebras",
    Chutes: "Chutes",
    Cirrascale: "Cirrascale",
    Clarifai: "Clarifai",
    Cloudflare: "Cloudflare",
    Cohere: "Cohere",
    Crusoe: "Crusoe",
    DeepInfra: "DeepInfra",
    DeepSeek: "DeepSeek",
    DekaLLM: "DekaLLM",
    Featherless: "Featherless",
    Fireworks: "Fireworks",
    Friendli: "Friendli",
    GMICloud: "GMICloud",
    Google: "Google",
    GoogleAIStudio: "Google AI Studio",
    Groq: "Groq",
    Hyperbolic: "Hyperbolic",
    Inception: "Inception",
    Inceptron: "Inceptron",
    InferenceNet: "InferenceNet",
    Ionstream: "Ionstream",
    Infermatic: "Infermatic",
    IoNet: "Io Net",
    Inflection: "Inflection",
    Liquid: "Liquid",
    Mara: "Mara",
    Mancer2: "Mancer 2",
    Minimax: "Minimax",
    ModelRun: "ModelRun",
    Mistral: "Mistral",
    Modular: "Modular",
    MoonshotAI: "Moonshot AI",
    Morph: "Morph",
    NCompass: "NCompass",
    Nebius: "Nebius",
    NextBit: "NextBit",
    Novita: "Novita",
    Nvidia: "Nvidia",
    OpenAI: "OpenAI",
    OpenInference: "OpenInference",
    Parasail: "Parasail",
    Perplexity: "Perplexity",
    Phala: "Phala",
    Recraft: "Recraft",
    Reka: "Reka",
    Relace: "Relace",
    SambaNova: "SambaNova",
    Seed: "Seed",
    SiliconFlow: "SiliconFlow",
    Sourceful: "Sourceful",
    StepFun: "StepFun",
    Stealth: "Stealth",
    StreamLake: "StreamLake",
    Switchpoint: "Switchpoint",
    Together: "Together",
    Upstage: "Upstage",
    Venice: "Venice",
    WandB: "WandB",
    Xiaomi: "Xiaomi",
    XAI: "xAI",
    ZAi: "Z.AI",
    FakeProvider: "FakeProvider",
};
/** @internal */
export const ProviderResponseProviderName$inboundSchema = openEnums.inboundSchema(ProviderResponseProviderName);
/** @internal */
export const ProviderResponse$inboundSchema = z.object({
    endpoint_id: z.string().optional(),
    id: z.string().optional(),
    is_byok: z.boolean().optional(),
    latency: z.number().optional(),
    model_permaslug: z.string().optional(),
    provider_name: ProviderResponseProviderName$inboundSchema.optional(),
    status: z.nullable(z.number()),
}).transform((v) => {
    return remap$(v, {
        "endpoint_id": "endpointId",
        "is_byok": "isByok",
        "model_permaslug": "modelPermaslug",
        "provider_name": "providerName",
    });
});
export function providerResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ProviderResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ProviderResponse' from JSON`);
}
