// @openrouter/sdk/esm/models/publicpricing.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 0a44a1c3bab5
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
/** @internal */
export const PublicPricing$inboundSchema = z
    .object({
    audio: z.string().optional(),
    audio_output: z.string().optional(),
    completion: z.string(),
    discount: z.number().optional(),
    image: z.string().optional(),
    image_output: z.string().optional(),
    image_token: z.string().optional(),
    input_audio_cache: z.string().optional(),
    input_cache_read: z.string().optional(),
    input_cache_write: z.string().optional(),
    internal_reasoning: z.string().optional(),
    prompt: z.string(),
    request: z.string().optional(),
    web_search: z.string().optional(),
}).transform((v) => {
    return remap$(v, {
        "audio_output": "audioOutput",
        "image_output": "imageOutput",
        "image_token": "imageToken",
        "input_audio_cache": "inputAudioCache",
        "input_cache_read": "inputCacheRead",
        "input_cache_write": "inputCacheWrite",
        "internal_reasoning": "internalReasoning",
        "web_search": "webSearch",
    });
});
export function publicPricingFromJSON(jsonString) {
    return safeParse(jsonString, (x) => PublicPricing$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'PublicPricing' from JSON`);
}
