// @openrouter/sdk/esm/models/publicendpoint.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 396ce3186017
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import * as openEnums from "../types/enums.js";
import { EndpointStatus$inboundSchema, } from "./endpointstatus.js";
import { Parameter$inboundSchema } from "./parameter.js";
import { PercentileStats$inboundSchema, } from "./percentilestats.js";
import { ProviderName$inboundSchema } from "./providername.js";
export const PublicEndpointQuantization = {
    Int4: "int4",
    Int8: "int8",
    Fp4: "fp4",
    Fp6: "fp6",
    Fp8: "fp8",
    Fp16: "fp16",
    Bf16: "bf16",
    Fp32: "fp32",
    Unknown: "unknown",
};
/** @internal */
export const Pricing$inboundSchema = z.object({
    audio: z.string().optional(),
    audio_output: z.string().optional(),
    completion: z.string(),
    discount: z.number().optional(),
    image: z.string().optional(),
    image_output: z.string().optional(),
    image_token: z.string().optional(),
    input_audio_cache: z.string().optional(),
    input_cache_read: z.string().optional(),
    input_cache_write: z.string().optional(),
    internal_reasoning: z.string().optional(),
    prompt: z.string(),
    request: z.string().optional(),
    web_search: z.string().optional(),
}).transform((v) => {
    return remap$(v, {
        "audio_output": "audioOutput",
        "image_output": "imageOutput",
        "image_token": "imageToken",
        "input_audio_cache": "inputAudioCache",
        "input_cache_read": "inputCacheRead",
        "input_cache_write": "inputCacheWrite",
        "internal_reasoning": "internalReasoning",
        "web_search": "webSearch",
    });
});
export function pricingFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Pricing$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Pricing' from JSON`);
}
/** @internal */
export const PublicEndpointQuantization$inboundSchema = openEnums.inboundSchema(PublicEndpointQuantization);
/** @internal */
export const PublicEndpoint$inboundSchema = z.object({
    context_length: z.int(),
    latency_last_30m: z.nullable(PercentileStats$inboundSchema),
    max_completion_tokens: z.nullable(z.int()),
    max_prompt_tokens: z.nullable(z.int()),
    model_id: z.string(),
    model_name: z.string(),
    name: z.string(),
    pricing: z.lazy(() => Pricing$inboundSchema),
    provider_name: ProviderName$inboundSchema,
    quantization: z.nullable(PublicEndpointQuantization$inboundSchema),
    status: EndpointStatus$inboundSchema.optional(),
    supported_parameters: z.array(Parameter$inboundSchema),
    supports_implicit_caching: z.boolean(),
    tag: z.string(),
    throughput_last_30m: z.nullable(PercentileStats$inboundSchema),
    uptime_last_1d: z.nullable(z.number()),
    uptime_last_30m: z.nullable(z.number()),
    uptime_last_5m: z.nullable(z.number()),
}).transform((v) => {
    return remap$(v, {
        "context_length": "contextLength",
        "latency_last_30m": "latencyLast30m",
        "max_completion_tokens": "maxCompletionTokens",
        "max_prompt_tokens": "maxPromptTokens",
        "model_id": "modelId",
        "model_name": "modelName",
        "provider_name": "providerName",
        "supported_parameters": "supportedParameters",
        "supports_implicit_caching": "supportsImplicitCaching",
        "throughput_last_30m": "throughputLast30m",
        "uptime_last_1d": "uptimeLast1d",
        "uptime_last_30m": "uptimeLast30m",
        "uptime_last_5m": "uptimeLast5m",
    });
});
export function publicEndpointFromJSON(jsonString) {
    return safeParse(jsonString, (x) => PublicEndpoint$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'PublicEndpoint' from JSON`);
}
