// @openrouter/sdk/esm/models/reasoningdetailsummary.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 08e5d256b8c9
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { ReasoningFormat$inboundSchema, ReasoningFormat$outboundSchema, } from "./reasoningformat.js";
/** @internal */
export const ReasoningDetailSummary$inboundSchema = z.object({
    format: z.nullable(ReasoningFormat$inboundSchema).optional(),
    id: z.nullable(z.string()).optional(),
    index: z.int().optional(),
    summary: z.string(),
    type: z.literal("reasoning.summary"),
});
/** @internal */
export const ReasoningDetailSummary$outboundSchema = z.object({
    format: z.nullable(ReasoningFormat$outboundSchema).optional(),
    id: z.nullable(z.string()).optional(),
    index: z.int().optional(),
    summary: z.string(),
    type: z.literal("reasoning.summary"),
});
export function reasoningDetailSummaryToJSON(reasoningDetailSummary) {
    return JSON.stringify(ReasoningDetailSummary$outboundSchema.parse(reasoningDetailSummary));
}
export function reasoningDetailSummaryFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ReasoningDetailSummary$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ReasoningDetailSummary' from JSON`);
}
