// @openrouter/sdk/esm/models/reasoningtextcontent.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: e5520210e3cb
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
export const ReasoningTextContentType = {
    ReasoningText: "reasoning_text",
};
/** @internal */
export const ReasoningTextContentType$inboundSchema = z.enum(ReasoningTextContentType);
/** @internal */
export const ReasoningTextContentType$outboundSchema = ReasoningTextContentType$inboundSchema;
/** @internal */
export const ReasoningTextContent$inboundSchema = z.object({
    text: z.string(),
    type: ReasoningTextContentType$inboundSchema,
});
/** @internal */
export const ReasoningTextContent$outboundSchema = z.object({
    text: z.string(),
    type: ReasoningTextContentType$outboundSchema,
});
export function reasoningTextContentToJSON(reasoningTextContent) {
    return JSON.stringify(ReasoningTextContent$outboundSchema.parse(reasoningTextContent));
}
export function reasoningTextContentFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ReasoningTextContent$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ReasoningTextContent' from JSON`);
}
