// @openrouter/sdk/esm/models/reasoningdetailtext.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 8b6bd4c64a66
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { ReasoningFormat$inboundSchema, ReasoningFormat$outboundSchema, } from "./reasoningformat.js";
/** @internal */
export const ReasoningDetailText$inboundSchema = z.object({
    format: z.nullable(ReasoningFormat$inboundSchema).optional(),
    id: z.nullable(z.string()).optional(),
    index: z.int().optional(),
    signature: z.nullable(z.string()).optional(),
    text: z.nullable(z.string()).optional(),
    type: z.literal("reasoning.text"),
});
/** @internal */
export const ReasoningDetailText$outboundSchema = z.object({
    format: z.nullable(ReasoningFormat$outboundSchema).optional(),
    id: z.nullable(z.string()).optional(),
    index: z.int().optional(),
    signature: z.nullable(z.string()).optional(),
    text: z.nullable(z.string()).optional(),
    type: z.literal("reasoning.text"),
});
export function reasoningDetailTextToJSON(reasoningDetailText) {
    return JSON.stringify(ReasoningDetailText$outboundSchema.parse(reasoningDetailText));
}
export function reasoningDetailTextFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ReasoningDetailText$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ReasoningDetailText' from JSON`);
}
