// @openrouter/sdk/esm/models/reasoningdetailencrypted.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 2064e7820eab
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { ReasoningFormat$inboundSchema, ReasoningFormat$outboundSchema, } from "./reasoningformat.js";
/** @internal */
export const ReasoningDetailEncrypted$inboundSchema = z.object({
    data: z.string(),
    format: z.nullable(ReasoningFormat$inboundSchema).optional(),
    id: z.nullable(z.string()).optional(),
    index: z.int().optional(),
    type: z.literal("reasoning.encrypted"),
});
/** @internal */
export const ReasoningDetailEncrypted$outboundSchema = z.object({
    data: z.string(),
    format: z.nullable(ReasoningFormat$outboundSchema).optional(),
    id: z.nullable(z.string()).optional(),
    index: z.int().optional(),
    type: z.literal("reasoning.encrypted"),
});
export function reasoningDetailEncryptedToJSON(reasoningDetailEncrypted) {
    return JSON.stringify(ReasoningDetailEncrypted$outboundSchema.parse(reasoningDetailEncrypted));
}
export function reasoningDetailEncryptedFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ReasoningDetailEncrypted$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ReasoningDetailEncrypted' from JSON`);
}
