// @openrouter/sdk/esm/models/responseserrorfield.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 830ea3d3b590
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import * as openEnums from "../types/enums.js";
export const Code = {
    ServerError: "server_error",
    RateLimitExceeded: "rate_limit_exceeded",
    InvalidPrompt: "invalid_prompt",
    VectorStoreTimeout: "vector_store_timeout",
    InvalidImage: "invalid_image",
    InvalidImageFormat: "invalid_image_format",
    InvalidBase64Image: "invalid_base64_image",
    InvalidImageUrl: "invalid_image_url",
    ImageTooLarge: "image_too_large",
    ImageTooSmall: "image_too_small",
    ImageParseError: "image_parse_error",
    ImageContentPolicyViolation: "image_content_policy_violation",
    InvalidImageMode: "invalid_image_mode",
    ImageFileTooLarge: "image_file_too_large",
    UnsupportedImageMediaType: "unsupported_image_media_type",
    EmptyImageFile: "empty_image_file",
    FailedToDownloadImage: "failed_to_download_image",
    ImageFileNotFound: "image_file_not_found",
};
/** @internal */
export const Code$inboundSchema = openEnums
    .inboundSchema(Code);
/** @internal */
export const ResponsesErrorField$inboundSchema = z.object({
    code: Code$inboundSchema,
    message: z.string(),
});
export function responsesErrorFieldFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ResponsesErrorField$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ResponsesErrorField' from JSON`);
}
