// @openrouter/sdk/esm/models/responsesrequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 39799702f982
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import * as openEnums from "../types/enums.js";
import { ApplyPatchServerTool$outboundSchema, } from "./applypatchservertool.js";
import { AutoRouterPlugin$outboundSchema, } from "./autorouterplugin.js";
import { ChatSearchModelsServerTool$outboundSchema, } from "./chatsearchmodelsservertool.js";
import { CodeInterpreterServerTool$outboundSchema, } from "./codeinterpreterservertool.js";
import { CodexLocalShellTool$outboundSchema, } from "./codexlocalshelltool.js";
import { ComputerUseServerTool$outboundSchema, } from "./computeruseservertool.js";
import { ContextCompressionPlugin$outboundSchema, } from "./contextcompressionplugin.js";
import { CustomTool$outboundSchema, } from "./customtool.js";
import { DatetimeServerTool$outboundSchema, } from "./datetimeservertool.js";
import { FileParserPlugin$outboundSchema, } from "./fileparserplugin.js";
import { FileSearchServerTool$outboundSchema, } from "./filesearchservertool.js";
import { ImageConfig$outboundSchema, } from "./imageconfig.js";
import { ImageGenerationServerTool$outboundSchema, } from "./imagegenerationservertool.js";
import { ImageGenerationServerToolOpenRouter$outboundSchema, } from "./imagegenerationservertoolopenrouter.js";
import { InputsUnion$outboundSchema, } from "./inputsunion.js";
import { LegacyWebSearchServerTool$outboundSchema, } from "./legacywebsearchservertool.js";
import { McpServerTool$outboundSchema, } from "./mcpservertool.js";
import { ModerationPlugin$outboundSchema, } from "./moderationplugin.js";
import { OpenAIResponsesToolChoiceUnion$outboundSchema, } from "./openairesponsestoolchoiceunion.js";
import { OpenAIResponsesTruncation$outboundSchema, } from "./openairesponsestruncation.js";
import { OutputModalityEnum$outboundSchema, } from "./outputmodalityenum.js";
import { Preview20250311WebSearchServerTool$outboundSchema, } from "./preview20250311websearchservertool.js";
import { PreviewWebSearchServerTool$outboundSchema, } from "./previewwebsearchservertool.js";
import { ProviderPreferences$outboundSchema, } from "./providerpreferences.js";
import { ReasoningConfig$outboundSchema, } from "./reasoningconfig.js";
import { ResponseHealingPlugin$outboundSchema, } from "./responsehealingplugin.js";
import { ResponseIncludesEnum$outboundSchema, } from "./responseincludesenum.js";
import { ShellServerTool$outboundSchema, } from "./shellservertool.js";
import { StoredPromptTemplate$outboundSchema, } from "./storedprompttemplate.js";
import { TextExtendedConfig$outboundSchema, } from "./textextendedconfig.js";
import { TraceConfig$outboundSchema, } from "./traceconfig.js";
import { WebSearchPlugin$outboundSchema, } from "./websearchplugin.js";
import { WebSearchServerTool$outboundSchema, } from "./websearchservertool.js";
import { WebSearchServerToolOpenRouter$outboundSchema, } from "./websearchservertoolopenrouter.js";
export const ResponsesRequestServiceTier = {
    Auto: "auto",
    Default: "default",
    Flex: "flex",
    Priority: "priority",
    Scale: "scale",
};
/** @internal */
export const ResponsesRequestPlugin$outboundSchema = z.union([
    AutoRouterPlugin$outboundSchema,
    ContextCompressionPlugin$outboundSchema,
    FileParserPlugin$outboundSchema,
    ModerationPlugin$outboundSchema,
    ResponseHealingPlugin$outboundSchema,
    WebSearchPlugin$outboundSchema,
]);
export function responsesRequestPluginToJSON(responsesRequestPlugin) {
    return JSON.stringify(ResponsesRequestPlugin$outboundSchema.parse(responsesRequestPlugin));
}
/** @internal */
export const ResponsesRequestServiceTier$outboundSchema = openEnums.outboundSchema(ResponsesRequestServiceTier);
/** @internal */
export const ResponsesRequestToolFunction$outboundSchema = z.object({
    description: z.nullable(z.string()).optional(),
    name: z.string(),
    parameters: z.nullable(z.record(z.string(), z.nullable(z.any()))),
    strict: z.nullable(z.boolean()).optional(),
    type: z.literal("function"),
});
export function responsesRequestToolFunctionToJSON(responsesRequestToolFunction) {
    return JSON.stringify(ResponsesRequestToolFunction$outboundSchema.parse(responsesRequestToolFunction));
}
/** @internal */
export const ResponsesRequestToolUnion$outboundSchema = z.union([
    z.lazy(() => ResponsesRequestToolFunction$outboundSchema),
    PreviewWebSearchServerTool$outboundSchema,
    Preview20250311WebSearchServerTool$outboundSchema,
    LegacyWebSearchServerTool$outboundSchema,
    WebSearchServerTool$outboundSchema,
    FileSearchServerTool$outboundSchema,
    ComputerUseServerTool$outboundSchema,
    CodeInterpreterServerTool$outboundSchema,
    McpServerTool$outboundSchema,
    ImageGenerationServerTool$outboundSchema,
    CodexLocalShellTool$outboundSchema,
    ShellServerTool$outboundSchema,
    ApplyPatchServerTool$outboundSchema,
    CustomTool$outboundSchema,
    DatetimeServerTool$outboundSchema.and(z.object({ type: z.literal("openrouter:datetime") })),
    ImageGenerationServerToolOpenRouter$outboundSchema.and(z.object({ type: z.literal("openrouter:image_generation") })),
    ChatSearchModelsServerTool$outboundSchema.and(z.object({ type: z.literal("openrouter:experimental__search_models") })),
    WebSearchServerToolOpenRouter$outboundSchema,
]);
export function responsesRequestToolUnionToJSON(responsesRequestToolUnion) {
    return JSON.stringify(ResponsesRequestToolUnion$outboundSchema.parse(responsesRequestToolUnion));
}
/** @internal */
export const ResponsesRequest$outboundSchema = z.object({
    background: z.nullable(z.boolean()).optional(),
    frequencyPenalty: z.nullable(z.number()).optional(),
    imageConfig: z.record(z.string(), ImageConfig$outboundSchema).optional(),
    include: z.nullable(z.array(ResponseIncludesEnum$outboundSchema)).optional(),
    input: InputsUnion$outboundSchema.optional(),
    instructions: z.nullable(z.string()).optional(),
    maxOutputTokens: z.nullable(z.int()).optional(),
    maxToolCalls: z.nullable(z.int()).optional(),
    metadata: z.nullable(z.record(z.string(), z.string())).optional(),
    modalities: z.array(OutputModalityEnum$outboundSchema).optional(),
    model: z.string().optional(),
    models: z.array(z.string()).optional(),
    parallelToolCalls: z.nullable(z.boolean()).optional(),
    plugins: z.array(z.union([
        AutoRouterPlugin$outboundSchema,
        ContextCompressionPlugin$outboundSchema,
        FileParserPlugin$outboundSchema,
        ModerationPlugin$outboundSchema,
        ResponseHealingPlugin$outboundSchema,
        WebSearchPlugin$outboundSchema,
    ])).optional(),
    presencePenalty: z.nullable(z.number()).optional(),
    previousResponseId: z.nullable(z.string()).optional(),
    prompt: z.nullable(StoredPromptTemplate$outboundSchema).optional(),
    promptCacheKey: z.nullable(z.string()).optional(),
    provider: z.nullable(ProviderPreferences$outboundSchema).optional(),
    reasoning: z.nullable(ReasoningConfig$outboundSchema).optional(),
    safetyIdentifier: z.nullable(z.string()).optional(),
    serviceTier: z.nullable(ResponsesRequestServiceTier$outboundSchema.default("auto")),
    sessionId: z.string().optional(),
    store: z.literal(false).default(false),
    stream: z.boolean().default(false),
    temperature: z.nullable(z.number()).optional(),
    text: TextExtendedConfig$outboundSchema.optional(),
    toolChoice: OpenAIResponsesToolChoiceUnion$outboundSchema.optional(),
    tools: z.array(z.union([
        z.lazy(() => ResponsesRequestToolFunction$outboundSchema),
        PreviewWebSearchServerTool$outboundSchema,
        Preview20250311WebSearchServerTool$outboundSchema,
        LegacyWebSearchServerTool$outboundSchema,
        WebSearchServerTool$outboundSchema,
        FileSearchServerTool$outboundSchema,
        ComputerUseServerTool$outboundSchema,
        CodeInterpreterServerTool$outboundSchema,
        McpServerTool$outboundSchema,
        ImageGenerationServerTool$outboundSchema,
        CodexLocalShellTool$outboundSchema,
        ShellServerTool$outboundSchema,
        ApplyPatchServerTool$outboundSchema,
        CustomTool$outboundSchema,
        DatetimeServerTool$outboundSchema.and(z.object({ type: z.literal("openrouter:datetime") })),
        ImageGenerationServerToolOpenRouter$outboundSchema.and(z.object({ type: z.literal("openrouter:image_generation") })),
        ChatSearchModelsServerTool$outboundSchema.and(z.object({ type: z.literal("openrouter:experimental__search_models") })),
        WebSearchServerToolOpenRouter$outboundSchema,
    ])).optional(),
    topK: z.int().optional(),
    topLogprobs: z.nullable(z.int()).optional(),
    topP: z.nullable(z.number()).optional(),
    trace: TraceConfig$outboundSchema.optional(),
    truncation: z.nullable(OpenAIResponsesTruncation$outboundSchema).optional(),
    user: z.string().optional(),
}).transform((v) => {
    return remap$(v, {
        frequencyPenalty: "frequency_penalty",
        imageConfig: "image_config",
        maxOutputTokens: "max_output_tokens",
        maxToolCalls: "max_tool_calls",
        parallelToolCalls: "parallel_tool_calls",
        presencePenalty: "presence_penalty",
        previousResponseId: "previous_response_id",
        promptCacheKey: "prompt_cache_key",
        safetyIdentifier: "safety_identifier",
        serviceTier: "service_tier",
        sessionId: "session_id",
        toolChoice: "tool_choice",
        topK: "top_k",
        topLogprobs: "top_logprobs",
        topP: "top_p",
    });
});
export function responsesRequestToJSON(responsesRequest) {
    return JSON.stringify(ResponsesRequest$outboundSchema.parse(responsesRequest));
}
