// @openrouter/sdk/esm/models/responseoutputtext.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 2d5e61e53c46
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { OpenAIResponsesAnnotation$inboundSchema, OpenAIResponsesAnnotation$outboundSchema, } from "./openairesponsesannotation.js";
/** @internal */
export const ResponseOutputTextTopLogprob$inboundSchema = z.object({
    bytes: z.array(z.int()),
    logprob: z.number(),
    token: z.string(),
});
/** @internal */
export const ResponseOutputTextTopLogprob$outboundSchema = z.object({
    bytes: z.array(z.int()),
    logprob: z.number(),
    token: z.string(),
});
export function responseOutputTextTopLogprobToJSON(responseOutputTextTopLogprob) {
    return JSON.stringify(ResponseOutputTextTopLogprob$outboundSchema.parse(responseOutputTextTopLogprob));
}
export function responseOutputTextTopLogprobFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ResponseOutputTextTopLogprob$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ResponseOutputTextTopLogprob' from JSON`);
}
/** @internal */
export const Logprob$inboundSchema = z.object({
    bytes: z.array(z.int()),
    logprob: z.number(),
    token: z.string(),
    top_logprobs: z.array(z.lazy(() => ResponseOutputTextTopLogprob$inboundSchema)),
}).transform((v) => {
    return remap$(v, {
        "top_logprobs": "topLogprobs",
    });
});
/** @internal */
export const Logprob$outboundSchema = z
    .object({
    bytes: z.array(z.int()),
    logprob: z.number(),
    token: z.string(),
    topLogprobs: z.array(z.lazy(() => ResponseOutputTextTopLogprob$outboundSchema)),
}).transform((v) => {
    return remap$(v, {
        topLogprobs: "top_logprobs",
    });
});
export function logprobToJSON(logprob) {
    return JSON.stringify(Logprob$outboundSchema.parse(logprob));
}
export function logprobFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Logprob$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Logprob' from JSON`);
}
/** @internal */
export const ResponseOutputText$inboundSchema = z.object({
    annotations: z.array(OpenAIResponsesAnnotation$inboundSchema).optional(),
    logprobs: z.array(z.lazy(() => Logprob$inboundSchema)).optional(),
    text: z.string(),
    type: z.literal("output_text"),
});
/** @internal */
export const ResponseOutputText$outboundSchema = z.object({
    annotations: z.array(OpenAIResponsesAnnotation$outboundSchema).optional(),
    logprobs: z.array(z.lazy(() => Logprob$outboundSchema)).optional(),
    text: z.string(),
    type: z.literal("output_text"),
});
export function responseOutputTextToJSON(responseOutputText) {
    return JSON.stringify(ResponseOutputText$outboundSchema.parse(responseOutputText));
}
export function responseOutputTextFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ResponseOutputText$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ResponseOutputText' from JSON`);
}
