// @openrouter/sdk/esm/models/storedprompttemplate.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 4a3405746141
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import { InputFile$inboundSchema, InputFile$outboundSchema, } from "./inputfile.js";
import { InputImage$inboundSchema, InputImage$outboundSchema, } from "./inputimage.js";
import { InputText$inboundSchema, InputText$outboundSchema, } from "./inputtext.js";
/** @internal */
export const Variables$inboundSchema = z.union([
    InputText$inboundSchema,
    InputImage$inboundSchema,
    InputFile$inboundSchema,
    z.string(),
]);
/** @internal */
export const Variables$outboundSchema = z.union([
    InputText$outboundSchema,
    InputImage$outboundSchema,
    InputFile$outboundSchema,
    z.string(),
]);
export function variablesToJSON(variables) {
    return JSON.stringify(Variables$outboundSchema.parse(variables));
}
export function variablesFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Variables$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Variables' from JSON`);
}
/** @internal */
export const StoredPromptTemplate$inboundSchema = z.object({
    id: z.string(),
    variables: z.nullable(z.record(z.string(), z.union([
        InputText$inboundSchema,
        InputImage$inboundSchema,
        InputFile$inboundSchema,
        z.string(),
    ]))).optional(),
});
/** @internal */
export const StoredPromptTemplate$outboundSchema = z.object({
    id: z.string(),
    variables: z.nullable(z.record(z.string(), z.union([
        InputText$outboundSchema,
        InputImage$outboundSchema,
        InputFile$outboundSchema,
        z.string(),
    ]))).optional(),
});
export function storedPromptTemplateToJSON(storedPromptTemplate) {
    return JSON.stringify(StoredPromptTemplate$outboundSchema.parse(storedPromptTemplate));
}
export function storedPromptTemplateFromJSON(jsonString) {
    return safeParse(jsonString, (x) => StoredPromptTemplate$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'StoredPromptTemplate' from JSON`);
}
