// @openrouter/sdk/esm/models/textdoneevent.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 006e3645c287
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { StreamLogprob$inboundSchema } from "./streamlogprob.js";
/** @internal */
export const TextDoneEvent$inboundSchema = z
    .object({
    content_index: z.int(),
    item_id: z.string(),
    logprobs: z.array(StreamLogprob$inboundSchema),
    output_index: z.int(),
    sequence_number: z.int(),
    text: z.string(),
    type: z.literal("response.output_text.done"),
}).transform((v) => {
    return remap$(v, {
        "content_index": "contentIndex",
        "item_id": "itemId",
        "output_index": "outputIndex",
        "sequence_number": "sequenceNumber",
    });
});
export function textDoneEventFromJSON(jsonString) {
    return safeParse(jsonString, (x) => TextDoneEvent$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'TextDoneEvent' from JSON`);
}
