// @openrouter/sdk/esm/models/textdeltaevent.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 920ea48e0f90
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import { StreamLogprob$inboundSchema } from "./streamlogprob.js";
/** @internal */
export const TextDeltaEvent$inboundSchema = z.object({
    content_index: z.int(),
    delta: z.string(),
    item_id: z.string(),
    logprobs: z.array(StreamLogprob$inboundSchema),
    output_index: z.int(),
    sequence_number: z.int(),
    type: z.literal("response.output_text.delta"),
}).transform((v) => {
    return remap$(v, {
        "content_index": "contentIndex",
        "item_id": "itemId",
        "output_index": "outputIndex",
        "sequence_number": "sequenceNumber",
    });
});
export function textDeltaEventFromJSON(jsonString) {
    return safeParse(jsonString, (x) => TextDeltaEvent$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'TextDeltaEvent' from JSON`);
}
