// @openrouter/sdk/esm/models/textextendedconfig.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: e264e7981200
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
import * as openEnums from "../types/enums.js";
import { Formats$inboundSchema, Formats$outboundSchema, } from "./formats.js";
export const Verbosity = {
    Low: "low",
    Medium: "medium",
    High: "high",
    Xhigh: "xhigh",
    Max: "max",
};
/** @internal */
export const Verbosity$inboundSchema = openEnums
    .inboundSchema(Verbosity);
/** @internal */
export const Verbosity$outboundSchema = openEnums
    .outboundSchema(Verbosity);
/** @internal */
export const TextExtendedConfig$inboundSchema = z.object({
    format: Formats$inboundSchema.optional(),
    verbosity: z.nullable(Verbosity$inboundSchema).optional(),
});
/** @internal */
export const TextExtendedConfig$outboundSchema = z.object({
    format: Formats$outboundSchema.optional(),
    verbosity: z.nullable(Verbosity$outboundSchema).optional(),
});
export function textExtendedConfigToJSON(textExtendedConfig) {
    return JSON.stringify(TextExtendedConfig$outboundSchema.parse(textExtendedConfig));
}
export function textExtendedConfigFromJSON(jsonString) {
    return safeParse(jsonString, (x) => TextExtendedConfig$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'TextExtendedConfig' from JSON`);
}
