// @openrouter/sdk/esm/models/toolchoiceallowed.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 82ee236509d6
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
export const ModeRequired = {
    Required: "required",
};
export const ModeAuto = {
    Auto: "auto",
};
export const ToolChoiceAllowedType = {
    AllowedTools: "allowed_tools",
};
/** @internal */
export const ModeRequired$inboundSchema = z
    .enum(ModeRequired);
/** @internal */
export const ModeRequired$outboundSchema = ModeRequired$inboundSchema;
/** @internal */
export const ModeAuto$inboundSchema = z.enum(ModeAuto);
/** @internal */
export const ModeAuto$outboundSchema = ModeAuto$inboundSchema;
/** @internal */
export const Mode$inboundSchema = z.union([
    ModeAuto$inboundSchema,
    ModeRequired$inboundSchema,
]);
/** @internal */
export const Mode$outboundSchema = z.union([
    ModeAuto$outboundSchema,
    ModeRequired$outboundSchema,
]);
export function modeToJSON(mode) {
    return JSON.stringify(Mode$outboundSchema.parse(mode));
}
export function modeFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Mode$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Mode' from JSON`);
}
/** @internal */
export const ToolChoiceAllowedType$inboundSchema = z.enum(ToolChoiceAllowedType);
/** @internal */
export const ToolChoiceAllowedType$outboundSchema = ToolChoiceAllowedType$inboundSchema;
/** @internal */
export const ToolChoiceAllowed$inboundSchema = z.object({
    mode: z.union([ModeAuto$inboundSchema, ModeRequired$inboundSchema]),
    tools: z.array(z.record(z.string(), z.nullable(z.any()))),
    type: ToolChoiceAllowedType$inboundSchema,
});
/** @internal */
export const ToolChoiceAllowed$outboundSchema = z.object({
    mode: z.union([ModeAuto$outboundSchema, ModeRequired$outboundSchema]),
    tools: z.array(z.record(z.string(), z.nullable(z.any()))),
    type: ToolChoiceAllowedType$outboundSchema,
});
export function toolChoiceAllowedToJSON(toolChoiceAllowed) {
    return JSON.stringify(ToolChoiceAllowed$outboundSchema.parse(toolChoiceAllowed));
}
export function toolChoiceAllowedFromJSON(jsonString) {
    return safeParse(jsonString, (x) => ToolChoiceAllowed$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ToolChoiceAllowed' from JSON`);
}
