// @openrouter/sdk/esm/models/traceconfig.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 50ce6e8a7bcd
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
/** @internal */
export const TraceConfig$outboundSchema = z.object({
    generationName: z.string().optional(),
    parentSpanId: z.string().optional(),
    spanName: z.string().optional(),
    traceId: z.string().optional(),
    traceName: z.string().optional(),
    additionalProperties: z.record(z.string(), z.nullable(z.any())).optional(),
}).transform((v) => {
    return {
        ...v.additionalProperties,
        ...remap$(v, {
            generationName: "generation_name",
            parentSpanId: "parent_span_id",
            spanName: "span_name",
            traceId: "trace_id",
            traceName: "trace_name",
            additionalProperties: null,
        }),
    };
});
export function traceConfigToJSON(traceConfig) {
    return JSON.stringify(TraceConfig$outboundSchema.parse(traceConfig));
}
