// @openrouter/sdk/esm/models/updateguardrailrequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 8ae5b029f2d5
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { GuardrailInterval$outboundSchema, } from "./guardrailinterval.js";
/** @internal */
export const UpdateGuardrailRequest$outboundSchema = z.object({
    allowedModels: z.nullable(z.array(z.string())).optional(),
    allowedProviders: z.nullable(z.array(z.string())).optional(),
    description: z.nullable(z.string()).optional(),
    enforceZdr: z.nullable(z.boolean()).optional(),
    ignoredModels: z.nullable(z.array(z.string())).optional(),
    ignoredProviders: z.nullable(z.array(z.string())).optional(),
    limitUsd: z.nullable(z.number()).optional(),
    name: z.string().optional(),
    resetInterval: z.nullable(GuardrailInterval$outboundSchema).optional(),
}).transform((v) => {
    return remap$(v, {
        allowedModels: "allowed_models",
        allowedProviders: "allowed_providers",
        enforceZdr: "enforce_zdr",
        ignoredModels: "ignored_models",
        ignoredProviders: "ignored_providers",
        limitUsd: "limit_usd",
        resetInterval: "reset_interval",
    });
});
export function updateGuardrailRequestToJSON(updateGuardrailRequest) {
    return JSON.stringify(UpdateGuardrailRequest$outboundSchema.parse(updateGuardrailRequest));
}
