// @openrouter/sdk/esm/models/usage.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 7d556ab96faf
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
/** @internal */
export const InputTokensDetails$inboundSchema = z.object({
    cached_tokens: z.int(),
}).transform((v) => {
    return remap$(v, {
        "cached_tokens": "cachedTokens",
    });
});
export function inputTokensDetailsFromJSON(jsonString) {
    return safeParse(jsonString, (x) => InputTokensDetails$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'InputTokensDetails' from JSON`);
}
/** @internal */
export const OutputTokensDetails$inboundSchema = z.object({
    reasoning_tokens: z.int(),
}).transform((v) => {
    return remap$(v, {
        "reasoning_tokens": "reasoningTokens",
    });
});
export function outputTokensDetailsFromJSON(jsonString) {
    return safeParse(jsonString, (x) => OutputTokensDetails$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'OutputTokensDetails' from JSON`);
}
/** @internal */
export const CostDetails$inboundSchema = z
    .object({
    upstream_inference_cost: z.nullable(z.number()).optional(),
    upstream_inference_input_cost: z.number(),
    upstream_inference_output_cost: z.number(),
}).transform((v) => {
    return remap$(v, {
        "upstream_inference_cost": "upstreamInferenceCost",
        "upstream_inference_input_cost": "upstreamInferenceInputCost",
        "upstream_inference_output_cost": "upstreamInferenceOutputCost",
    });
});
export function costDetailsFromJSON(jsonString) {
    return safeParse(jsonString, (x) => CostDetails$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CostDetails' from JSON`);
}
/** @internal */
export const Usage$inboundSchema = z.object({
    input_tokens: z.int(),
    input_tokens_details: z.lazy(() => InputTokensDetails$inboundSchema),
    output_tokens: z.int(),
    output_tokens_details: z.lazy(() => OutputTokensDetails$inboundSchema),
    total_tokens: z.int(),
    cost: z.nullable(z.number()).optional(),
    cost_details: z.lazy(() => CostDetails$inboundSchema).optional(),
    is_byok: z.boolean().optional(),
}).transform((v) => {
    return remap$(v, {
        "input_tokens": "inputTokens",
        "input_tokens_details": "inputTokensDetails",
        "output_tokens": "outputTokens",
        "output_tokens_details": "outputTokensDetails",
        "total_tokens": "totalTokens",
        "cost_details": "costDetails",
        "is_byok": "isByok",
    });
});
export function usageFromJSON(jsonString) {
    return safeParse(jsonString, (x) => Usage$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Usage' from JSON`);
}
