// @openrouter/sdk/esm/models/videomodel.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 3761c095ec55
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import * as openEnums from "../types/enums.js";
export const SupportedAspectRatio = {
    OneHundredAndSixtyNine: "16:9",
    NineHundredAndSixteen: "9:16",
    Eleven: "1:1",
    FortyThree: "4:3",
    ThirtyFour: "3:4",
    TwoHundredAndNineteen: "21:9",
    NineHundredAndTwentyOne: "9:21",
};
export const SupportedFrameImage = {
    FirstFrame: "first_frame",
    LastFrame: "last_frame",
};
export const SupportedResolution = {
    FourHundredAndEightyp: "480p",
    SevenHundredAndTwentyp: "720p",
    OneThousandAndEightyp: "1080p",
    OneK: "1K",
    TwoK: "2K",
    FourK: "4K",
};
export const SupportedSize = {
    FourHundredAndEightyx480: "480x480",
    FourHundredAndEightyx640: "480x640",
    FourHundredAndEightyx854: "480x854",
    FourHundredAndEightyx1120: "480x1120",
    SixHundredAndFortyx480: "640x480",
    SevenHundredAndTwentyx720: "720x720",
    SevenHundredAndTwentyx960: "720x960",
    SevenHundredAndTwentyx1280: "720x1280",
    SevenHundredAndTwentyx1680: "720x1680",
    EightHundredAndFiftyFourx480: "854x480",
    NineHundredAndSixtyx720: "960x720",
    OneThousandAndEightyx1080: "1080x1080",
    OneThousandAndEightyx1440: "1080x1440",
    OneThousandAndEightyx1920: "1080x1920",
    OneThousandAndEightyx2520: "1080x2520",
    OneThousandOneHundredAndTwentyx480: "1120x480",
    OneThousandTwoHundredAndEightyx720: "1280x720",
    OneThousandFourHundredAndFortyx1080: "1440x1080",
    OneThousandSixHundredAndEightyx720: "1680x720",
    OneThousandNineHundredAndTwentyx1080: "1920x1080",
    TwoThousandOneHundredAndSixtyx2160: "2160x2160",
    TwoThousandOneHundredAndSixtyx2880: "2160x2880",
    TwoThousandOneHundredAndSixtyx3840: "2160x3840",
    TwoThousandOneHundredAndSixtyx5040: "2160x5040",
    TwoThousandFiveHundredAndTwentyx1080: "2520x1080",
    TwoThousandEightHundredAndEightyx2160: "2880x2160",
    ThreeThousandEightHundredAndFortyx2160: "3840x2160",
    FiveThousandAndFortyx2160: "5040x2160",
};
/** @internal */
export const SupportedAspectRatio$inboundSchema = openEnums.inboundSchema(SupportedAspectRatio);
/** @internal */
export const SupportedFrameImage$inboundSchema = openEnums.inboundSchema(SupportedFrameImage);
/** @internal */
export const SupportedResolution$inboundSchema = openEnums.inboundSchema(SupportedResolution);
/** @internal */
export const SupportedSize$inboundSchema = openEnums.inboundSchema(SupportedSize);
/** @internal */
export const VideoModel$inboundSchema = z
    .object({
    allowed_passthrough_parameters: z.array(z.string()),
    canonical_slug: z.string(),
    created: z.int(),
    description: z.string().optional(),
    generate_audio: z.nullable(z.boolean()),
    hugging_face_id: z.nullable(z.string()).optional(),
    id: z.string(),
    name: z.string(),
    pricing_skus: z.nullable(z.record(z.string(), z.string())).optional(),
    seed: z.nullable(z.boolean()),
    supported_aspect_ratios: z.nullable(z.array(SupportedAspectRatio$inboundSchema)),
    supported_durations: z.nullable(z.array(z.int())),
    supported_frame_images: z.nullable(z.array(SupportedFrameImage$inboundSchema)),
    supported_resolutions: z.nullable(z.array(SupportedResolution$inboundSchema)),
    supported_sizes: z.nullable(z.array(SupportedSize$inboundSchema)),
}).transform((v) => {
    return remap$(v, {
        "allowed_passthrough_parameters": "allowedPassthroughParameters",
        "canonical_slug": "canonicalSlug",
        "generate_audio": "generateAudio",
        "hugging_face_id": "huggingFaceId",
        "pricing_skus": "pricingSkus",
        "supported_aspect_ratios": "supportedAspectRatios",
        "supported_durations": "supportedDurations",
        "supported_frame_images": "supportedFrameImages",
        "supported_resolutions": "supportedResolutions",
        "supported_sizes": "supportedSizes",
    });
});
export function videoModelFromJSON(jsonString) {
    return safeParse(jsonString, (x) => VideoModel$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'VideoModel' from JSON`);
}
