// @openrouter/sdk/esm/models/videogenerationresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 0a96c6ee9eea
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
import * as openEnums from "../types/enums.js";
import { VideoGenerationUsage$inboundSchema, } from "./videogenerationusage.js";
export const VideoGenerationResponseStatus = {
    Pending: "pending",
    InProgress: "in_progress",
    Completed: "completed",
    Failed: "failed",
    Cancelled: "cancelled",
    Expired: "expired",
};
/** @internal */
export const VideoGenerationResponseStatus$inboundSchema = openEnums.inboundSchema(VideoGenerationResponseStatus);
/** @internal */
export const VideoGenerationResponse$inboundSchema = z.object({
    error: z.string().optional(),
    generation_id: z.string().optional(),
    id: z.string(),
    polling_url: z.string(),
    status: VideoGenerationResponseStatus$inboundSchema,
    unsigned_urls: z.array(z.string()).optional(),
    usage: VideoGenerationUsage$inboundSchema.optional(),
}).transform((v) => {
    return remap$(v, {
        "generation_id": "generationId",
        "polling_url": "pollingUrl",
        "unsigned_urls": "unsignedUrls",
    });
});
export function videoGenerationResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => VideoGenerationResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'VideoGenerationResponse' from JSON`);
}
