// @openrouter/sdk/esm/models/websearchservertoolopenrouter.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 99dffefc1c11
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
/** @internal */
export const ParametersT$outboundSchema = z.object({
    maxResults: z.int().optional(),
    maxTotalResults: z.int().optional(),
}).transform((v) => {
    return remap$(v, {
        maxResults: "max_results",
        maxTotalResults: "max_total_results",
    });
});
export function parametersToJSON(parametersT) {
    return JSON.stringify(ParametersT$outboundSchema.parse(parametersT));
}
/** @internal */
export const WebSearchServerToolOpenRouter$outboundSchema = z.object({
    parameters: z.lazy(() => ParametersT$outboundSchema).optional(),
    type: z.literal("openrouter:web_search"),
});
export function webSearchServerToolOpenRouterToJSON(webSearchServerToolOpenRouter) {
    return JSON.stringify(WebSearchServerToolOpenRouter$outboundSchema.parse(webSearchServerToolOpenRouter));
}
