// @openrouter/sdk/esm/models/websearchsource.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 2a64353a399c
 */
import * as z from "zod/v4";
import { safeParse } from "../lib/schemas.js";
export const WebSearchSourceType = {
    Url: "url",
};
/** @internal */
export const WebSearchSourceType$inboundSchema = z.enum(WebSearchSourceType);
/** @internal */
export const WebSearchSourceType$outboundSchema = WebSearchSourceType$inboundSchema;
/** @internal */
export const WebSearchSource$inboundSchema = z.object({
    type: WebSearchSourceType$inboundSchema,
    url: z.string(),
});
/** @internal */
export const WebSearchSource$outboundSchema = z.object({
    type: WebSearchSourceType$outboundSchema,
    url: z.string(),
});
export function webSearchSourceToJSON(webSearchSource) {
    return JSON.stringify(WebSearchSource$outboundSchema.parse(webSearchSource));
}
export function webSearchSourceFromJSON(jsonString) {
    return safeParse(jsonString, (x) => WebSearchSource$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'WebSearchSource' from JSON`);
}
