// @openrouter/sdk/esm/models/websearchdomainfilter.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 827204408f55
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { safeParse } from "../lib/schemas.js";
/** @internal */
export const WebSearchDomainFilter$inboundSchema = z.object({
    allowed_domains: z.nullable(z.array(z.string())).optional(),
    excluded_domains: z.nullable(z.array(z.string())).optional(),
}).transform((v) => {
    return remap$(v, {
        "allowed_domains": "allowedDomains",
        "excluded_domains": "excludedDomains",
    });
});
/** @internal */
export const WebSearchDomainFilter$outboundSchema = z.object({
    allowedDomains: z.nullable(z.array(z.string())).optional(),
    excludedDomains: z.nullable(z.array(z.string())).optional(),
}).transform((v) => {
    return remap$(v, {
        allowedDomains: "allowed_domains",
        excludedDomains: "excluded_domains",
    });
});
export function webSearchDomainFilterToJSON(webSearchDomainFilter) {
    return JSON.stringify(WebSearchDomainFilter$outboundSchema.parse(webSearchDomainFilter));
}
export function webSearchDomainFilterFromJSON(jsonString) {
    return safeParse(jsonString, (x) => WebSearchDomainFilter$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'WebSearchDomainFilter' from JSON`);
}
