// @openrouter/sdk/esm/models/websearchconfig.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: d43f37b9a707
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import { SearchQualityLevel$outboundSchema, } from "./searchqualitylevel.js";
import { WebSearchEngineEnum$outboundSchema, } from "./websearchengineenum.js";
import { WebSearchUserLocationServerTool$outboundSchema, } from "./websearchuserlocationservertool.js";
/** @internal */
export const WebSearchConfig$outboundSchema = z.object({
    allowedDomains: z.array(z.string()).optional(),
    engine: WebSearchEngineEnum$outboundSchema.optional(),
    excludedDomains: z.array(z.string()).optional(),
    maxResults: z.int().optional(),
    maxTotalResults: z.int().optional(),
    searchContextSize: SearchQualityLevel$outboundSchema.optional(),
    userLocation: WebSearchUserLocationServerTool$outboundSchema.optional(),
}).transform((v) => {
    return remap$(v, {
        allowedDomains: "allowed_domains",
        excludedDomains: "excluded_domains",
        maxResults: "max_results",
        maxTotalResults: "max_total_results",
        searchContextSize: "search_context_size",
        userLocation: "user_location",
    });
});
export function webSearchConfigToJSON(webSearchConfig) {
    return JSON.stringify(WebSearchConfig$outboundSchema.parse(webSearchConfig));
}
